---
title: "Clustering of TF isoform expression patterns"
author: "Sviatoslav Sidorov"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
   html_document:
     code_folding: hide
     collapsed: no
     fig_align: center
     fig_caption: yes
     highlight: haddock
     keep_md: yes
     number_sections: yes
     smooth_scroll: no
     toc: yes
     toc_depth: 3
     toc_float: yes
---

```{r setup, include=FALSE}
require("knitr")
opts_knit$set(root.dir = "/home/rstudio")
knitr::opts_chunk$set(echo = TRUE)
stringsAsFactors = F

library(htmltools)
library(dplyr) 
library(ggplot2)
library(ggrepel)
library(tibble)
library(stringr)
library(magrittr)
library(kableExtra)
library(cluster)
library(fpc)
library(gridExtra)
library(ComplexHeatmap)
library(rlist)
library(tictoc)
library(hash)
library(circlize)
library(tidyr)
```

## Remove overlapping InterPro entries (need to move into the domain analysis notebook)

Find heavily overlapping InterPro entries:

```{r, include=T}
overlap.cutoff = 0.7

find_overlapping_iprs = function(tr.df) {
  tr.df.matches = tr.df %>%
    dplyr::select(ipr_accession, ipr_description, int_start, int_stop, domain_type)

  if (is.na(tr.df.matches[1, "ipr_accession"]) | (nrow(tr.df.matches) == 1)) {
    return(c(""))
  }
  
  result = c()
  i = 1
  i.max = nrow(tr.df.matches) - 1
  j.max = i.max + 1
  while (i <= i.max) {
    overlapped.iprs = ""
    j = i + 1
    while (j <= j.max) {
      i.start = tr.df.matches[i, "int_start"]
      i.stop = tr.df.matches[i, "int_stop"]
      j.start = tr.df.matches[j, "int_start"]
      j.stop = tr.df.matches[j, "int_stop"]
      i.length = i.stop - i.start + 1
      j.length = j.stop - j.start + 1
      ij.overlap.length = min(i.stop, j.stop) - max(i.start, j.start) + 1
      if ((ij.overlap.length / i.length >= overlap.cutoff) | 
          (ij.overlap.length / j.length >= overlap.cutoff)) {
        j.ipr = tr.df.matches[j, "ipr_accession"]
        j.descr = tr.df.matches[j, "ipr_description"]
        j.domain.type = tr.df.matches[j, "domain_type"]
        is.overlap.reciprocal = ((ij.overlap.length / i.length >= overlap.cutoff) & 
                                 (ij.overlap.length / j.length >= overlap.cutoff))
        overlapped.iprs = paste0(overlapped.iprs,
                                 ifelse(overlapped.iprs == "", "", ";"),
                                 paste0(j.ipr, "_", j.descr, "_", 
                                        j.start, "_", j.stop, "_", 
                                        j.domain.type, "_", is.overlap.reciprocal))
      }
      j = j + 1
    }
    result = c(result, overlapped.iprs)
    i = i + 1
  }
  result = c(result, "")
  
  return(result)
}

tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl = read.table(file = "data/results/domain_analysis/tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl.tsv",
            sep = "\t",
            quote = "\"",
            header = T)

overlapping_iprs_table = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  dplyr::select(ensembl_gene_id, ensembl_transcript_id, ipr_accession, ipr_description, int_start, int_stop, domain_type) %>%
  distinct() %>%
  group_by(ensembl_gene_id, ensembl_transcript_id) %>%
  do(mutate(., overlapping_iprs = find_overlapping_iprs(.))) %>%
  ungroup() %>%
  arrange(ensembl_gene_id, ensembl_transcript_id, as.numeric(int_start), as.numeric(int_stop)) %>%
  filter(overlapping_iprs != "") %>%
  arrange(ipr_accession)
```

Manually inspect the `overlapping_iprs_table` table and write down entries that are duplicates and need to be filtered out:

a) If entries represent the same type of domain, keep the longer match, because it may be an array of domains or a pair of domains at a long distance, while the short match could be just one unit of the array:

```{r, include=T}
# Keep the longest match out of overlapping ipr1 and ipr2
keep.longer.match.df = data.frame(ipr1 = c("IPR000008", "IPR000014", "IPR000014", "IPR000306", "IPR000313", "IPR000433", 
                                           "IPR000536", "IPR000953", "IPR000953", "IPR000980", "IPR001356", "IPR001356",
                                           "IPR001487", "IPR001660", "IPR001660", "IPR001781", "IPR001841", "IPR001841",
                                           "IPR001841", "IPR001841", "IPR001909", "IPR001965", "IPR001965", "IPR001965",
                                           "IPR001965", "IPR001965", "IPR001965", "IPR001965", "IPR002713", "IPR002999",
                                           "IPR002999", "IPR002999", "IPR002999", "IPR003118", "IPR003118", "IPR003118",
                                           "IPR003604", "IPR011989", "IPR012921", "IPR013087", "IPR013087", "IPR013087",
                                           "IPR013087", "IPR013763", "IPR013783", "IPR013520", "IPR014001", "IPR015943",
                                           "IPR017877", "IPR019787", "IPR019787", "IPR019787", "IPR021935", "IPR023780",
                                           "IPR026932", "IPR032397", "IPR037518", "IPR037598", "IPR037599", "IPR037600"),
                                  ipr2 = c("IPR037772", "IPR013767", "IPR013655", "IPR017455", "IPR035501", "IPR041981",
                                           "IPR021064", "IPR023780", "IPR017984", "IPR035859", "IPR024578", "IPR008422",
                                           "IPR037374", "IPR037611", "IPR037612", "IPR042688", "IPR027370", "IPR018957", 
                                           "IPR041888", "IPR040380", "IPR003655", "IPR019787", "IPR042023", "IPR042025",
                                           "IPR042017", "IPR042014", "IPR042015", "IPR042580", "IPR032835", "IPR041297",
                                           "IPR040477", "IPR041291", "IPR041292", "IPR041886", "IPR035573", "IPR042693",
                                           "IPR022755", "IPR016024", "IPR010912", "IPR041697", "IPR003604", "IPR036236",
                                           "IPR022755", "IPR013150", "IPR014756", "IPR037431", "IPR000330", "IPR036322",
                                           "IPR017930", "IPR042023", "IPR042025", "IPR042014", "IPR037745", "IPR017984",
                                           "IPR030392", "IPR033926", "IPR000555", "IPR041418", "IPR041418", "IPR041418"))
```

Print overlap examples:

- Specific overlaps general:

```{r, include=T}
same.domain.gen.spec = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(ensembl_transcript_id == "ENST00000318003") %>%
  filter(ipr_accession %in% c("IPR000008", "IPR037772")) %>%
  dplyr::select(ipr_accession, ipr_description, int_start, int_stop, ipr_status)

same.domain.gen.spec
```

- Specific overlaps specific:

```{r, include=T}
same.domain.spec.spec = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(ensembl_transcript_id == "ENST00000322349") %>%
  filter(ipr_accession %in% c("IPR000306", "IPR017455")) %>%
  dplyr::select(ipr_accession, ipr_description, int_start, int_stop, ipr_status)

same.domain.spec.spec
```

Plot overlap examples:

- Specific overlaps general:

```{r, include=T}
p = ggplot(same.domain.gen.spec) +
  geom_rect(mapping = aes(xmin = int_start,
                          xmax = int_stop,
                          ymin = c(10, 30),
                          ymax = c(20, 40),
                          fill = ipr_accession)) +
  geom_text(aes(x = min(int_start) - 30,
                y = c(15, 35),
                label = paste(ipr_accession, ipr_description)),
            size = 3) +
  # Ruler
  scale_x_continuous(n.breaks = 7, 
                     limits = c(600, max(same.domain.gen.spec$int_stop))) +
  # Y axis scale
  scale_y_continuous(limits = c(0, 100)) +
  # Formatting
  theme_classic() + 
  theme(axis.line.y = element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "none")

ggsave(filename = "data/results/same_domain_general_specific_overlap.pdf",
       plot = p,
       width = 4,
       height = 3)

p
```

- Specific overlaps specific:

```{r, include=T}
p = ggplot(same.domain.spec.spec) +
  geom_rect(mapping = aes(xmin = int_start,
                          xmax = int_stop,
                          ymin = c(10, 30),
                          ymax = c(20, 40),
                          fill = ipr_accession)) +
  geom_text(aes(x = min(int_start) - 20,
                y = c(15, 35),
                label = paste(ipr_accession, ipr_description)),
            size = 3) +
  # Ruler
  scale_x_continuous(n.breaks = 7, 
                     limits = c(1300, max(same.domain.spec.spec$int_stop))) +
  # Y axis scale
  scale_y_continuous(limits = c(0, 100)) +
  # Formatting
  theme_classic() + 
  theme(axis.line.y = element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "none")

ggsave(filename = "data/results/same_domain_specific_specific_overlap.pdf",
       plot = p,
       width = 4,
       height = 3)

p
```

b) If entries represent different types of domains (or at least it's unclear if they represent the same one), then: 

- if they overlap reciprocally by the set cutoff (and so represent the same type of domain under different names), then retain the more biologically relevant or better described entry in each TF:

```{r, include=T}
# Always retain ipr1
keep.more.relevant.df = data.frame(ipr1 = c("IPR001841", "IPR003347", "IPR013087", "IPR013087", "IPR013087", "IPR013087", 
                                            "IPR013087"),
                                   ipr2 = c("IPR019787", "IPR041667", "IPR008598", "IPR034731", "IPR041643", "IPR034729", 
                                            "IPR041011"))
```

Print an overlap example:

```{r, include=T}
diff.domains.recip = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(ensembl_transcript_id == "ENST00000318524") %>%
  filter(ipr_accession %in% c("IPR001841", "IPR019787")) %>%
  dplyr::select(ipr_accession, ipr_description, int_start, int_stop, ipr_status)

diff.domains.recip
```

Plot the overlap example:

```{r, include=T}
p = ggplot(diff.domains.recip) +
  geom_rect(mapping = aes(xmin = int_start,
                          xmax = int_stop,
                          ymin = c(10, 30),
                          ymax = c(20, 40),
                          fill = ipr_accession)) +
  geom_text(aes(x = min(int_start) - 20,
                y = c(15, 35),
                label = paste(ipr_accession, ipr_description)),
            size = 3) +
  # Ruler
  scale_x_continuous(n.breaks = 7, 
                     limits = c(300, max(diff.domains.recip$int_stop))) +
  # Y axis scale
  scale_y_continuous(limits = c(0, 100)) +
  # Formatting
  theme_classic() + 
  theme(axis.line.y = element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "none")

ggsave(filename = "data/results/different_domains_reciprocal_overlap.pdf",
       plot = p,
       width = 4,
       height = 3)

p
```

- if they do not overlap reciprocally by the set cutoff (and so the match of one of them is considerably longer than the other), then they represent two different domains, one being partially or fully a part of the other, and hence we need to keep both of them (alternative scenarios are that a smaller domain resides within an array represented as a bigger domain, or domains are shifted relative to each other).

Print an overlap example:

```{r, include=T}
diff.domains.nonrecip = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(ensembl_transcript_id == "ENST00000235790") %>%
  filter(ipr_accession %in% c("IPR003347", "IPR019787")) %>%
  dplyr::select(ipr_accession, ipr_description, int_start, int_stop, start, stop, ipr_status)

diff.domains.nonrecip
```

Plot an overlap example:

```{r, include=T}
p = ggplot(diff.domains.nonrecip) +
  # InterPro entries
  geom_rect(diff.domains.nonrecip %>% 
              dplyr::select(-start,
                            -stop) %>%
              distinct(),
            mapping = aes(xmin = int_start,
                          xmax = int_stop,
                          ymin = c(30, 70),
                          ymax = c(40, 80),
                          fill = ipr_accession)) +
  geom_text(diff.domains.nonrecip %>% 
              dplyr::select(-start,
                            -stop) %>%
              distinct(), 
            mapping = aes(x = min(int_start) - 20,
                          y = c(35, 75),
                          label = paste(ipr_accession, ipr_description)),
            size = 3) +
  geom_rect(diff.domains.nonrecip %>% 
              dplyr::select(-int_start,
                            -int_stop),
            mapping = aes(xmin = start,
                          xmax = stop,
                          ymin = c(10, 50, 50),
                          ymax = c(20, 60, 60),
                          fill = ipr_accession)) +
  # Ruler
  scale_x_continuous(n.breaks = 7, 
                     limits = c(0, max(diff.domains.nonrecip$int_stop))) +
  # Y axis scale
  scale_y_continuous(limits = c(0, 100)) +
  # Formatting
  theme_classic() + 
  theme(axis.line.y = element_blank(),
        axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.x = element_blank(),
        legend.position = "none")

ggsave(filename = "data/results/different_domains_nonreciprocal_overlap.pdf",
       plot = p,
       width = 4,
       height = 3)

p
```

Remove the duplicates:

```{r, include=T}
overlap.cutoff = 0.7

remove_ipr = function(gene.id, transcript.id, ipr, ipr.start, ipr.stop) {

  tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl.deduped <<- subset(tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl.deduped, 
       !((ensembl_gene_id == gene.id) & 
         (ensembl_transcript_id == transcript.id) & 
         (ipr_accession == ipr) & 
         (int_start == ipr.start) &
         (int_stop == ipr.stop)))
}

dedup_overlapping_domains = function(tr.df) {
  gene.id = unique(tr.df$ensembl_gene_id)
  
  transcript.id = unique(tr.df$ensembl_transcript_id)
  
  tr.df.matches = tr.df %>%
    dplyr::select(ipr_accession, ipr_description, int_start, int_stop, domain_type)

  if (is.na(tr.df.matches[1, "ipr_accession"]) | (nrow(tr.df.matches) == 1)) {
    return(tr.df)
  }
  
  i = 1
  i.max = nrow(tr.df.matches) - 1
  j.max = i.max + 1
  while (i <= i.max) {
    j = i + 1
    while (j <= j.max) {
      i.start = unlist(tr.df.matches[i, "int_start"])
      i.stop = unlist(tr.df.matches[i, "int_stop"])
      j.start = unlist(tr.df.matches[j, "int_start"])
      j.stop = unlist(tr.df.matches[j, "int_stop"])
      i.length = i.stop - i.start + 1
      j.length = j.stop - j.start + 1
      ij.overlap.length = min(i.stop, j.stop) - max(i.start, j.start) + 1
      if ((ij.overlap.length / i.length >= overlap.cutoff) | 
          (ij.overlap.length / j.length >= overlap.cutoff)) {
        i.ipr = unlist(tr.df.matches[i, "ipr_accession"])
        j.ipr = unlist(tr.df.matches[j, "ipr_accession"])
        if (nrow(keep.longer.match.df %>% filter((ipr1 == i.ipr) & (ipr2 == j.ipr))) > 0 |
            nrow(keep.longer.match.df %>% filter((ipr1 == j.ipr) & (ipr2 == i.ipr))) > 0) {
          if (i.length > j.length) {
            remove_ipr(gene.id, transcript.id, j.ipr, j.start, j.stop)
          } else {
            remove_ipr(gene.id, transcript.id, i.ipr, i.start, i.stop)
          }
        } else if (nrow(keep.more.relevant.df %>% filter((ipr1 == i.ipr) & (ipr2 == j.ipr))) > 0) {
          remove_ipr(gene.id, transcript.id, j.ipr, j.start, j.stop)
        } else if (nrow(keep.more.relevant.df %>% filter((ipr1 == j.ipr) & (ipr2 == i.ipr))) > 0) {
          remove_ipr(gene.id, transcript.id, i.ipr, i.start, i.stop)
        }
      }
      j = j + 1
    }
    i = i + 1
  }
  
  return(tr.df)
}

tic("Dedup overlapping domains")
tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl.deduped = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl

tmp.table = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  dplyr::select(ensembl_gene_id, ensembl_transcript_id, ipr_accession, ipr_description, int_start, int_stop, domain_type) %>%
  distinct() %>%
  group_by(ensembl_gene_id, ensembl_transcript_id) %>%
  do(dedup_overlapping_domains(.))

rm(tmp.table)

tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl.deduped
toc()
# ~20 sec
```

Check filtering results for selected examples:

```{r, include=T}
tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(ensembl_transcript_id %in% c("ENST00000318003", "ENST00000322349", "ENST00000318524", "ENST00000235790")) %>%
  dplyr::select(humantfs_gene_name, ensembl_transcript_id, ipr_accession, ipr_description, int_start, int_stop) %>%
  distinct()
  # "ENST00000318003": "IPR037772" but not "IPR000008"
  # "ENST00000322349": "IPR000306" but not "IPR017455"
  # "ENST00000318524": "IPR001841" but not "IPR019787"
  # "ENST00000235790": both: "IPR003347", "IPR019787"
```

The filtering worked as expected.

Obtain amino acid sequences of all DBDs and non-DBDs (file "data/ensembl99/Homo_sapiens.GRCh38.pep.all.99.tsv" was obtained by running the command from "data/ensembl99/fa2tsv.txt" in "data/ensembl99/"):

```{r, include=T}
ens99_pep = read.delim("data/ensembl99/Homo_sapiens.GRCh38.pep.all.99.tsv", header = F) %>%
              dplyr::rename("protein_accession" = "V1",
                            "protein_sequence" = "V2") %>%
  mutate(protein_sequence = as.character(protein_sequence))

canonical_dbd_iprs = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(!is.na(represents_family)) %>%
  dplyr::select(ipr_accession, ipr_description, represents_family) %>%
  arrange(represents_family, ipr_accession) %>%
  distinct() %>%
  pull(ipr_accession)

dbd_seq_table = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(!DBD %in% c("Unknown", "Excluded")) %>%
  filter(ipr_accession %in% canonical_dbd_iprs) %>%
  dplyr::select(ensembl_gene_id, 
                ensembl_transcript_id, 
                protein_accession, 
                ipr_accession, 
                ipr_description, 
                db, 
                domain_accession, 
                start, 
                stop) %>%
  left_join(ens99_pep, by = c("protein_accession" = "protein_accession")) %>%
  mutate(dbd_sequence = stringr::str_sub(protein_sequence, start = start, end = stop)) %>%
  arrange(ensembl_gene_id, ensembl_transcript_id, as.numeric(start), as.numeric(stop))

write.table(dbd_seq_table,
            file = "data/results/dbd_seq_table.tsv",
            sep = "\t",
            quote = F,
            row.names = F)

nondbd_seq_table = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(!DBD %in% c("Unknown", "Excluded")) %>%
  filter(!ipr_accession %in% canonical_dbd_iprs) %>%
  dplyr::select(ensembl_gene_id, 
                ensembl_transcript_id, 
                protein_accession, 
                ipr_accession, 
                ipr_description, 
                db, 
                domain_accession, 
                start, 
                stop) %>%
  left_join(ens99_pep, by = c("protein_accession" = "protein_accession")) %>%
  mutate(nondbd_sequence = stringr::str_sub(protein_sequence, start = start, end = stop)) %>%
  arrange(ensembl_gene_id, ensembl_transcript_id, as.numeric(start), as.numeric(stop))

write.table(nondbd_seq_table,
            file = "data/results/nondbd_seq_table.tsv",
            sep = "\t",
            quote = F,
            row.names = F)
```

## Describe changes in major TF isoforms

First of all, set the expression values < 1 TPM to 0 (so all non-expressed isoforms have 0 TPM):

```{r, include=T}
# Clean up table names!

isoform.expression.tf.annot = read.delim("data/results/isoform_expression_tf_annot_1.tsv",
                                         header = T,
                                         sep = "\t",
                                         stringsAsFactors = F)

isoform.expression.tf.annot.bin = cbind(isoform.expression.tf.annot[, 1:5],
                                        apply(isoform.expression.tf.annot[, -c(1:5)],
                                              c(1, 2),
                                              function(x) {
                                                ifelse(x < 1, 0, x)
                                              }))
```

Next, calculate isoform ranks for each TF in each tissue. If an isoform is not expressed in a tissue, it will have a rank 10000 in that tissue:

```{r, include=T}
tissue.names = names(isoform.expression.tf.annot.bin)[6:(ncol(isoform.expression.tf.annot.bin))]

tic("Calculate ranks of isoforms for all TFs in all tissues")
isoform.expression.tf.annot.ranks = bind_rows(lapply(isoform.expression.tf.annot.bin %>%
                                                       pull(ensembl_gene_id) %>%
                                                       unique(),
                                                     function(gene_id) {
                                                       tf.df = isoform.expression.tf.annot.bin %>%
                                                                 filter(ensembl_gene_id == gene_id)
                                                       tf.rank.df = isoform.expression.tf.annot.bin %>%
                                                                      filter(ensembl_gene_id == gene_id) %>%
                                                                      dplyr::select(ensembl_gene_id,
                                                                                    humantfs_gene_name,
                                                                                    ensembl_transcript_id,
                                                                                    tf_family,
                                                                                    n_dbd) %>%
                                                                    bind_cols(lapply(tissue.names,
                                                                                     function(tissue.name) {
                                                                                       rank.df.temp = data.frame(a = rank(-tf.df[[tissue.name]],
                                                                                                                          ties.method = "min"))
                                                                                       return(data.frame(a = unlist(lapply(1:nrow(rank.df.temp),
                                                                                                                           function(i) {
                                                                                                                             ifelse(tf.df[i, tissue.name] == 0,
                                                                                                                                    10000,
                                                                                                                                    rank.df.temp[i, 1])
                                                                                                                           }))))
                                                                                     }))
                                                       names(tf.rank.df) = c(names(tf.rank.df)[1:5], tissue.names)
                                                       return(tf.rank.df)
                                                     }))
toc()
#~31 sec

write.table(isoform.expression.tf.annot.ranks,
            file = "data/results/isoform_expression_ranks_ties_min_bin.tsv",
            quote = F,
            row.names = F,
            sep = "\t")
```

Load the table of isoform ranks:

```{r, include=T}
isoform.expression.tf.annot.ranks = read.delim("data/results/isoform_expression_ranks_ties_min_bin.tsv",
                                               header = T,
                                               sep = "\t",
                                               stringsAsFactors = F)
```

Calculate the fractions of TFs with 0, 1, 2, 3, >=4 expressed isoforms:

```{r, include=T}
isoform.expression.tf.annot.ranks.expressed = isoform.expression.tf.annot.ranks %>%
  rowwise() %>%
  mutate(isoform.expressed = any(c_across(all_of(tissue.names)) != 10000)) %>%
  group_by(ensembl_gene_id) %>%
  mutate(expressed.isoform.n = length(isoform.expressed[isoform.expressed])) %>%
  ungroup()

# Overall number of TFs in biological families
tf.n = isoform.expression.tf.annot.ranks.expressed %>%
  pull(ensembl_gene_id) %>% 
  unique() %>%
  length()

expressed.isoforms.n.table = isoform.expression.tf.annot.ranks.expressed %>%
  dplyr::select(ensembl_gene_id, expressed.isoform.n) %>%
  distinct() %>%
  mutate(isoform_n_summary = ifelse(expressed.isoform.n >= 4, ">=4", expressed.isoform.n)) %>%
  dplyr::select(-expressed.isoform.n) %>%
  mutate(feature = "Expr. isof.") %>%
  group_by(isoform_n_summary) %>%
  do(mutate(., summary_fraction = nrow(.) / tf.n)) %>%
  ungroup() %>%
  dplyr::select(-ensembl_gene_id) %>%
  distinct()
```

Next, find major isoforms in each TF:

```{r, include=T}
calc_if_dominant = function(isoform.df) {
  isoform.df.tpms = isoform.df %>%
    dplyr::select(-ensembl_gene_id,
                  -humantfs_gene_name,
                  -ensembl_transcript_id,
                  -tf_family,
                  -n_dbd,
                  -isoform.expressed,
                  -expressed.isoform.n)
  return(any(isoform.df.tpms[1, ] == 1))
}

tic("Find and select major isoforms")
isoform.expression.tf.annot.dom.isof = isoform.expression.tf.annot.ranks.expressed %>%
  group_by(ensembl_transcript_id) %>%
  do(mutate(., is_dominant = calc_if_dominant(.))) %>%
  ungroup() %>%
  filter(is_dominant)
toc()
# ~19 sec
```

Now it is convenient to calculate the proportions of DBD+ and DBD- isoforms that are expressed in at least one tissue and get major in at least one tissue:

```{r, include=T}
isoform.expression.tf.annot.ranks.expressed.stats = isoform.expression.tf.annot.ranks.expressed %>%
  group_by(ensembl_transcript_id) %>%
  do(mutate(., is_dominant = calc_if_dominant(.))) %>%
  ungroup() %>%
  dplyr::select(ensembl_transcript_id,
                n_dbd,
                isoform.expressed,
                is_dominant) %>%
  mutate(dbd_status = ifelse(n_dbd == 0, "-", "+")) %>%
  add_count(dbd_status) %>%
  dplyr::rename("count_by_dbd" = "n") %>%
  add_count(dbd_status, isoform.expressed) %>%
  dplyr::rename("count_by_dbd_expr" = "n") %>%
  add_count(dbd_status, isoform.expressed, is_dominant) %>%
  dplyr::rename("count_by_dbd_expr_dominant" = "n") %>%
  mutate(expression_ratio = count_by_dbd_expr / count_by_dbd) %>%
  mutate(dominance_ratio = count_by_dbd_expr_dominant / count_by_dbd_expr) %>%
  dplyr::select(dbd_status, 
                isoform.expressed, 
                is_dominant, 
                count_by_dbd, 
                count_by_dbd_expr, 
                count_by_dbd_expr_dominant, 
                expression_ratio,
                dominance_ratio) %>%
  distinct()

isoform.expression.tf.annot.ranks.expressed.stats

dbdpos_expressed_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "+") & (isoform.expressed)) %>%
  pull(expression_ratio) %>%
  unique()

dbdpos_notexpressed_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "+") & (!isoform.expressed)) %>%
  pull(expression_ratio)

dbdneg_expressed_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "-") & (isoform.expressed)) %>%
  pull(expression_ratio) %>%
  unique()

dbdneg_notexpressed_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "-") & (!isoform.expressed)) %>%
  pull(expression_ratio)

dbdpos_dominant_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "+") & (is_dominant)) %>%
  pull(dominance_ratio)

dbdpos_notdominant_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "+") & (isoform.expressed) & (!is_dominant)) %>%
  pull(dominance_ratio)

dbdneg_dominant_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "-") & (is_dominant)) %>%
  pull(dominance_ratio)

dbdneg_notdominant_prop = isoform.expression.tf.annot.ranks.expressed.stats %>%
  filter((dbd_status == "-") & (isoform.expressed) & (!is_dominant)) %>%
  pull(dominance_ratio)

p = data.frame(bar_id = c(rep("e+", 2),
                          rep("m+", 2),
                          rep("e-", 2),
                          rep("m-", 2)),
               isoform_category = rep(c("Yes", "No"), 4),
               isoform_proportion = c(dbdpos_expressed_prop,
                                      dbdpos_notexpressed_prop,
                                      dbdpos_dominant_prop,
                                      dbdpos_notdominant_prop,
                                      dbdneg_expressed_prop,
                                      dbdneg_notexpressed_prop,
                                      dbdneg_dominant_prop,
                                      dbdneg_notdominant_prop)) %>%
  mutate(bar_id = factor(bar_id, levels = unique(bar_id))) %>%
  mutate(isoform_category = factor(isoform_category, levels = c("Yes", "No"))) %>%
  ggplot() +
    geom_col(aes(x = bar_id,
                 y = isoform_proportion,
                 fill = isoform_category)) +
    theme_classic()

ggsave(filename = "data/results/dbd_pos_neg_expression_dominance_barplot.pdf",
       plot = p,
       width = 4,
       height = 3)
```

Summarise functional types of domains in DBD- isoforms (separately in C2H2 ZF and all other TFs):

```{r, include=T}
domain_functional_classification = read.delim("analysis/classify_domains/output/classify_domains/Other_domains_manually_curated_filled_generalized.csv")

dbdminus.domain.table = isoform.expression.tf.annot.ranks.expressed %>%
  dplyr::select(ensembl_gene_id, 
                humantfs_gene_name, 
                tf_family, 
                ensembl_transcript_id, 
                n_dbd) %>%
  filter(n_dbd == 0) %>%
  left_join(tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
              dplyr::select(ensembl_gene_id, 
                            ensembl_transcript_id, 
                            ipr_accession, 
                            ipr_description,
                            represents_family) %>%
              distinct() %>%
              filter(is.na(represents_family)),
            by = c("ensembl_gene_id" = "ensembl_gene_id",
                   "ensembl_transcript_id" = "ensembl_transcript_id")) %>%
  left_join(domain_functional_classification,
            by = c("ipr_accession" = "ipr_accession",
                   "ipr_description" = "ipr_description")) %>%
  mutate(Molecular_mechanism = ifelse(is.na(Molecular_mechanism), "No domains", Molecular_mechanism),
         Molecular_function = ifelse(is.na(Molecular_function), "No domains", Molecular_function),
         Molecular_function_general = ifelse(is.na(Molecular_function_general), "No domains", Molecular_function_general))

dbdminus.domain.table.per_function = dplyr::bind_rows(lapply(1:(nrow(dbdminus.domain.table)),
                                                               function(i) {
                                                                 df.i = dbdminus.domain.table[i, ]
                                                                 return(data.frame(ensembl_gene_id = df.i$ensembl_gene_id,
                                                                                   humantfs_gene_name = df.i$humantfs_gene_name,
                                                                                   tf_family = df.i$tf_family,
                                                                                   ensembl_transcript_id = df.i$ensembl_transcript_id,
                                                                                   n_dbd = df.i$n_dbd,
                                                                                   ipr_accession = df.i$ipr_accession,
                                                                                   ipr_description = df.i$ipr_description,
                                                                                   Molecular_function_general = unlist(stringr::str_split(df.i$Molecular_function_general, 
                                                                                                                                  fixed("; ")))))
                                                               })) %>%
  group_by(ensembl_transcript_id) %>%
  mutate(general_function_list = paste0(unique(sort(Molecular_function_general)), collapse = "; ")) %>%
  ungroup()
```

Plot the summaries (Fig. S5A-B):

```{r, include=T}
# C2H2 ZFs
# Fig. S5A
p = dbdminus.domain.table.per_function %>%
  filter(tf_family == "C2H2 ZF") %>%
  dplyr::select(ensembl_gene_id, ensembl_transcript_id, Molecular_function_general) %>%
  distinct() %>%
  mutate(total_n = length(unique(.$ensembl_transcript_id))) %>%
  add_count(Molecular_function_general) %>%
  mutate(frequency = n / total_n) %>%
  dplyr::select(Molecular_function_general, frequency) %>%
  distinct() %>%
  arrange(desc(frequency)) %>%
  mutate(Molecular_function_general = factor(Molecular_function_general, levels = Molecular_function_general)) %>%
  ggplot(aes(x = Molecular_function_general, y = frequency)) +
    geom_col(fill = "orange") +
    scale_y_continuous(limits = c(0, 0.6), breaks = scales::pretty_breaks(7)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# Clean up plot names!

ggsave(filename = "data/results/dbdminus_per_nondbd_functions_c2h2zf_fractions_barplot_3.pdf",
       plot = p,
       width = 6,
       height = 6)

# Other TFs
# Fig. S5B
p = dbdminus.domain.table.per_function %>%
  filter(tf_family != "C2H2 ZF") %>%
  dplyr::select(ensembl_gene_id, ensembl_transcript_id, Molecular_function_general) %>%
  distinct() %>%
  mutate(total_n = length(unique(.$ensembl_transcript_id))) %>%
  add_count(Molecular_function_general) %>%
  mutate(frequency = n / total_n) %>%
  dplyr::select(Molecular_function_general, frequency) %>%
  distinct() %>%
  arrange(desc(frequency)) %>%
  mutate(Molecular_function_general = factor(Molecular_function_general, levels = Molecular_function_general)) %>%
  ggplot(aes(x = Molecular_function_general, y = frequency)) +
    geom_col(fill = "blue") +
    scale_y_continuous(limits = c(0, 0.6), breaks = scales::pretty_breaks(7)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# Clean up plot names!

ggsave(filename = "data/results/dbdminus_per_nondbd_functions_non-c2h2zf_fractions_barplot_3.pdf",
       plot = p,
       width = 6,
       height = 6)
```

Transcription repression domains in C2H2 ZFs are as follows:

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family == "C2H2 ZF") %>%
  filter(general_function_list == "Transcription repression") %>%
  add_count(ipr_description) %>%
  dplyr::select(ipr_accession, ipr_description, n) %>%
  distinct() %>%
  arrange(desc(n))
```

Oligomerization domains in C2H2 ZFs are as follows:

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family == "C2H2 ZF") %>%
  filter(general_function_list == "Oligomerization") %>%
  add_count(ipr_description) %>%
  dplyr::select(ipr_accession, ipr_description, n) %>%
  distinct() %>%
  arrange(desc(n))
```

Dimerization domains in C2H2 ZFs are as follows:

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family == "C2H2 ZF") %>%
  filter(general_function_list == "Dimerization") %>%
  add_count(ipr_description) %>%
  dplyr::select(ipr_accession, ipr_description, n) %>%
  distinct() %>%
  arrange(desc(n))
```

Dimerization domains in other TFs are as follows:

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family != "C2H2 ZF") %>%
  filter(general_function_list == "Dimerization") %>%
  add_count(ipr_description) %>%
  dplyr::select(ipr_accession, ipr_description, n) %>%
  distinct() %>%
  arrange(desc(n))
```

Ligand-binding domains in other TFs are as follows:

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family != "C2H2 ZF") %>%
  filter(general_function_list == "Ligand binding") %>%
  add_count(ipr_description) %>%
  dplyr::select(ipr_accession, ipr_description, n) %>%
  distinct() %>%
  arrange(desc(n))
```

Transcription activation domains in other TFs are as follows:

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family != "C2H2 ZF") %>%
  filter(general_function_list == "Transcription activation") %>%
  add_count(ipr_description) %>%
  dplyr::select(ipr_accession, ipr_description, n) %>%
  distinct() %>%
  arrange(desc(n))
```

The number of DBD- isoforms produced by C2H2 ZFs is

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family == "C2H2 ZF") %>%
  pull(ensembl_transcript_id) %>%
  unique() %>%
  length()
```

and the number of DBD- isoforms produced by other TFs is

```{r, include=T}
dbdminus.domain.table.per_function %>%
  filter(tf_family != "C2H2 ZF") %>%
  pull(ensembl_transcript_id) %>%
  unique() %>%
  length()
```

There is only

```{r, include=T}
dbdminus.domain.table.per_function %>% 
  filter((Molecular_function_general == "DBD") | (Molecular_function_general == "?DBD")) %>% 
  pull(ensembl_transcript_id) %>% 
  length()
```

DBD- isoform that contains domains qualified as "DBD" or "?DBD" (these domains are "non-canonical" DBDs). This domain is:

```{r, include=T}
dbdminus.domain.table.per_function %>% 
  filter((Molecular_function_general == "DBD") | (Molecular_function_general == "?DBD")) %>% 
  dplyr::select(ipr_accession, ipr_description)
```

which can contact linker DNA between two nucleosomes as a part of a larger protein complex (see IPR028941 in InterPro). Let us exclude the "DBD-" isoform 

```{r, include=T}
dbdminus.domain.table.per_function %>% 
  filter((Molecular_function_general == "DBD") | (Molecular_function_general == "?DBD")) %>% 
  pull(ensembl_transcript_id)
```

that contains this domain from further analysis of DBD- isoforms:

```{r, include=T}
dbdminus.domain.table.per_function = dbdminus.domain.table.per_function %>%
  filter(ensembl_transcript_id != "ENST00000549787")
```

Now, using the functional annotation of non-DBDs, we can make general predictions about functions of DBD- isoforms.

First, write down the general lists of DBD- functions:

```{r, include=T}
write.table(dbdminus.domain.table.per_function %>%
              dplyr::select(general_function_list) %>%
              distinct(),
            file = "data/results/dbdminus_domain_table_per_function_raw.tsv",
            quote = F,
            row.names = F,
            sep = "\t")
```

Next, we annotated the list of non-DBD combinations manually based on known examples (the result is in the `data/results/dbdminus_domain_table_per_function_filled.tsv` table).

Finally, plot the proportions of DBD- isoforms with different predicted functions (Fig. 5A-B):

```{r, include=T}
dbdminus.domain.table.per_function.filled = read.delim("data/results/dbdminus_domain_table_per_function_filled.tsv",
                                                       header = T,
                                                       sep = "\t")

dbdminus.domain.table.per_function.annot = dbdminus.domain.table.per_function %>%
  left_join(dbdminus.domain.table.per_function.filled,
            by = c("general_function_list" = "general_function_list"))

# C2H2 ZFs
# Fig. 5A
p = dbdminus.domain.table.per_function.annot %>%
  filter(tf_family == "C2H2 ZF") %>%
  dplyr::select(ensembl_transcript_id, isoform_function) %>%
  distinct() %>%
  add_count(isoform_function) %>%
  dplyr::select(-ensembl_transcript_id) %>%
  distinct() %>%
  arrange(desc(n)) %>%
  mutate(total_n = sum(n)) %>%
  mutate(function_freq = n / total_n) %>%
  mutate(isoform_function = factor(isoform_function, levels = isoform_function)) %>%
  ggplot(aes(x = isoform_function, y = function_freq)) +
    geom_col(fill = "orange") +
    scale_y_continuous(limits = c(0, 0.55), breaks = scales::pretty_breaks(6)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# Clean up plot names!

ggsave(filename = "data/results/nondbd_isoform_functions_c2h2zf_fractions_barplot_3.pdf",
       plot = p,
       width = 6,
       height = 6)

# Other TFs
# Fig. 5B
p = dbdminus.domain.table.per_function.annot %>%
  filter(tf_family != "C2H2 ZF") %>%
  dplyr::select(ensembl_transcript_id, isoform_function) %>%
  distinct() %>%
  add_count(isoform_function) %>%
  dplyr::select(-ensembl_transcript_id) %>%
  distinct() %>%
  arrange(desc(n)) %>%
  mutate(total_n = sum(n)) %>%
  mutate(function_freq = n / total_n) %>%
  mutate(isoform_function = factor(isoform_function, levels = isoform_function)) %>%
  ggplot(aes(x = isoform_function, y = function_freq)) +
    geom_col(fill = "blue") +
    scale_y_continuous(limits = c(0, 0.7), breaks = scales::pretty_breaks(7)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# Clean up plot names!
 
ggsave(filename = "data/results/nondbd_isoform_functions_non-c2h2zf_fractions_barplot_3.pdf",
       plot = p,
       width = 6,
       height = 6)
```

37% of C2H2 ZF and 62% of non-C2H2 ZF DBD- isoforms have an "Unclear" function. Summarise domain functions in these DBD- isoforms to see why:

```{r, include=T}
# C2H2 ZF
# Fig. S3...
p = dbdminus.domain.table.per_function.annot %>%
  filter(tf_family == "C2H2 ZF") %>%
  filter(isoform_function == "Unclear") %>%
  dplyr::select(ensembl_transcript_id, general_function_list) %>%
  distinct() %>%
  add_count(general_function_list) %>%
  dplyr::select(-ensembl_transcript_id) %>%
  distinct() %>%
  arrange(desc(n)) %>%
  mutate(total_n = sum(n)) %>%
  mutate(function_freq = n / total_n) %>%
  mutate(general_function_list = factor(general_function_list, levels = general_function_list)) %>%
  ggplot(aes(x = general_function_list, y = function_freq)) +
    geom_col(fill = "orange") +
    #scale_y_continuous(limits = c(0, 0.55), breaks = scales::pretty_breaks(6)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# Clean up plot names!

ggsave(filename = "data/results/nondbd_unclear_c2h2zf_fractions_barplot_3.pdf",
       plot = p,
       width = 6,
       height = 6)

# Non-C2H2 ZF
# Fig. S3...
p = dbdminus.domain.table.per_function.annot %>%
  filter(tf_family != "C2H2 ZF") %>%
  filter(isoform_function == "Unclear") %>%
  dplyr::select(ensembl_transcript_id, general_function_list) %>%
  distinct() %>%
  add_count(general_function_list) %>%
  dplyr::select(-ensembl_transcript_id) %>%
  distinct() %>%
  arrange(desc(n)) %>%
  mutate(total_n = sum(n)) %>%
  mutate(function_freq = n / total_n) %>%
  mutate(general_function_list = factor(general_function_list, levels = general_function_list)) %>%
  ggplot(aes(x = general_function_list, y = function_freq)) +
    geom_col(fill = "blue") +
    #scale_y_continuous(limits = c(0, 0.55), breaks = scales::pretty_breaks(6)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# Clean up plot names!

ggsave(filename = "data/results/nondbd_unclear_non-c2h2zf_fractions_barplot_3.pdf",
       plot = p,
       width = 6,
       height = 6)
```

Hence, the vast majority of DBD- isoforms with an "Unclear" function just do not have domains. Consequently, we need to check if they have IDRs and so still could be functional.

## Summarise major isoform switches

List all switches:

```{r, include=T}
diff.cutoff = 0.1

find_switch = function(isof.pair) {
  isof.id1 = isof.pair[1]
  isof.id2 = isof.pair[2]
  
  isof.id1.tissues = tissue.names[as.numeric(isoform.expression.tf.annot.dom.isof %>%
                                               filter(ensembl_transcript_id == isof.id1) %>%
                                               dplyr::select(all_of(tissue.names))) == 1]
  
  isof.id2.tissues = tissue.names[as.numeric(isoform.expression.tf.annot.dom.isof %>%
                                               filter(ensembl_transcript_id == isof.id2) %>%
                                               dplyr::select(all_of(tissue.names))) == 1]
  
  isof.tissues = sort(unique(c(isof.id1.tissues, isof.id2.tissues)))
  
  isof.df.1 = isoform.expression.tf.annot.bin %>%
    filter(ensembl_transcript_id == isof.id1) %>%
    dplyr::select(all_of(isof.tissues))
  
  isof.df.2 = isoform.expression.tf.annot.bin %>%
    filter(ensembl_transcript_id == isof.id2) %>%
    dplyr::select(all_of(isof.tissues))
  
  a = as.numeric(isof.df.1[1, ])
  
  b = as.numeric(isof.df.2[1, ])
  
  prop.a.b = a / b
  
  prop.b.a = b / a
  
  prop.a.b.test = (prop.a.b <= 1 - diff.cutoff)
  
  prop.b.a.test = (prop.b.a <= 1 - diff.cutoff)
  
  switch.exists = as.numeric((length(prop.a.b.test[prop.a.b.test]) >= 1) & 
                             (length(prop.b.a.test[prop.b.a.test]) >= 1))
  
  if (switch.exists) {
    isof2.tissues = paste(isof.tissues[prop.a.b.test], collapse = ",")
    isof1.tissues = paste(isof.tissues[prop.b.a.test], collapse = ",")
  } else {
    isof2.tissues = NA
    isof1.tissues = NA
  }
  
  return(paste(c(switch.exists, isof1.tissues, isof2.tissues),
               collapse = "|"))
}

detect_switches = function(gene_id) {
  tf.df = isoform.expression.tf.annot.dom.isof %>%
    filter(ensembl_gene_id == gene_id)
  
  if (nrow(tf.df) == 1) {
    return(data.frame(ensembl_gene_id = gene_id,
                      isof_id1 = (tf.df %>% pull(ensembl_transcript_id)),
                      isof_id2 = (tf.df %>% pull(ensembl_transcript_id)),
                      switch_detected = "0",
                      isof1_tissues = NA,
                      isof2_tissues = NA))
  }
  
  isof.pair.df.tmp = t(combn(tf.df$ensembl_transcript_id, 2))
  
  isof.pair.df = data.frame(isof.pair.df.tmp[, 1],
                            isof.pair.df.tmp[, 2])
  
  names(isof.pair.df) = c("isof_id1", "isof_id2")
  
  isof.pair.df %<>%
    group_by(isof_id1, isof_id2) %>%
    do(mutate(., switch = find_switch(as.character(.)))) %>%
    ungroup() %>%
    tidyr::separate(col = "switch",
                    into = c("switch_detected", "isof1_tissues", "isof2_tissues"),
                    sep = "[|]") %>%
    mutate(ensembl_gene_id = gene_id) %>%
    dplyr::select(ensembl_gene_id, isof_id1, isof_id2, switch_detected, isof1_tissues, isof2_tissues)
    
  return(isof.pair.df)
}

gene_ids = isoform.expression.tf.annot.dom.isof %>%
  filter(expressed.isoform.n > 1) %>%
  pull(ensembl_gene_id) %>%
  unique()

switch.table = bind_rows(lapply(gene_ids,
                                function(gene_id) {
                                  detect_switches(gene_id)
                                })) %>%
  filter(switch_detected == 1)
```

Hence, we have 

```{r, include=T}
switch.table %>%
  nrow()
```

switches in 

```{r, include=T}
switch.table %>%
  pull(ensembl_gene_id) %>%
  unique() %>%
  length()
```

TFs. 

List major isoforms that take part in switches:

```{r, include=T}
switching.isoforms = c(switch.table %>%
                         filter(switch_detected == 1) %>%
                         pull(isof_id1),
                       switch.table %>%
                         filter(switch_detected == 1) %>%
                         pull(isof_id2)) %>%
  unique()
```

Hence, we have

```{r, include=T}
switching.isoforms %>%
  length()
```

major isoforms that take part in at least one switch.

Show how frequently TFs with different numbers of major isoforms switch them:

```{r, include=T}
library(ggsankey)

tf.majors.switches = isoform.expression.tf.annot.ranks.expressed %>%
  dplyr::select(ensembl_gene_id,
                humantfs_gene_name,
                tf_family) %>%
  distinct() %>%
  left_join(isoform.expression.tf.annot.dom.isof %>%
              count(ensembl_gene_id) %>%
              dplyr::rename("major_n" = "n"),
            by = c("ensembl_gene_id" = "ensembl_gene_id")) %>%
  left_join(switch.table %>%
              count(ensembl_gene_id) %>%
              dplyr::rename("switch_n" = "n"),
            by = c("ensembl_gene_id" = "ensembl_gene_id")) %>%
  replace(is.na(.), 0) %>%
  mutate(major_n_summary = ifelse(major_n >= 4, ">=4", major_n)) %>%
  mutate(switch_n_summary = ifelse(switch_n >= 4, ">=4", switch_n)) %>%
  dplyr::rename("major" = "major_n_summary",
                "switch" = "switch_n_summary")

tf.majors.switches.long = tf.majors.switches %>%
  make_long(major, switch)

p = ggplot(tf.majors.switches.long, 
       aes(x = x, 
           next_x = next_x, 
           node = node, 
           next_node = next_node,
           fill = factor(node),
           label = node)) +
  geom_sankey(flow.alpha = 0.5, width = 0.5) + 
  theme_sankey(base_size = 12)

ggsave(filename = "data/results/majors_to_switches_sankey_plot.pdf",
       plot = p,
       width = 4,
       height = 5)
```

Hence, we have 

```{r, include=T}
possibly.switching.tfs.n = tf.majors.switches %>%
  filter((major != "0") & (major != "1")) %>%
  pull(ensembl_gene_id) %>%
  unique() %>%
  length()

possibly.switching.tfs.n
```

TFs that can switch their major isoforms (i. e., express more than one major isoform). They comprise

```{r, include=T}
all.tfs.n = tf.majors.switches %>%
  pull(ensembl_gene_id) %>%
  unique() %>%
  length()

possibly.switching.tfs.n / all.tfs.n * 100
```

percent of all TFs in our analysis.

From these,

```{r, include=T}
switching.tfs.n = tf.majors.switches %>%
  filter(switch != "0") %>%
  pull(ensembl_gene_id) %>%
  unique() %>%
  length()

switching.tfs.n
```

TFs switch their major isoforms across adult human tissues. They comprise

```{r, include=T}
switching.tfs.n / possibly.switching.tfs.n * 100
```

percent of TFs that could switch their major isoforms and

```{r, include=T}
switching.tfs.n / all.tfs.n * 100
```

percent of all TFs.

Here are the numbers of TFs stratified by the number of major isoforms and the number of switches (this table was plotted above as a Sankey plot):

```{r, include=T}
tf.majors.switches %>%
  add_count(major, switch) %>%
  dplyr::select(major, switch, n) %>%
  distinct() %>%
  arrange(major)
```

And here are the numbers and percentages of TFs with different numbers of major isoforms and switches:

```{r, include=T}
# Major isoforms
tf.majors.switches %>%
  count(major) %>%
  mutate(perc = n / sum(n) * 100)
  
# Switches
tf.majors.switches %>%
  count(switch) %>%
  mutate(perc = n / sum(n) * 100)
```

In total, the majority of TFs in our analysis (78%) either express just one major isoform or are not expressed in adult human tissues and consequently do not switch their major isoforms. The majority of the rest express two major isoforms (18%), and most of them (82%) switch the isoforms, while a minor part do not. This is explained by the fact that the difference in expression levels of major isoforms of TFs from this minor part is not substantial enough in any tissue to pass our filtering for switches. A yet smaller proportion of TFs (3%) express three major isoforms, and most of them (63%) have the maximal combinatorially possible number of switches (3). However, a minority of these TFs have one or two switches, because pairs of some of their major isoforms do not have a big enough difference in the transcription level to pass our filtering. Of note, we can formally prove that for any number $n$ of major isoforms such transcription levels are possible that these isoforms produce any number of switches, from 0 to ($n(n-1)/2$).

Interestingly, the proportion of TFs with 2 switches (0.87%) is smaller than the proportion of TFs with 3 switches (1.95%). This could be explained by the fact that to produce 2 switches and to produce 3 switches each TF in our dataset needs 3 major isoforms that comprise 3 pairs. But in a big number of tissues (there are more than 40 tissues in our dataset), for each pair of major isoforms the chance to pass our filtering criterion for a switch in _at least one_ pair of tissues could be higher than the chance not to pass it in _all_ pairs of tissues. Finally, a tiny minority of TFs (0.54%) have 4 or more major isoforms and produce 4 or more switches. None of these TFs produces 3 switches or less, which could be explained in the same way: The bigger the number of major isoforms, the bigger the number of their pairs, and in a big number of tissues the chance for each pair to pass our filtering in at least one pair of tissues could be higher than the chance not to pass it in all pairs of tissues. To formally prove this, we would need to assess the probability of a switch between two major isoforms of a TF that is expressed in a certain number of tissues. However, we are not going into this question at present.

Next, summarise switches via ranking each TF in each tissue according to the major isoform of the TF in the tissue. Namely, if the major isoform of a TF in a tissue is the TF's widest-expressed major isoform and it switches in this tissue, then the TF gets rank 1 or -1, depending on whether the major isoform is DBD+ or DBD-. In the same way, the TF gets rank 2 (-2) or 3 (-3) if its major isoform in the tissue is the TF's second- or third-widest expressed major isoform and it switches in the tissue. If the major isoform of a TF in a tissue is the TF's fourth-, fifth-, etc. widest-expressed major isoform and it switches in this tissue, then the TF gets rank 5, independently of the presence of a DBD in the isoform. If the major isoform of a TF does not switch in a tissue, then in that tissue the TF gets rank 4 (if the two widest-expressed major isoforms are tightly co-expressed in that tissue) or number 5 (otherwise). If a TF is not expressed in a tissue, it gets rank 0 in that tissue. Overall, ranks +/-1, +/-2 and +/-3 designate switches of the major isoforms of a TF, rank 4 designates tight co-expression of the first and second widest-expressed major isoforms of a TF, rank 0 signifies that a TF is not expressed in a tissue, and rank 5 denotes all other possibilities (rare and uninteresting).

```{r, include=T}
generate_tf_cluster_row = function(tf.id) {
  tf.df = isoform.expression.tf.annot.dom.isof.dom_annot %>%
    filter(ensembl_gene_id == tf.id)
  
  switching.tissues = sort(unique(unlist(stringr::str_split(tf.df$merged_tissue_list, fixed(",")))))
  
  tf.cluster.row = tf.df[1, ]
  for (tissue.name in tissue.names) {
    if (!Reduce(function(x, y) x | y, tf.df[, tissue.name] == 1)) { # TF is not expressed in the tissue
      dominant.isoform.rank = 0
    } else if (tissue.name %in% switching.tissues) { # TF is expressed, and the major isoform takes part in switches
      dominant.isoform.rank = unlist(tf.df[tf.df[[tissue.name]] == 1, "isoform_rank"])[1]
      dominant.isoform.rank = ifelse(abs(dominant.isoform.rank) >= 4, 5, dominant.isoform.rank)
    } else { # TF is expressed, and the major isoform does not take part in switches (i. e., is tightly co-expressed with another major isoform)
      dominant.isoform.rank = unlist(tf.df[tf.df[[tissue.name]] == 1, "isoform_rank"])[1]
      ranks.in.tissue = tf.df[[tissue.name]]
      below.top = ranks.in.tissue[ranks.in.tissue != 1]
      if (length(below.top) == 0) { # two major isoforms are expressed at exactly the same level in a tissue
        second.isoform.rank = unlist(tf.df[tf.df[[tissue.name]] == 1, "isoform_rank"])[2]
      } else { # major isoforms are expressed at different levels in a tissue
        min.second = min(ranks.in.tissue[ranks.in.tissue != 1]) 
        second.isoform.rank = unlist(tf.df[tf.df[[tissue.name]] == min.second, "isoform_rank"])[1]
      }
      if ((dominant.isoform.rank %in% c(1, -1, 2, -2)) & (second.isoform.rank %in% c(1, -1, 2, -2))) { # the two top major isoforms are coexpressed
        dominant.isoform.rank = 4
      } else { # other major isoforms are co-expressed
        dominant.isoform.rank = 5
      }
    }
    tf.cluster.row[1, tissue.name] = dominant.isoform.rank
  }
  return(tf.cluster.row)
}

merge_tissue_lists = function(isof.df) {
  if (nrow(isof.df) == 1) {
    return(isof.df$tissue_list)
  }
  
  return(paste(sort(unique(unlist(stringr::str_split(isof.df$tissue_list, fixed(","))))),
               collapse = ","))
}

mi.tissues.table = data.frame(ensembl_transcript_id = c(switch.table$isof_id1,
                                                        switch.table$isof_id2),
                              tissue_list = c(switch.table$isof1_tissues,
                                              switch.table$isof2_tissues)) %>%
  distinct() %>%
  group_by(ensembl_transcript_id) %>%
  do(mutate(., merged_tissue_list = merge_tissue_lists(.))) %>%
  ungroup() %>%
  arrange(ensembl_transcript_id) %>%
  dplyr::select(-tissue_list) %>%
  distinct()

tic("Generate the matrix of numbers of switching major isoforms")
isoform.expression.tf.annot.dom.isof.dom_annot = isoform.expression.tf.annot.dom.isof %>%
  filter(ensembl_transcript_id %in% switching.isoforms) %>%
  left_join(mi.tissues.table,
            by = c("ensembl_transcript_id" = "ensembl_transcript_id")) %>%
  group_by(ensembl_transcript_id) %>%
  mutate(dominance_tissues_n = length(unlist(stringr::str_split(merged_tissue_list, fixed(","))))) %>%
  ungroup() %>%
  group_by(ensembl_gene_id) %>%
  do(mutate(., isoform_rank = rank(-.$dominance_tissues_n,
                                   ties.method = "first"))) %>%
  ungroup() %>%
  group_by(ensembl_transcript_id) %>%
  mutate(isoform_rank = ifelse(n_dbd > 0, 
                               isoform_rank,
                               -isoform_rank)) %>%
  ungroup()

dominant.isoform.number.matrix = bind_rows(lapply(isoform.expression.tf.annot.dom.isof.dom_annot %>%
                                                    pull(ensembl_gene_id) %>%
                                                    unique(),
                                                  function(tf.id) {
                                                    generate_tf_cluster_row(tf.id)
                                                  }))
toc()
# ~23 sec

write.table(dominant.isoform.number.matrix %>%
              dplyr::select(-ensembl_transcript_id,
                            -n_dbd,
                            -isoform.expressed,
                            -expressed.isoform.n,
                            -is_dominant,
                            -dominance_tissues_n,
                            -isoform_rank),
            file = "data/results/dominant_isoform_number_matrix_1.tsv",
            quote = F,
            row.names = F,
            sep = "\t")

write.table(isoform.expression.tf.annot.dom.isof.dom_annot %>%
              dplyr::select(ensembl_gene_id,
                            humantfs_gene_name,
                            ensembl_transcript_id,
                            tf_family,
                            n_dbd,
                            isoform_rank),
            file = "data/results/dominant_isoform_ids_and_numbers_1.tsv",
            quote = F,
            row.names = F,
            sep = "\t")
```

Different isoform ranks:

```{r, include=T}
sort(unique(unlist(dominant.isoform.number.matrix[, c(6:48)])))
```

Cluster the matrix of isoform ranks by the Hamming distance and visualise it:

```{r, include=T}
hamming_dist = function(x, y) {
  match.vector = (x != y)
  return(length(match.vector[match.vector]))
}

heatmap.df = dominant.isoform.number.matrix %>%
  dplyr::select(c(1, c(6:48))) %>%
  tibble::column_to_rownames(var = "ensembl_gene_id")

cluster.colors = structure(c("grey",
                             "dark red", "dark blue",
                             "red", "blue",
                             "#FF9696", "#9696FF", # light red, light blue
                             "dark green", "orange"), 
                             names = c(0, 
                                       1, -1,
                                       2, -2,
                                       3, -3, 
                                       4, 5))

h = Heatmap(heatmap.df, 
            name = "Isoform ranks", 
            clustering_distance_columns = function(x, y) hamming_dist(x, y),
            clustering_distance_rows = function(x, y) hamming_dist(x, y),
            col = cluster.colors,
            row_labels = rep("", nrow(heatmap.df)),
            row_split = 7, # 11
            column_split = 5) # 4

pdf(file = "data/results/clustered_switch_heatmap_2_tissues5.pdf",
    width = 10,
    height = 15)
h
dev.off()
```

Here, we visually split the TF hierarchy into 7 clusters and the tissue hierarchy into 5 clusters.

Tissue clusters:

1) Testis

2) Muscle, Liver and Pancreas (Non-brain 1)

3) The majority of other non-brain tissues (Non-brain 2)

4) Brain except for the cerebellum (Brain)

5) Cerebellum, pituitary and prostate (Other)

The split between cluster Non-brain 2 and everything else happens earlier than the split between brain and non-brain tissues, which suggests a functional significance of the difference between clusters (2) and (3). Below, we show in a PCA plot that the clustering of the non-brain 1 tissues is not due to similar transcription profiles, which would be an artifact in GTEx. Also, interestingly, the cerebellum, pituitary and prostate cluster separately (but merge with the majority of the brain tissues when 4 clusters are chosen). In the examples of TF isoform expression that I saw, I indeed frequently found the cerebellum to be unlike other brain tissues; the pituitary gland is attached to the brain (and so could be involved in some brain-related processes) but also pituitary hormones can control the prostate [Reiter et al., 1999](https://onlinelibrary.wiley.com/doi/epdf/10.1002/%28SICI%291097-0045%2819990201%2938%3A2%3C159%3A%3AAID-PROS10%3E3.0.CO%3B2-5). Hence there could be physiological reasons for these organs to exhibit similar switch patterns. However, below I check on the same PCA plot that the cerebellum, the pituitary gland and the prostate do not have highly overlapping expression profiles, which would be an artifact in GTEx. From the heatmap, switch patterns in the "Other" tissues look more like those in the non-brain tissues 2 than those in the brain or non-brain 1 tissues.

TF clusters:

1) The main isoform is DBD-, and it switches with a DBD+ one (or is coexpressed with it) mostly outside of the brain and the "Other" tissues.

2) The main isoform is also DBD-, and it switches with a DBD+ or another DBD- one in various tissues. However, these isoforms are almost never closely coexpressed.

3) TFs in this cluster are mainly expressed in the non-brain 2 and "Other" tissues, and their main isoforms are DBD+. Switches mostly happen between the main DBD+ isoform and another DBD+ isoform in various tissues. Tight coexpression of switching isoforms is rare.

4) The main isoform is DBD+, and it switches with another DBD+ isoform mostly outside of the brain. Also, switching isoforms are frequently tightly coexpressed.

5) The main isoform is DBD+, and it switches with another DBD+ or a DBD- isoform mostly in the brain. This cluster looks heterogeneous but I would need to increase the number of clusters a lot in order to split it. 

6) TFs in this cluster are extremely tissue-specific. They switch DBD+ and DBD- isoforms across various tissues.

7) The main isoform is DBD+, and it switches with another DBD+ (or is tightly coexpressed with it) mostly outside of the brain. In this way, this cluster is symmetrical to the cluster 1 where the main isoform is DBD-.

Very interestingly, switching TFs are mostly either expressed or not expressed at the same time in the brain and in the non-brain 1 tissues. It would be interesting to check which TFs are mostly not expressed in the brain and the non-brain 1 tissues but are expressed elsewhere (these are all TFs from the cluster 3 and the lower part of the TF cluster 2).

Annotate all switching TFs with their clusters (Fig. 5E and S5E):

```{r, include=T}
h = draw(h)

cluster.list = row_order(h)

cluster.annot.df = data.frame(ensembl_gene_id = character(),
                              cluster_number = numeric())

for (i in 1:(length(cluster.list))) {
  cluster.df = data.frame(ensembl_gene_id = dominant.isoform.number.matrix[cluster.list[[i]], "ensembl_gene_id"],
                          cluster_number = i)
  cluster.annot.df = bind_rows(cluster.annot.df,
                               cluster.df)
}

cluster.annot.df.final = dominant.isoform.number.matrix %>%
  left_join(cluster.annot.df, 
            by = c("ensembl_gene_id" = "ensembl_gene_id")) %>%
  dplyr::select(c(1:5, 55)) %>%
  dplyr::select(-ensembl_transcript_id, -n_dbd)

write.table(cluster.annot.df.final,
            file = "data/results/cluster_annot_df_final.tsv",
            quote = F,
            row.names = F,
            sep = "\t")
```

Print out TFs from each cluster:

```{r, include=T}
cluster.annot.df.final %>%
  filter(cluster_number == 1)
cluster.annot.df.final %>%
  filter(cluster_number == 2)
cluster.annot.df.final %>%
  filter(cluster_number == 3)
cluster.annot.df.final %>%
  filter(cluster_number == 4)
cluster.annot.df.final %>%
  filter(cluster_number == 5)
cluster.annot.df.final %>%
  filter(cluster_number == 6)
cluster.annot.df.final %>%
  filter(cluster_number == 7)
```

```{r, include=T}
write.table(cluster.annot.df.final %>%
              filter(cluster_number == 1),
            file = "data/results/cluster1_tissues5.tsv",
            quote = F,
            row.names = F,
            sep = "\t")

write.table(cluster.annot.df.final %>%
              filter(cluster_number == 2),
            file = "data/results/cluster2_tissues5.tsv",
            quote = F,
            row.names = F,
            sep = "\t")

write.table(cluster.annot.df.final %>%
              filter(cluster_number == 3),
            file = "data/results/cluster3_tissues5.tsv",
            quote = F,
            row.names = F,
            sep = "\t")

write.table(cluster.annot.df.final %>%
              filter(cluster_number == 4),
            file = "data/results/cluster4_tissues5.tsv",
            quote = F,
            row.names = F,
            sep = "\t")

write.table(cluster.annot.df.final %>%
              filter(cluster_number == 5),
            file = "data/results/cluster5_tissues5.tsv",
            quote = F,
            row.names = F,
            sep = "\t")

write.table(cluster.annot.df.final %>%
              filter(cluster_number == 6),
            file = "data/results/cluster6_tissues5.tsv",
            quote = F,
            row.names = F,
            sep = "\t")

write.table(cluster.annot.df.final %>%
              filter(cluster_number == 7),
            file = "data/results/cluster7_tissues5.tsv",
            quote = F,
            row.names = F,
            sep = "\t")
```

## Functionally characterise switches of major TF isoforms

Generate a master table of switches between major TF isoforms:

```{r, include=T}
domain_functional_classification = read.delim("analysis/classify_domains/output/classify_domains/Other_domains_manually_curated_filled_generalized.csv")

dbd_seq_table = read.delim("data/results/dbd_seq_table.tsv",
                           header = T,
                           sep = "\t")

ens99_pep = read.delim("data/ensembl99/Homo_sapiens.GRCh38.pep.all.99.tsv", header = F) %>%
  dplyr::rename("protein_accession" = "V1",
                "protein_sequence" = "V2") %>%
  mutate(protein_sequence = as.character(protein_sequence))

per.isof.domain.table = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
              dplyr::select(ensembl_gene_id, protein_accession, ensembl_transcript_id,  
                            ipr_accession, ipr_description, 
                            start, stop) %>%
              left_join(domain_functional_classification,
                        by = c("ipr_accession" = "ipr_accession",
                               "ipr_description" = "ipr_description")) %>%
              mutate(Molecular_mechanism = ifelse(is.na(Molecular_mechanism) & (ipr_accession %in% canonical_dbd_iprs),
                                                  "Canonical DBD",
                                                  Molecular_mechanism)) %>%
              mutate(Molecular_function = ifelse(is.na(Molecular_function) & (ipr_accession %in% canonical_dbd_iprs),
                                                 "Canonical DBD",
                                                 Molecular_function)) %>%
              mutate(Molecular_function_general = ifelse(is.na(Molecular_function_general) & (ipr_accession %in% canonical_dbd_iprs),
                                                         "Canonical DBD",
                                                         Molecular_function_general)) %>%
              left_join(dbd_seq_table %>%
                          dplyr::select(ensembl_gene_id, ensembl_transcript_id,
                                        ipr_accession, ipr_description,
                                        start, stop,
                                        dbd_sequence),
                        by = c("ensembl_gene_id" = "ensembl_gene_id", 
                               "ensembl_transcript_id" = "ensembl_transcript_id",
                               "ipr_accession" = "ipr_accession", 
                               "ipr_description" = "ipr_description",
                               "start" = "start", 
                               "stop" = "stop")) %>%
              left_join(nondbd_seq_table %>%
                          dplyr::select(ensembl_gene_id, ensembl_transcript_id,
                                        ipr_accession, ipr_description,
                                        start, stop,
                                        nondbd_sequence),
                        by = c("ensembl_gene_id" = "ensembl_gene_id", 
                               "ensembl_transcript_id" = "ensembl_transcript_id",
                               "ipr_accession" = "ipr_accession", 
                               "ipr_description" = "ipr_description",
                               "start" = "start", 
                               "stop" = "stop")) %>%              
              group_by(ensembl_gene_id, ensembl_transcript_id) %>%
              arrange(ipr_accession, start, stop, Molecular_function_general) %>%
              ungroup() %>%
              group_by(ensembl_gene_id, ensembl_transcript_id) %>%
              mutate(ipr_accession_list = paste0(ipr_accession,
                                                 collapse = "|"),
                     ipr_description_list = paste0(ipr_description,
                                                   collapse = "|"),
                     mol_mechanism_list = paste0(Molecular_mechanism,
                                                 collapse = "|"),
                     mol_function_general_list = paste0(Molecular_function_general,
                                                        collapse = "|"),
                     dbd_seq_list = paste0(dbd_sequence[!is.na(dbd_sequence)], 
                                           collapse = "|"),
                     nondbd_seq_list = paste0(nondbd_sequence[!is.na(nondbd_sequence)], 
                                              collapse = "|")) %>%
              ungroup() %>%
              dplyr::select(-ipr_accession, -ipr_description, -start, -stop,
                            -Molecular_mechanism, -Molecular_function, -Molecular_function_general,
                            -Comment, -dbd_sequence, -nondbd_sequence) %>%
              distinct()

isoform.switch.master.table = switch.table %>%
  dplyr::select(-switch_detected) %>%
  dplyr::rename("ensembl_transcript_id1" = "isof_id1",
                "ensembl_transcript_id2" = "isof_id2",
                "tissue_list1" = "isof1_tissues",
                "tissue_list2" = "isof2_tissues") %>%
  left_join(isoform.expression.tf.annot.dom.isof.dom_annot %>%
              dplyr::select(ensembl_gene_id, humantfs_gene_name, tf_family,
                            ensembl_transcript_id, n_dbd, isoform_rank),
            by = c("ensembl_gene_id" = "ensembl_gene_id",
                   "ensembl_transcript_id1" = "ensembl_transcript_id")) %>%
  dplyr::rename("n_dbd1" = "n_dbd",
                "isoform_rank1" = "isoform_rank") %>%
  left_join(isoform.expression.tf.annot.dom.isof.dom_annot %>%
              dplyr::select(ensembl_gene_id, 
                            ensembl_transcript_id, n_dbd, isoform_rank),
            by = c("ensembl_gene_id" = "ensembl_gene_id",
                   "ensembl_transcript_id2" = "ensembl_transcript_id")) %>%
  dplyr::rename("n_dbd2" = "n_dbd",
                "isoform_rank2" = "isoform_rank") %>%
  left_join(per.isof.domain.table,
            by = c("ensembl_gene_id" = "ensembl_gene_id",
                   "ensembl_transcript_id1" = "ensembl_transcript_id")) %>%
  dplyr::rename("ipr_accession_list1" = "ipr_accession_list",
                "ipr_description_list1" = "ipr_description_list",
                "mol_mechanism_list1" = "mol_mechanism_list",
                "mol_function_general_list1" = "mol_function_general_list",
                "dbd_seq_list1" = "dbd_seq_list",
                "nondbd_seq_list1" = "nondbd_seq_list",
                "protein_accession1" = "protein_accession") %>%
  left_join(per.isof.domain.table,
            by = c("ensembl_gene_id" = "ensembl_gene_id",
                   "ensembl_transcript_id2" = "ensembl_transcript_id")) %>%
  dplyr::rename("ipr_accession_list2" = "ipr_accession_list",
                "ipr_description_list2" = "ipr_description_list",
                "mol_mechanism_list2" = "mol_mechanism_list",
                "mol_function_general_list2" = "mol_function_general_list",
                "dbd_seq_list2" = "dbd_seq_list",
                "nondbd_seq_list2" = "nondbd_seq_list",
                "protein_accession2" = "protein_accession") %>%
  left_join(ens99_pep %>%
              dplyr::select(protein_accession, protein_sequence), 
            by = c("protein_accession1" = "protein_accession")) %>%
  dplyr::rename("protein_sequence1" = "protein_sequence") %>%
  left_join(ens99_pep %>%
              dplyr::select(protein_accession, protein_sequence), 
            by = c("protein_accession2" = "protein_accession")) %>%
  dplyr::rename("protein_sequence2" = "protein_sequence") %>%
  left_join(cluster.annot.df.final,
            by = c("ensembl_gene_id" = "ensembl_gene_id",
                   "humantfs_gene_name" = "humantfs_gene_name",
                   "tf_family" = "tf_family")) %>%
  dplyr::select(ensembl_gene_id,
                humantfs_gene_name,
                tf_family,
                cluster_number,
                ensembl_transcript_id1,
                ensembl_transcript_id2,
                protein_accession1,
                protein_accession2,
                n_dbd1,
                n_dbd2,
                isoform_rank1,
                isoform_rank2,
                tissue_list1,
                tissue_list2,
                ipr_accession_list1,
                ipr_accession_list2,
                ipr_description_list1,
                ipr_description_list2,
                mol_mechanism_list1,
                mol_mechanism_list2,
                mol_function_general_list1,
                mol_function_general_list2,
                dbd_seq_list1,
                dbd_seq_list2,
                nondbd_seq_list1,
                nondbd_seq_list2,
                protein_sequence1,
                protein_sequence2) %>%
  arrange(ensembl_gene_id,
          cluster_number,
          ensembl_transcript_id1,
          ensembl_transcript_id2)
```

Annotate the master switch table with different switch properties:

```{r, include=T}
add_element_counts = function(v) {
  return(as.data.frame(table(v)) %>% 
           rowwise() %>% 
           mutate(with_count = paste0(v, " (", Freq, ")")) %>%
           pull(with_count))
}

normalize_switches = function(switched_domains) {
  switch_dict = c()
  switched_domains_norm = c()
  
  for (s in switched_domains) {
    if (s %in% switch_dict) {
      switched_domains_norm = c(switched_domains_norm, s)
    } else {
      s.rev = paste0(unlist(stringr::str_split(s, fixed(" | ")))[2], " | ", unlist(stringr::str_split(s, fixed(" | ")))[1])
      if (s.rev %in% switch_dict) {
        switched_domains_norm = c(switched_domains_norm, s.rev)
      } else {
        switch_dict = c(switch_dict, s)
        switched_domains_norm = c(switched_domains_norm, s)
      }
    }
  }
  
  return(switched_domains_norm)
}

find_changed_nondbds = function(switch.df) {
  mol.function.general.list = unlist(stringr::str_split(switch.df[1, "mol_function_general_list1"], fixed("|")))
  nondbd.seq.list1 = unlist(stringr::str_split(switch.df[1, "nondbd_seq_list1"], fixed("|")))
  nondbd.seq.list2 = unlist(stringr::str_split(switch.df[1, "nondbd_seq_list2"], fixed("|")))
  
  i = 1
  j = 1
  i.max = length(mol.function.general.list)
  result.left.side = ""
  result.right.side = ""
  while (i <= i.max) {
    if (mol.function.general.list[i] != "Canonical DBD") {
      if (nondbd.seq.list1[j] != nondbd.seq.list2[j]) {
        result.left.side = paste0(result.left.side, 
                                  ifelse(result.left.side == "", "", "; "),
                                  mol.function.general.list[i], 
                                  " 1")
        result.right.side = paste0(result.right.side, 
                                   ifelse(result.right.side == "", "", "; "),
                                   mol.function.general.list[i], 
                                   " 2")
      }
      j = j + 1
    }
    i = i + 1
  }
  
  result.whole = paste0(result.left.side, " | ", result.right.side)
  return(result.whole)
}

summarise_switch = function(clean.switch) {
  if (clean.switch == "|") {
    return(clean.switch)
  }
  
  left.side = unlist(stringr::str_split(clean.switch, fixed(" | ")))[1]
  right.side = unlist(stringr::str_split(clean.switch, fixed(" | ")))[2]
  if (left.side == "") {
    summarised.switch = paste0(right.side, " | ")
  } else {
    summarised.switch = clean.switch
  }
  
  return(summarised.switch)
}

isoform.switch.master.table.annot = isoform.switch.master.table %>%
  mutate(switch_type = ifelse(stringr::str_detect(mol_function_general_list1, "DBD") & 
                              stringr::str_detect(mol_function_general_list2, "DBD"),
                              "+/+",
                              ifelse(!stringr::str_detect(mol_function_general_list1, "DBD") & 
                                     !stringr::str_detect(mol_function_general_list2, "DBD"),
                                     "-/-",
                                     "+/-"))) %>%
  mutate(protein_seq_changed = (protein_sequence1 != protein_sequence2)) %>%
  mutate(dbd_seq_changed = (dbd_seq_list1 != dbd_seq_list2)) %>%
  mutate(nondbd_seq_changed = (nondbd_seq_list1 != nondbd_seq_list2)) %>%
  mutate(mol_mechanism_list1 = ifelse(mol_mechanism_list1 == "NA", NA, mol_mechanism_list1),
         mol_mechanism_list2 = ifelse(mol_mechanism_list2 == "NA", NA, mol_mechanism_list2),
         mol_function_general_list1 = ifelse(mol_function_general_list1 == "NA", NA, mol_function_general_list1),
         mol_function_general_list2 = ifelse(mol_function_general_list2 == "NA", NA, mol_function_general_list2)) %>%
  rowwise() %>%
  mutate(switched_domains = paste0(c(paste0(add_element_counts(unlist(stringr::str_split(mol_function_general_list1, fixed("|")))),
                                            collapse = "; "),
                                     paste0(add_element_counts(unlist(stringr::str_split(mol_function_general_list2, fixed("|")))),
                                            collapse = "; ")),
                                   collapse = " | ")) %>%
  bind_cols(data.frame(switched_domains_norm = normalize_switches(.$switched_domains))) %>%
  mutate(switched_domains_norm_clean = ifelse(unlist(stringr::str_split(switched_domains_norm, fixed(" | ")))[1] ==
                                              unlist(stringr::str_split(switched_domains_norm, fixed(" | ")))[2],
                                              "|",
                                              switched_domains_norm)) %>%
  mutate(switched_domains_norm_clean = ifelse((switched_domains_norm_clean == "|") & dbd_seq_changed,
                                              "Canonical DBD(s) 1 | Canonical DBD(s) 2",
                                              switched_domains_norm_clean)) %>%
  group_by(ensembl_gene_id, ensembl_transcript_id1, ensembl_transcript_id2,
           ipr_accession_list1, ipr_accession_list2,
           protein_sequence1, protein_sequence2) %>%
  do(mutate(., nondbd_seq_switch = ifelse((switched_domains_norm_clean == "|") & nondbd_seq_changed,
                                            find_changed_nondbds(.),
                                            ""))) %>%
  ungroup() %>% 
  mutate(switched_domains_norm_clean_seq = ifelse((switched_domains_norm_clean == "|") & (nondbd_seq_switch != ""),
                                                  nondbd_seq_switch,
                                                  switched_domains_norm_clean)) %>%
  mutate(switched_domains_norm_clean_seq = ifelse(switched_domains_norm_clean_seq == "|", 
                                                  switched_domains_norm, 
                                                  switched_domains_norm_clean_seq)) %>%
  group_by(switched_domains_norm_clean_seq) %>%
  do(mutate(., summarised_switch = summarise_switch(unique(.$switched_domains_norm_clean_seq)))) %>%
  ungroup()
```

Print out switch stats:

```{r, include=T}
cat("Number of TFs that switch major isoforms: ", 
    isoform.switch.master.table.annot %>% 
      pull(ensembl_gene_id) %>% 
      unique() %>% 
      length(),
    "\n")

cat("Total number of switches: ",
    nrow(isoform.switch.master.table.annot),
    "\n")

cat("Number of DBD+ to DBD+ switches: ",
    isoform.switch.master.table.annot %>%
      filter(switch_type == "+/+") %>%
      nrow(),
    "\n")

cat("Number of DBD+ to DBD- switches: ",
    isoform.switch.master.table.annot %>%
      filter(switch_type == "+/-") %>%
      nrow(),
    "\n")

cat("Number of DBD- to DBD- switches: ",
    isoform.switch.master.table.annot %>%
      filter(switch_type == "-/-") %>%
      nrow(),
    "\n")

cat("Number of switches (among all +/+ and -/-) that do not change domains: ",
    isoform.switch.master.table.annot %>%
      filter(!dbd_seq_changed & !nondbd_seq_changed) %>%
      nrow(),
    "\n")

cat("Number of switches (among all +/+ and -/-) not changing protein sequence: ",
    isoform.switch.master.table.annot %>%
      filter(!protein_seq_changed) %>%
      nrow(),
    "\n")

cat("Number of +/+ switches that do not change domains: ",
    isoform.switch.master.table.annot %>%
      filter(switch_type == "+/+") %>%
      filter(!dbd_seq_changed & !nondbd_seq_changed) %>%
      nrow(),
    "\n")

cat("Number of +/+ switches that do not change protein sequence: ",
    isoform.switch.master.table.annot %>%
      filter(switch_type == "+/+") %>%
      filter(!protein_seq_changed) %>%
      nrow(),
    "\n")
```

Hence, the majority of switches occur between two DBD+ major isoforms. Interestingly, in 181 switches (which comprise 61% of all +/+ switches!) domains do not change; 181 - 63 = 118 of these switches change the protein sequence between domains (hence, potentially changing functional IDRs) and the other 63 switches change only UTRs (in this way, potentially changing mRNA stability or translational efficiency).

Plot the fractions of +/-, -/- and +/+ switches, overall and per cluster (Fig. S5D, S5H). Also generate a plot that shows the proportions of +/+ switches that (a) change domains, (b) change the proteins sequence outside of domains, (c) change UTRs only. I excluded -/- switches from the last plot because their number is minute, and it will not be clear if they take part in the proportions (a), (b) and (c) at all.

```{r, include=T}
p = isoform.switch.master.table.annot %>%
  count(switch_type) %>%
  mutate(switch_type_fraction = n / sum(n)) %>%
  mutate(plot_feature = "switch") %>%
  mutate(switch_type = factor(switch_type, levels = c("+/-", "-/-", "+/+"))) %>%
  ggplot() +
    geom_col(aes(x = plot_feature, y = switch_type_fraction, fill = switch_type)) +
    scale_fill_manual(values = c("purple", "blue", "red")) +
    theme_classic()

ggsave(filename = "data/results/switch_type_fractions_barplot_2.pdf",
       plot = p,
       width = 4,
       height = 6)

p = isoform.switch.master.table.annot %>%
  add_count(switch_type) %>%
  dplyr::rename("total_n" = "n") %>%
  filter(switch_type == "+/+") %>%
  mutate(switch_group = ifelse(dbd_seq_changed | nondbd_seq_changed,
                               "Change in domains",
                               ifelse(protein_seq_changed,
                                      "Change outside of domains",
                                      "Change only in UTRs"))) %>%
  add_count(switch_group) %>%
  dplyr::select(switch_group, n, total_n) %>%
  distinct() %>%
  mutate(switch_group_fraction = n / total_n) %>%
  mutate(plot_feature = "switch") %>%
  mutate(switch_group = factor(switch_group, levels = rev(c("Change in domains", 
                                                            "Change outside of domains", 
                                                            "Change only in UTRs")))) %>%
  ggplot() +
    geom_col(aes(x = plot_feature, y = switch_group_fraction, fill = switch_group)) +
    scale_fill_manual(values = c("orange", "dark green", "grey")) +
    theme_classic()

ggsave(filename = "data/results/dbdpos_dbdpos_switch_group_fractions_barplot_2.pdf",
       plot = p,
       width = 4,
       height = 6)

per.cluster.type.summaries = isoform.switch.master.table.annot %>%
  mutate(switch_type = factor(switch_type, levels = rev(c("+/-", "-/-", "+/+")))) %>%
  # mutate(switch_type = factor(switch_type, levels = c("+/-", "-/-", "+/+"))) %>%
  dplyr::select(cluster_number,
                switch_type) %>%
  group_by(cluster_number) %>%
  add_count(switch_type) %>%
  ungroup() %>%
  arrange(cluster_number, switch_type) %>%
  distinct() %>%
  dplyr::rename("switch_count" = "n") %>%
  group_by(cluster_number) %>%
  mutate(total_switch_count = sum(switch_count)) %>%
  ungroup() %>%
  mutate(switch_fraction = switch_count / total_switch_count)

p = per.cluster.type.summaries %>%
  ggplot() +
    geom_col(aes(x = cluster_number, y = switch_fraction, fill = switch_type, group = cluster_number)) +
    scale_fill_manual(values = rev(c("purple", "blue", "red"))) +
    # scale_fill_manual(values = rev(c("#EE766F", "#3DAB3D", "#6E94CD"))) +
    theme_classic()

ggsave(filename = "data/results/switch_type_fractions_per_cluster_barplot_2.pdf",
       plot = p,
       width = 15,
       height = 6)
```

Rate the most common switch types:

```{r, include=T}
isoform.switch.master.table.annot %>%
  count(summarised_switch) %>%
  arrange(desc(n)) %>%
  head(10)
```

Generate a table of switches for manual functional classification:

```{r, include=T}
reformat_domain_list = function(domain.list) {
  return(paste0(unique(sort(unlist(stringr::str_split(unlist(stringr::str_split(domain.list, 
                                                                                fixed("|"))),
                                                      fixed("; "))))), 
                collapse = "; "))
}

switch.types.table = isoform.switch.master.table.annot %>%
  dplyr::select(switch_type,
                mol_function_general_list1,
                mol_function_general_list2,
                dbd_seq_changed,
                nondbd_seq_changed,
                protein_seq_changed,
                summarised_switch) %>%
  rowwise() %>%
  mutate(., mol_function_general_list1 = reformat_domain_list(mol_function_general_list1)) %>%
  mutate(., mol_function_general_list2 = reformat_domain_list(mol_function_general_list2)) %>%
  left_join(dbdminus.domain.table.per_function.filled %>%
              dplyr::rename("general_function_list1" = "general_function_list",
                            "isoform_function1" = "isoform_function") %>%
              dplyr::select(-prediction_comment),
            by = c("mol_function_general_list1" = "general_function_list1")) %>%
  left_join(dbdminus.domain.table.per_function.filled %>%
              dplyr::rename("general_function_list2" = "general_function_list",
                            "isoform_function2" = "isoform_function") %>%
              dplyr::select(-prediction_comment),
            by = c("mol_function_general_list2" = "general_function_list2")) %>%
  mutate(dbdminus_isoform_function = ifelse(!is.na(isoform_function1) & !is.na(isoform_function2),
                                            paste0(c(isoform_function1, isoform_function2), collapse = " | "),
                                            ifelse(is.na(isoform_function1),
                                                   isoform_function2,
                                                   isoform_function1))) %>%
  dplyr::select(-mol_function_general_list1,
                -mol_function_general_list2,
                -isoform_function1,
                -isoform_function2) %>%
  arrange(switch_type,
          protein_seq_changed,
          dbd_seq_changed,
          nondbd_seq_changed,
          summarised_switch) %>%
  distinct()

# Clean up table names

write.table(switch.types.table,
            file = "data/results/switch_types_table_raw8.tsv",
            quote = F,
            row.names = F,
            sep = "\t")
```

Rate the most common switch functions per switch type (Fig. 5D):

```{r, include=T}
switch.types.table.filled = read.delim("data/results/switch_types_table_filled9.tsv",
                                       header = T,
                                       sep = "\t",
                                       stringsAsFactors = F)

isoform.switch.master.table.annot.with_functions = isoform.switch.master.table.annot %>%
    left_join(switch.types.table.filled %>%
                dplyr::select(-dbdminus_isoform_function) %>%
                distinct(),
            by = c("switch_type" = "switch_type",
                   "dbd_seq_changed" = "dbd_seq_changed",
                   "nondbd_seq_changed" = "nondbd_seq_changed",
                   "protein_seq_changed" = "protein_seq_changed",
                   "summarised_switch" = "summarised_switch")) %>%
  mutate(general_functional_consequence = stringr::str_trim(general_functional_consequence, side = "both")) %>%
  mutate(general_functional_consequence = stringr::str_replace(general_functional_consequence,
                                                               fixed("+/- Catalytic activity"),
                                                               "Catalytic activity presence/absence")) %>%
  mutate(general_functional_consequence = stringr::str_replace(general_functional_consequence,
                                                               fixed("+/- Nucleic acid binding and change in DNA binding"),
                                                               "Nucleic acid binding presence/absence and change in DNA binding")) %>%
  mutate(general_functional_consequence = stringr::str_replace(general_functional_consequence,
                                                               fixed("+/- Protein binding"),
                                                               "Protein binding presence/absence")) %>%
  mutate(general_functional_consequence = stringr::str_replace(general_functional_consequence,
                                                               fixed("+/- Catalytic activity and change in DNA binding"),
                                                               "Catalytic activity presence/absence and change in DNA binding")) %>%
  mutate(general_functional_consequence = stringr::str_replace(general_functional_consequence,
                                                               fixed("+/- Oligomerization and change in DNA binding"),
                                                               "Oligomerization presence/absence and change in DNA binding")) %>%
  mutate(general_functional_consequence = stringr::str_replace(general_functional_consequence,
                                                               fixed("+/- Protein or nucleic acid binding and change in DNA binding"),
                                                               "Protein or nucleic acid binding presence/absence and change in DNA binding"))

switch.categories.overall.table = isoform.switch.master.table.annot.with_functions %>%
  dplyr::select(switch_type,
                general_functional_consequence) %>%
  group_by(switch_type) %>%
  add_count(general_functional_consequence) %>%
  ungroup() %>%
  dplyr::rename("switch_count" = "n",
                "switch_category" = "general_functional_consequence") %>%
  distinct() %>%
  arrange(switch_type, desc(switch_count)) %>%
  group_by(switch_type) %>%
  mutate(total_switch_count = sum(switch_count)) %>%
  ungroup() %>%
  mutate(switch_fraction = switch_count / total_switch_count) %>%
  mutate(switch_category = ifelse(switch_fraction >= 0.03, switch_category, "Other")) %>%
  group_by(switch_type, switch_category) %>%
  mutate(switch_fraction = ifelse(switch_category != "Other", switch_fraction, sum(switch_fraction))) %>%
  ungroup() %>%
  dplyr::select(-switch_count,
                -total_switch_count) %>%
  distinct()

switch.categories.overall.list = sort(unique(switch.categories.overall.table$switch_category))

p = switch.categories.overall.table %>%
  mutate(switch_category = factor(switch_category, levels = switch.categories.overall.list)) %>%
  mutate(switch_type = factor(switch_type, levels = c("+/+", "-/-", "+/-"))) %>%
  ggplot() +
    geom_point(aes(x = switch_category, y = switch_type, colour = switch_type, 
                   size = switch_fraction)) +
    scale_colour_manual(values = c("red", "blue", "purple")) +
    scale_size("Switch fraction", breaks = c(0.1, 0.3, 0.5)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

ggsave(filename = "data/results/function_consequence_fractions_scatter_plot_3pct.pdf",
       plot = p,
       width = 6,
       height = 4)
```

The exact frequencies are as follows:

```{r, include=T}
switch.categories.overall.table %>%
  mutate(switch_category = factor(switch_category, levels = switch.categories.overall.list)) %>%
  mutate(switch_type = factor(switch_type, levels = c("+/+", "-/-", "+/-")))
```

Hence, among the 298 DBD+/DBD+ switches, the majority do not change domains ("Change outside of domains", 40%, and "Change in UTRs", 21%; ~61% in total). The first of these two categories motivates us to look into possible changes in intrinsically disordered regions (IDRs), while the second category suggests that switching major isoforms may have different stability or translational efficiency. Also, 18% of DBD+/DBD+ switches change the DBD sequence(s). Consequently, we will look into DBD changes and try to interpret them.

The 13 DBD-/DBD- switches are spread among four categories, and "Change outside of domains" (54%) and "Activator / Domainless protein" (23%) are the two biggest ones. The "Activator / Domainless protein" category here is generated by two C2H2 zinc fingers, ZNF268 and ZNF655, whose switching DBD- isoforms gain / lose the KRAB domain:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter((switch_type == "-/-") & (general_functional_consequence == "Activator / Domainless protein")) %>%
  dplyr::select(ensembl_gene_id, 
                humantfs_gene_name, 
                tf_family, 
                cluster_number, 
                ensembl_transcript_id1, 
                ensembl_transcript_id2,
                switch_type,
                ipr_description_list1,
                ipr_description_list2,
                general_functional_consequence,
                consequence_comment)
```

The fact that domainless isoforms become major suggests that we need to check if these isoforms have IDRs (and so can still be functional even without domains).

Finally, among the 89 DBD+/DBD- switches the vast majority are in the "Reverse regulation" (48%) and "TF / Domainless protein" (35%) categories (83% in total). The "Reverse regulation" category here is derived based on the known cases of alternative isoforms with different functions from Fig. 2B-C. The switch category "TF / Domainless protein" is generated by TFs from different families, but more than half of them are C2H2 ZFs who gain / lose their C2H2 zinc fingers:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter((switch_type == "+/-") & (general_functional_consequence == "TF / Domainless protein")) %>%
  dplyr::select(ensembl_gene_id, 
                humantfs_gene_name, 
                tf_family, 
                cluster_number, 
                ensembl_transcript_id1, 
                ensembl_transcript_id2,
                switch_type,
                ipr_description_list1,
                ipr_description_list2,
                general_functional_consequence,
                consequence_comment)
```

The fact that this category is the second biggest in the DBD+/DBD- switches again supports our plan to check the presence of IDRs in domainless TF isoforms (not only major ones). Importantly, among DBD+/DBD- switches we can also have those changing non-DBDs, UTRs and sequences between domains. However, if DBDs change in a switch, we regard this change as the most important one and do not account for other possible changes in the same switch.

Plot the same category summaries per the most frequent switch type in each TF cluster (Fig. 5C):

```{r, include=T}
switch.categories.per_type_per_cluster.table = isoform.switch.master.table.annot.with_functions %>%
  dplyr::select(switch_type,
                cluster_number,
                general_functional_consequence) %>%
  group_by(switch_type, cluster_number) %>%
  add_count(general_functional_consequence) %>%
  ungroup() %>%
  dplyr::rename("switch_count" = "n",
                "switch_category" = "general_functional_consequence") %>%
  distinct() %>%
  group_by(switch_type, cluster_number) %>%
  mutate(total_switch_count = sum(switch_count)) %>%
  ungroup() %>%
  mutate(switch_fraction = switch_count / total_switch_count) %>%
  mutate(switch_category = ifelse(switch_fraction >= 0.1, switch_category, "Other")) %>%
  group_by(switch_type, cluster_number, switch_category) %>%
  mutate(switch_fraction = ifelse(switch_category != "Other", switch_fraction, sum(switch_fraction))) %>%
  ungroup() %>%
  dplyr::select(-switch_count,
                -total_switch_count) %>%
  distinct() %>%
  filter((cluster_number == 1) & (switch_type == "+/-") |
         (cluster_number == 2) & (switch_type == "+/-") |
         (cluster_number == 3) & (switch_type == "+/+") |
         (cluster_number == 4) & (switch_type == "+/+") |
         (cluster_number == 5) & (switch_type == "+/+") |
         (cluster_number == 6) & (switch_type == "+/+") |
         (cluster_number == 7) & (switch_type == "+/+"))
  
switch.categories.per_type_per_cluster.list = sort(unique(switch.categories.per_type_per_cluster.table$switch_category))

p.per_type_per_cluster = switch.categories.per_type_per_cluster.table %>%
  mutate(switch_category = factor(switch_category, levels = switch.categories.per_type_per_cluster.list)) %>%
  mutate(switch_type = factor(switch_type, levels = c("+/+", "+/-"))) %>%
  ggplot() +
    geom_point(aes(x = cluster_number, y = switch_category, colour = switch_type,
                   size = switch_fraction)) +
    scale_colour_manual(values = c("red", "purple")) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

ggsave(filename = "data/results/function_consequence_fractions_scatter_plot_per_type_per_cluster_10pct.pdf",
       plot = p.per_type_per_cluster,
       width = 14,
       height = 4)
```

Hence, the most frequent categories of DBD-/DBD- switches in clusters 1 and 2 just reflect the overall frequencies of categories. Interestingly, among DBD+/DBD+ switches only those in clusters 3, 4 and 5 change DNA binding. These clusters contain TFs that are broadly expressed either across all tissues or across the non-brain 2 cluster. Additionally, DBD+/DBD+ switches in cluster 4 and some parts of cluster 5 occur between the brain and non-brain tissues, and consequently DNA-binding is changing between these groups of tissues. Also very interestingly, cluster 7 is the only one (out of clusters 3-7 where we consider DBD+/DBD+ swithces) whose +/+ switches change transcription repression, and intriguingly this happens between the brain and non-brain tissues. We will look into these clusters in more detail below. 

Plot all category summaries per switch type (+/+, +/-, -/-) per TF cluster (Fig. S5H):

```{r, include=T}
switch.categories.per_type.table = isoform.switch.master.table.annot.with_functions %>%
  dplyr::select(switch_type,
                cluster_number,
                general_functional_consequence) %>%
  group_by(switch_type, cluster_number) %>%
  add_count(general_functional_consequence) %>%
  ungroup() %>%
  dplyr::rename("switch_count" = "n",
                "switch_category" = "general_functional_consequence") %>%
  distinct() %>%
  group_by(switch_type, cluster_number) %>%
  mutate(total_switch_count = sum(switch_count)) %>%
  ungroup() %>%
  mutate(switch_fraction = switch_count / total_switch_count) %>%
  mutate(switch_category = ifelse(switch_fraction >= 0.1, switch_category, "Other")) %>%
  group_by(switch_type, cluster_number, switch_category) %>%
  mutate(switch_fraction = ifelse(switch_category != "Other", switch_fraction, sum(switch_fraction))) %>%
  ungroup() %>%
  dplyr::select(-switch_count,
                -total_switch_count) %>%
  distinct() %>%
  mutate(switch_type = factor(switch_type, levels = c("+/+", "-/-", "+/-"))) %>%
  arrange(switch_type, switch_category, cluster_number) %>%
  mutate(switch_type_category = paste0(switch_type, "_", switch_category))
  
switch.type.categories.per_type.list = unique(switch.categories.per_type.table$switch_type_category)

p.per_type = switch.categories.per_type.table %>%
  mutate(switch_type_category = factor(switch_type_category, levels = switch.type.categories.per_type.list)) %>%
  mutate(switch_type = factor(switch_type, levels = c("+/+", "-/-", "+/-"))) %>%
  ggplot() +
    geom_point(aes(x = cluster_number, y = switch_type_category, colour = switch_type,
                   size = switch_fraction)) +
    #scale_colour_manual(values = rev(c("#EE766F", "#3DAB3D", "#6E94CD"))) +
    scale_colour_manual(values = c("red", "blue", "purple")) +
    scale_size("Switch fraction", breaks = c(0.15, 0.25, 0.50, 0.75, 1.00)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

ggsave(filename = "data/results/function_consequence_fractions_scatter_plot_per_type_10pct.pdf",
       plot = p.per_type,
       width = 14,
       height = 8)
```

## Analyse overrepresentation of switching TFs in TF families and GO terms

Use a hypergeometric test to find if any TF families or GO terms (gene sets) are over-represented among switching TFs (a signature), in comparison to all multi-isoform TFs (the background):

```{r, include=T}
library(hypeR)

# switching.tfs.test.table = isoform.expression.tf.annot.dom.isof %>%
#   dplyr::select(-isoform.expressed, -is_dominant) %>%
#   filter(expressed.isoform.n > 1) %>%
#   group_by(ensembl_gene_id) %>%
#   do(mutate(., dom_isoform_count = nrow(.))) %>%
#   ungroup() %>%
#   dplyr::select(tf_family,
#                 ensembl_gene_id,
#                 humantfs_gene_name,
#                 expressed.isoform.n,
#                 dom_isoform_count) %>%
#   distinct()

# switching.tfs = switching.tfs.test.table %>%
#   filter(dom_isoform_count > 1) %>%
#   pull(humantfs_gene_name) %>%
#   unique()

switching.tfs = isoform.switch.master.table.annot.with_functions %>%
  pull(humantfs_gene_name) %>%
  unique()
# 
# multiisoform.tfs = switching.tfs.test.table %>%
#   pull(humantfs_gene_name) %>%
#   unique()

multiisoform.tfs = isoform.expression.tf.annot.ranks.expressed %>%
  dplyr::select(humantfs_gene_name, expressed.isoform.n) %>%
  distinct() %>%
  filter(expressed.isoform.n > 1) %>%
  pull(humantfs_gene_name)

# multiisoform.tf.families = switching.tfs.test.table %>%
#   pull(tf_family) %>%
#   unique()

# multiisoform.tf.families = isoform.expression.tf.annot.ranks.expressed %>%
#   dplyr::select(tf_family, expressed.isoform.n) %>%
#   distinct() %>%
#   filter(expressed.isoform.n > 1) %>%
#   pull(tf_family) %>%
#   unique()

tf.families.expressed = isoform.expression.tf.annot.ranks.expressed %>%
  filter(isoform.expressed) %>%
  pull(tf_family) %>%
  unique()


tf.family.sets = lapply(tf.families.expressed, # multiisoform.tf.families,
                        function(family.name) {
                          isoform.expression.tf.annot.ranks.expressed %>%
                            #filter(expressed.isoform.n > 1) %>%
                            filter(tf_family == family.name) %>%
                            filter(isoform.expressed) %>%
                            pull(humantfs_gene_name) %>%
                            unique()
                          # switching.tfs.test.table %>%
                          #   filter(tf_family == family.name) %>%
                          #   pull(humantfs_gene_name)
                        })

names(tf.family.sets) = tf.families.expressed

tic("Hypergeometric test of the overrepresentation of TF families among switching TFs")
sw.tf.overr = hypeR(signature = switching.tfs,
                    genesets = tf.family.sets,
                    test = "hypergeometric",
                    background = multiisoform.tfs,
                    fdr = 0.05,
                    quiet = FALSE)
toc()

hyp_show(sw.tf.overr)
```

Hence, no TF families are over-represented among switching TFs with adjusted p-value (FDR) < 5%. 

Check GO Biological Process terms:

```{r, include=T}
go.bp.sets = msigdb_gsets(species = "Homo sapiens", category = "C5", subcategory = "GO:BP")

tic("Hypergeometric test of the overrepresentation of GO BP terms among switching TFs")
sw.go.bp.overr = hypeR(signature = switching.tfs,
                       genesets = go.bp.sets,
                       test = "hypergeometric",
                       background = multiisoform.tfs,
                       fdr = 0.05,
                       quiet = FALSE)
toc()

hyp_show(sw.go.bp.overr)
```

No over-represented GO BP terms with adjusted p-value (FDR) < 5%.

Check GO Molecular Function terms:

```{r, include=T}
go.mf.sets = msigdb_gsets(species = "Homo sapiens", category = "C5", subcategory = "GO:MF")

tic("Hypergeometric test of the overrepresentation of GO MF terms among switching TFs")
sw.go.mf.overr = hypeR(signature = switching.tfs,
                       genesets = go.mf.sets,
                       test = "hypergeometric",
                       background = multiisoform.tfs,
                       fdr = 0.05,
                       quiet = FALSE)
toc()

hyp_show(sw.go.mf.overr)
```

No over-represented GO MF terms with adjusted p-value (FDR) < 5%.

Finally, check GO Cellular Component terms:

```{r, include=T}
go.cc.sets = msigdb_gsets(species = "Homo sapiens", category = "C5", subcategory = "GO:CC")

tic("Hypergeometric test of the overrepresentation of GO CC terms among switching TFs")
sw.go.cc.overr = hypeR(signature = switching.tfs,
                       genesets = go.cc.sets,
                       test = "hypergeometric",
                       background = multiisoform.tfs,
                       fdr = 0.05,
                       quiet = FALSE)
toc()

hyp_show(sw.go.cc.overr)
```

No over-represented GO CC terms with adjusted p-value (FDR) < 5%.

Next, check if any TF families or GO terms are over-represented in each of the switching TF clusters.

First, check TF families:

```{r, include=T}
switching.tf.families = isoform.expression.tf.annot.ranks.expressed %>%
  filter(humantfs_gene_name %in% switching.tfs) %>%
  pull(tf_family) %>%
  unique()

tf.family.sw.sets = lapply(switching.tf.families,
                           function(family.name) {
                             isoform.expression.tf.annot.ranks.expressed %>%
                               filter(tf_family == family.name) %>%
                               filter(isoform.expressed) %>%
                               pull(humantfs_gene_name) %>%
                               unique()
                             # switching.tfs.test.table %>%
                             #   filter(dom_isoform_count > 1) %>%
                             #   filter(tf_family == family.name) %>%
                             #   pull(humantfs_gene_name) %>%
                             #   unique()
                           })

names(tf.family.sw.sets) = switching.tf.families

tic("Hypergeometric test of the overrepresentation of TF families among switching TFs")
cluster.numbers = sort(unique(isoform.switch.master.table.annot.with_functions$cluster_number))
hyp_obj_list = lapply(cluster.numbers,
                      function(cn) {
                        cluster.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(cluster_number == cn) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.cl.overr = hypeR(signature = cluster.tfs,
                                            genesets = tf.family.sw.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            fdr = 0.05,
                                            quiet = FALSE)
                        sw.cl.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
# hyp_show(hyp_obj_list[[8]])
# hyp_show(hyp_obj_list[[9]])
# hyp_show(hyp_obj_list[[10]])
# hyp_show(hyp_obj_list[[11]])
toc()
```

So, there is no over-representation of any TF family in any cluster.

Next, check if any GO Biological Process terms are over-represented in any of the clusters:

```{r, include=T}
tic("Hypergeometric test of the overrepresentation of GO BP terms in each cluster of switching TFs")
cluster.numbers = sort(unique(isoform.switch.master.table.annot.with_functions$cluster_number))
hyp_obj_list = lapply(cluster.numbers,
                      function(cn) {
                        cluster.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(cluster_number == cn) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.cl.overr = hypeR(signature = cluster.tfs,
                                            genesets = go.bp.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            fdr = 0.05,
                                            quiet = FALSE)
                        sw.cl.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
toc()
```

No over-represented GO BP terms with adjusted p-value (FDR) < 5%.

Next, check if any GO Molecular Function terms are over-represented in any of the clusters:

```{r, include=T}
tic("Hypergeometric test of the overrepresentation of GO MF terms in each cluster of switching TFs")
cluster.numbers = sort(unique(isoform.switch.master.table.annot.with_functions$cluster_number))
hyp_obj_list = lapply(cluster.numbers,
                      function(cn) {
                        cluster.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(cluster_number == cn) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.cl.overr = hypeR(signature = cluster.tfs,
                                            genesets = go.mf.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            fdr = 0.05,
                                            quiet = FALSE)
                        sw.cl.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
toc()
```

No over-represented GO MF terms with adjusted p-value (FDR) < 5%.

Finally, check if any GO Cellular Compartment terms are over-represented in any of the clusters:

```{r, include=T}
tic("Hypergeometric test of the overrepresentation of GO CC terms in each cluster of switching TFs")
cluster.numbers = sort(unique(isoform.switch.master.table.annot.with_functions$cluster_number))
hyp_obj_list = lapply(cluster.numbers,
                      function(cn) {
                        cluster.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(cluster_number == cn) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.cl.overr = hypeR(signature = cluster.tfs,
                                            genesets = go.cc.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            fdr = 0.05,
                                            quiet = FALSE)
                        sw.cl.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
toc()
```

No over-represented GO CC terms with adjusted p-value (FDR) < 5%.

Overall, no particular TF families or GO terms are over-represented among all switching TFs or any clusters of switching TFs. It means that TFs switch their major isoforms across adult human tissues independently of their family, or roles in particular pathways, or tissue specificity of their expression (as we saw in the heatmap of the clusters, some switching TFs are expressed in all tissues while others are very tissue-specific). Additionally, switching TFs exist in 

```{r, include=T}
switching.tf.families.n = isoform.expression.tf.annot.ranks.expressed %>%
  filter(humantfs_gene_name %in% switching.tfs) %>%
  pull(tf_family) %>%
  unique() %>%
  length()

switching.tf.families.n
```

out of

```{r, include=T}
expressing.tf.families.n = isoform.expression.tf.annot.ranks.expressed %>%
  filter(isoform.expressed) %>%
  pull(tf_family) %>%
  unique() %>%
  length()

expressing.tf.families.n
```

TF families that are expressed in adult human tissues. So, switching TF families comprise 

```{r, include=T}
switching.tf.families.n / expressing.tf.families.n * 100
```

percent of all TF families expressed in adult human tissues. Hence, major isoform switching is wide-spread across TF families.

Finally, let us check if any TF families or GO terms are over-represented in functional categories of switches.

First, check if any TF families are over-represented in functional categories of switches:

```{r, include=T}
tic("Hypergeometric test of the overrepresentation of TF families among switch functional categories")
switch.categories = sort(unique(isoform.switch.master.table.annot.with_functions$general_functional_consequence))
hyp_obj_list = lapply(switch.categories,
                      function(sc) {
                        switch.category.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(general_functional_consequence == sc) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.sc.overr = hypeR(signature = switch.category.tfs,
                                            genesets = tf.family.sw.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            #fdr = 0.05,
                                            quiet = FALSE)
                        sw.sc.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
hyp_show(hyp_obj_list[[8]])
hyp_show(hyp_obj_list[[9]])
hyp_show(hyp_obj_list[[10]])
hyp_show(hyp_obj_list[[11]])
hyp_show(hyp_obj_list[[12]])
hyp_show(hyp_obj_list[[13]])
hyp_show(hyp_obj_list[[14]])
hyp_show(hyp_obj_list[[15]])
hyp_show(hyp_obj_list[[16]])
hyp_show(hyp_obj_list[[17]])
hyp_show(hyp_obj_list[[18]])
hyp_show(hyp_obj_list[[19]])
hyp_show(hyp_obj_list[[20]])
hyp_show(hyp_obj_list[[21]])
hyp_show(hyp_obj_list[[22]])
hyp_show(hyp_obj_list[[23]])
hyp_show(hyp_obj_list[[24]])
hyp_show(hyp_obj_list[[25]])
hyp_show(hyp_obj_list[[26]])
hyp_show(hyp_obj_list[[27]])
hyp_show(hyp_obj_list[[28]])
hyp_show(hyp_obj_list[[29]])
hyp_show(hyp_obj_list[[30]])
hyp_show(hyp_obj_list[[31]])
hyp_show(hyp_obj_list[[32]])
toc()
```

```{r, include=T}
raw.pvalues.df = data.frame(functional.category = character(),
                            tf.family = character(),
                            raw.pvalue = numeric())

for (i in seq(1, 32)) {
  raw.pvalues.df %<>%
    bind_rows(data.frame(functional.category = switch.categories[i],
                         tf.family = hyp_obj_list[[i]]$data$label,
                         raw.pvalue = hyp_obj_list[[i]]$data$pval))
}

raw.pvalues.df %<>%
  mutate(padj = p.adjust(raw.pvalue, method = "BH"))

# FDR = 5%
raw.pvalues.df %>%
  filter(padj < 0.05)

# FDR = 10%
raw.pvalues.df %>%
  filter(padj < 0.1)

# FDR = 20%
raw.pvalues.df %>%
  filter(padj < 0.2)
```

Hence, some TF families are over-represented in some switch functional categories under FDR < 5%:

1. Nuclear receptors are over-represented in the category "Change in ligand binding." 

This over-representation is probably explained by the fact that the majority of ligand-binding domains are annotated in nuclear receptors:

```{r, include=T}
ligand.binding.domain.distrib = tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
  filter(ensembl_transcript_id %in% (isoform.expression.tf.annot.ranks.expressed %>%
                                       filter(isoform.expressed) %>%
                                       pull(ensembl_transcript_id))) %>% 
  left_join(domain_functional_classification,
            by = c("ipr_accession" = "ipr_accession")) %>%
  filter(stringr::str_detect(Molecular_function_general, fixed("Ligand binding"))) %>%
  dplyr::select(ensembl_gene_id, DBD) %>%
  distinct() %>%
  count(DBD) %>%
  arrange(desc(n))

ligand.binding.domain.distrib
```

Make a bar plot:

```{r, include=T}
p = ligand.binding.domain.distrib %>%
  mutate(N = sum(n)) %>%
  mutate(freq = n / N) %>%
  mutate(feature = "Family") %>%
  dplyr::select(DBD, freq, feature) %>%
  arrange(desc(freq)) %>%
  mutate(DBD = factor(DBD, levels = rev(DBD))) %>%
  ggplot() +
    geom_col(aes(x = feature,
                 y = freq,
                 fill = DBD)) +
    theme_classic()

ggsave(filename = "data/results/ligand_binding_domain_distribution_across_tf_families_barplot.pdf",
       plot = p,
       width = 3,
       height = 3)    
```

But it is important that nuclear receptors change their ligand-binding domains, because it can affect their function.

The change happens in the sequence of the domain:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(tf_family == "Nuclear receptor") %>%
  filter(general_functional_consequence == "Change in ligand binding") %>%
  dplyr::select(ensembl_gene_id, 
                humantfs_gene_name,
                ensembl_transcript_id1,
                ensembl_transcript_id2,
                summarised_switch)
```

2. MADS box TFs are over-represented in the category "Change in protein binding". The change is as follows:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(tf_family == "MADS box") %>%
  filter(general_functional_consequence == "Change in protein binding") %>%
  dplyr::select(ensembl_gene_id, 
                humantfs_gene_name,
                ensembl_transcript_id1,
                ensembl_transcript_id2,
                summarised_switch)
```

So, these two MADS boxes have some general protein binding domains whose sequence is changed in the switches.

3. C2H2 ZFs are over-represented in the "Reverse regulation" category:

```{r, include=T}
c2h2zfs.rr.switches = isoform.switch.master.table.annot.with_functions %>%
  filter(tf_family == "C2H2 ZF") %>%
  filter(general_functional_consequence == "Reverse regulation") %>%
  dplyr::select(ensembl_gene_id, 
                humantfs_gene_name,
                ensembl_transcript_id1,
                ensembl_transcript_id2,
                summarised_switch) %>%
  mutate(summarised_switch_clean = stringr::str_replace_all(summarised_switch, "\\s\\([0-9]++\\)", "")) %>%
  mutate(loss_type = ifelse(summarised_switch_clean %in% c("Canonical DBD; Transcription repression | Transcription repression",
                                                           "Canonical DBD; Dimerization | Dimerization",
                                                           "Canonical DBD; Oligomerization | Oligomerization",
                                                           "Oligomerization; Transcription repression | Canonical DBD; Oligomerization; Transcription repression",
                                                           "Transcription repression | Canonical DBD; Transcription repression"),
                            "Lost DBD",
                            ifelse(summarised_switch_clean %in% c("Canonical DBD; Transcription repression | Canonical DBD",
                                                                  "Canonical DBD | Canonical DBD; Transcription repression"),
                                   "Lost TR domain",
                                   "Other")))

c2h2zfs.rr.switches
```

Hence, most of the time the reversal of the regulation happens because of the loss of DBDs (C2H2 zinc fingers) or a transcription repression domain:

```{r, include=T}
p = c2h2zfs.rr.switches %>%
  mutate(N = nrow(.)) %>%
  group_by(loss_type) %>%
  do(mutate(., n = nrow(.))) %>%
  ungroup() %>%
  mutate(freq = n / N) %>%
  arrange(loss_type) %>%
  dplyr::select(loss_type, freq) %>%
  distinct() %>%
  mutate(feature = "switches") %>%
  mutate(loss_type = factor(loss_type, levels = rev(loss_type))) %>%
    ggplot() +
    geom_col(aes(x = feature,
                 y = freq,
                 fill = loss_type)) +
    theme_classic()

ggsave(filename = "data/results/c2h2zf_reverse_regulation_switches_barplot.pdf",
       plot = p,
       width = 3,
       height = 3)
```

Next, check if any GO Biological Process terms are over-represented in any of the switch functional categories:

```{r, include=T}
tic("Hypergeometric test of the overrepresentation of GO BP terms in each switch functional category of TFs")
hyp_obj_list = lapply(switch.categories,
                      function(sc) {
                        switch.category.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(general_functional_consequence == sc) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.sc.overr = hypeR(signature = switch.category.tfs,
                                            genesets = go.bp.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            fdr = 0.05,
                                            quiet = FALSE)
                        sw.sc.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
hyp_show(hyp_obj_list[[8]])
hyp_show(hyp_obj_list[[9]])
hyp_show(hyp_obj_list[[10]])
hyp_show(hyp_obj_list[[11]])
hyp_show(hyp_obj_list[[12]])
hyp_show(hyp_obj_list[[13]])
hyp_show(hyp_obj_list[[14]])
hyp_show(hyp_obj_list[[15]])
hyp_show(hyp_obj_list[[16]])
hyp_show(hyp_obj_list[[17]])
hyp_show(hyp_obj_list[[18]])
hyp_show(hyp_obj_list[[19]])
hyp_show(hyp_obj_list[[20]])
hyp_show(hyp_obj_list[[21]])
hyp_show(hyp_obj_list[[22]])
hyp_show(hyp_obj_list[[23]])
hyp_show(hyp_obj_list[[24]])
hyp_show(hyp_obj_list[[25]])
hyp_show(hyp_obj_list[[26]])
hyp_show(hyp_obj_list[[27]])
hyp_show(hyp_obj_list[[28]])
hyp_show(hyp_obj_list[[29]])
hyp_show(hyp_obj_list[[30]])
hyp_show(hyp_obj_list[[31]])
hyp_show(hyp_obj_list[[32]])
toc()
```

GO BP terms related to receptors and hormone signalling are over-represented in the category "Change in ligand binding," which is trivially explained by the fact that nuclear receptors are over-represented in this category.

Next, check if any GO Molecular Function terms are over-represented in any of the switch functional categories:

```{r, include=T}
tic("Hypergeometric test of the overrepresentation of GO MF terms in each switch functional category of TFs")
hyp_obj_list = lapply(switch.categories,
                      function(sc) {
                        switch.category.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(general_functional_consequence == sc) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.sc.overr = hypeR(signature = switch.category.tfs,
                                            genesets = go.mf.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            fdr = 0.05,
                                            quiet = FALSE)
                        sw.sc.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
hyp_show(hyp_obj_list[[8]])
hyp_show(hyp_obj_list[[9]])
hyp_show(hyp_obj_list[[10]])
hyp_show(hyp_obj_list[[11]])
hyp_show(hyp_obj_list[[12]])
hyp_show(hyp_obj_list[[13]])
hyp_show(hyp_obj_list[[14]])
hyp_show(hyp_obj_list[[15]])
hyp_show(hyp_obj_list[[16]])
hyp_show(hyp_obj_list[[17]])
hyp_show(hyp_obj_list[[18]])
hyp_show(hyp_obj_list[[19]])
hyp_show(hyp_obj_list[[20]])
hyp_show(hyp_obj_list[[21]])
hyp_show(hyp_obj_list[[22]])
hyp_show(hyp_obj_list[[23]])
hyp_show(hyp_obj_list[[24]])
hyp_show(hyp_obj_list[[25]])
hyp_show(hyp_obj_list[[26]])
hyp_show(hyp_obj_list[[27]])
hyp_show(hyp_obj_list[[28]])
hyp_show(hyp_obj_list[[29]])
hyp_show(hyp_obj_list[[30]])
hyp_show(hyp_obj_list[[31]])
hyp_show(hyp_obj_list[[32]])
toc()
```

GO MR terms GOMF_LIGAND_ACTIVATED_TRANSCRIPTION_FACTOR_ACTIVITY and GOMF_STEROID_HORMONE_RECEPTOR_ACTIVITY are over-represented in the category "Change in ligand binding," which is trivially explained by the fact that nuclear receptors are over-represented in this category.

Finally, check if any GO Cellular Compartment terms are over-represented in any of the switch functional categories:

```{r, include=T}
tic("Hypergeometric test of the overrepresentation of GO CC terms in each switch functional category of TFs")
hyp_obj_list = lapply(switch.categories,
                      function(sc) {
                        switch.category.tfs = isoform.switch.master.table.annot.with_functions %>%
                          filter(general_functional_consequence == sc) %>%
                          pull(humantfs_gene_name) %>%
                          unique()
                        sw.sc.overr = hypeR(signature = switch.category.tfs,
                                            genesets = go.cc.sets,
                                            test = "hypergeometric",
                                            background = switching.tfs,
                                            fdr = 0.05,
                                            quiet = FALSE)
                        sw.sc.overr
                      })

hyp_show(hyp_obj_list[[1]])
hyp_show(hyp_obj_list[[2]])
hyp_show(hyp_obj_list[[3]])
hyp_show(hyp_obj_list[[4]])
hyp_show(hyp_obj_list[[5]])
hyp_show(hyp_obj_list[[6]])
hyp_show(hyp_obj_list[[7]])
hyp_show(hyp_obj_list[[8]])
hyp_show(hyp_obj_list[[9]])
hyp_show(hyp_obj_list[[10]])
hyp_show(hyp_obj_list[[11]])
hyp_show(hyp_obj_list[[12]])
hyp_show(hyp_obj_list[[13]])
hyp_show(hyp_obj_list[[14]])
hyp_show(hyp_obj_list[[15]])
hyp_show(hyp_obj_list[[16]])
hyp_show(hyp_obj_list[[17]])
hyp_show(hyp_obj_list[[18]])
hyp_show(hyp_obj_list[[19]])
hyp_show(hyp_obj_list[[20]])
hyp_show(hyp_obj_list[[21]])
hyp_show(hyp_obj_list[[22]])
hyp_show(hyp_obj_list[[23]])
hyp_show(hyp_obj_list[[24]])
hyp_show(hyp_obj_list[[25]])
hyp_show(hyp_obj_list[[26]])
hyp_show(hyp_obj_list[[27]])
hyp_show(hyp_obj_list[[28]])
hyp_show(hyp_obj_list[[29]])
hyp_show(hyp_obj_list[[30]])
hyp_show(hyp_obj_list[[31]])
hyp_show(hyp_obj_list[[32]])
toc()
```

No over-representation of any GO CC terms in any switch functional category.

Plot the over-representation of TF families in switch functional categories. As a measure $\mu$ of over-representation, use the ratio of two proportions: (1) Proportion of TFs from a category among all TFs from a family; (2) Proportion of all TFs from a category to all TFs from the background.

First, plot all TF families in the category "Reverse regulation" to make sure that our measure of over-representation makes sense:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  dplyr::select(ensembl_gene_id,
                humantfs_gene_name,
                tf_family,
                general_functional_consequence) %>%
  distinct() %>%
  mutate(N = length(unique(ensembl_gene_id))) %>%
  group_by(tf_family) %>%
  do(mutate(., n = length(unique(.$ensembl_gene_id)))) %>%
  ungroup() %>%
  filter(general_functional_consequence == "Reverse regulation") %>%
  mutate(K = length(unique(ensembl_gene_id))) %>%
  add_count(tf_family) %>%
  dplyr::rename("k" = "nn") %>%
  mutate(mu = k / n * N / K) %>%
  mutate(mu0 = 1) %>%
  dplyr::select(tf_family, N, K, n, k, mu, mu0) %>%
  distinct()
```

C2H2 ZFs have $\mu > 1$ as expected (they are over-represented in the selected category). Other TF families with $\mu > 1$ are not over-represented in the category, according to the hypergeometric test above, probably because they have too few switching TFs. Nuclear receptors are also not over-represented in this category, and their observed count in the category is less than expected, which fits. 

Next, check the same metric for the "Change in ligand binding":

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  dplyr::select(ensembl_gene_id,
                humantfs_gene_name,
                tf_family,
                general_functional_consequence) %>%
  distinct() %>%
  mutate(N = length(unique(ensembl_gene_id))) %>%
  group_by(tf_family) %>%
  do(mutate(., n = length(unique(.$ensembl_gene_id)))) %>%
  ungroup() %>%
  filter(general_functional_consequence == "Change in ligand binding") %>%
  mutate(K = length(unique(ensembl_gene_id))) %>%
  add_count(tf_family) %>%
  dplyr::rename("k" = "nn") %>%
  mutate(mu = k / n * N / K) %>%
  mutate(mu0 = 1) %>%
  dplyr::select(tf_family, N, K, n, k, mu, mu0) %>%
  distinct()
```

Nuclear receptors have a very high $\mu$ in this category, which makes sense, as this family is over-represented in the selected category.

Hence, our measure of over-representation $\mu$ makes sense, and we will use it to make a volcano plot of over-represented TF families:

```{r, include=T}
overr.table = isoform.switch.master.table.annot.with_functions %>%
  dplyr::select(ensembl_gene_id,
                humantfs_gene_name,
                tf_family,
                general_functional_consequence) %>%
  distinct() %>%
  mutate(N = length(unique(ensembl_gene_id))) %>% # All switching TFs - the background
  group_by(tf_family) %>%
  do(mutate(., n = length(unique(.$ensembl_gene_id)))) %>% # Switching TFs within each family - the gene set
  ungroup() %>%
  group_by(general_functional_consequence) %>%
  do(mutate(., K = length(unique(.$ensembl_gene_id)))) %>% # Switching TFs within each category - the signature
  ungroup() %>%
  group_by(general_functional_consequence, tf_family) %>%
  do(mutate(., k = length(unique(.$ensembl_gene_id)))) %>% # The overlap between the gene set and the signature
  ungroup() %>%
  mutate(r = k / n * N / K) %>%
  dplyr::select(tf_family, general_functional_consequence, r) %>%
  distinct()

p = raw.pvalues.df %>%
  left_join(overr.table,
            by = c("tf.family" = "tf_family",
                   "functional.category" = "general_functional_consequence")) %>%
  mutate(r = ifelse(is.na(r), 0, r)) %>%
  mutate(overr = paste0(tf.family, " - ", functional.category)) %>%
  mutate(overr.label = ifelse(padj < 0.05, overr, "")) %>%
  mutate(sign = ifelse(padj < 0.05, "T", "F")) %>%
  mutate(sign = factor(sign, levels = c("T", "F"))) %>%
  ggplot(aes(x = log10(r), # r
             y = -log10(padj))) +
    geom_point(aes(color = sign), #overr),
               size = 2) + # 4
    geom_text_repel(aes(label = overr.label),
                    size = 2,
                    nudge_y = 0.2) + #,
                    # nudge_x = 0.5,
                    # nudge_y = 0.5) +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed", size = 0.5) + 
    scale_y_continuous(limits = c(0, 3.5)) + # , breaks = scales::pretty_breaks(7)
    #scale_x_continuous(limits = c(0, 300)) +
    theme_classic()

ggsave("data/results/tf_family_overr_in_switch_categories2.pdf", 
       p, 
       height = 3, 
       width = 3)
```

Investigate overlaps between functional switch categories:

```{r, include=T}
sw.category.list = list()

for (sc in switch.categories) {
  sc.tfs = isoform.switch.master.table.annot.with_functions %>%
      filter(general_functional_consequence == sc) %>%
      pull(humantfs_gene_name) %>%
      unique()
  sw.category.list[[sc]] =  sc.tfs
}

overlap.vector = numeric()
for (sc1 in names(sw.category.list)) {
  for (sc2 in names(sw.category.list)) {
    if (sc1 == sc2) {
      next
    }
    sc1.tfs = sw.category.list[[sc1]]
    sc2.tfs = sw.category.list[[sc2]]
    overlap.vector = c(overlap.vector, length(intersect(sc1.tfs, sc2.tfs)) / length(sc1.tfs))
  }
}

# Number of categories
cat("Number of signatures, n =", length(sw.category.list), "\n")
# Number of overlap proportions
cat("Number of overlap proportions, n * (n - 1) =", length(overlap.vector), "\n")
# Number of switching TF families that I test as gene sets
cat("Number of gene sets:", length(tf.family.sw.sets), "\n")
# Counts of different overlap proportions
as.data.frame(table(overlap.vector)) %>%
  arrange(desc(overlap.vector))
```

Do the categories with significant over-representations overlap?

```{r, include=T}
sw.category.list.sign = list()

for (sc in switch.categories) {
  if (!sc %in% c("Change in ligand binding",
                 "Change in protein binding",
                 "Change in transcription repression",
                 "Reverse regulation",
                 "TF / Unclear")) {
    next
  }
  sc.tfs = isoform.switch.master.table.annot.with_functions %>%
      filter(general_functional_consequence == sc) %>%
      pull(humantfs_gene_name) %>%
      unique()
  sw.category.list.sign[[sc]] =  sc.tfs
}

overlap.vector.sign = numeric()
for (sc1 in names(sw.category.list.sign)) {
  for (sc2 in names(sw.category.list.sign)) {
    if (sc1 == sc2) {
      next
    }
    sc1.tfs = sw.category.list.sign[[sc1]]
    sc2.tfs = sw.category.list.sign[[sc2]]
    overlap.vector.sign = c(overlap.vector.sign, length(intersect(sc1.tfs, sc2.tfs)) / length(sc1.tfs))
  }
}

as.data.frame(table(overlap.vector.sign)) %>%
  arrange(desc(overlap.vector.sign))
```

Check that tissues from the "Non-brain 1" and the "Other" clusters do not cluster separately from other tissues simply based on their expression profile. For this, make a PCA plot of tissue samples based on log2-transformed TPMs of the top 2000 protein-coding isoforms of both TFs and non-TF genes (Fig. S5F): 

```{r, include=T}
explain.variation = function(pca) {
  pca.summary = summary(pca)
  pca.pc1.annotation = paste("PC1 (", toString(round(pca.summary$importance[2] * 100, digits=1)), "%)", sep="")
  pca.pc2.annotation = paste("PC2 (", toString(round(pca.summary$importance[5] * 100, digits=1)), "%)", sep="")
  return(c(pca.pc1.annotation, pca.pc2.annotation))
}

gtex.annot.subset.filtered = read.delim("data/results/gtex8_processed/gtex_annot_subset_filtered.tsv")

gtex.annot.subset.filtered.target = gtex.annot.subset.filtered %>%
  mutate(SMTSD = ifelse(stringr::str_detect(SMTSD, fixed("Brain -")) & (SMTSD != "Brain - Cerebellum"), 
                        "Brain", 
                        ifelse(!SMTSD %in% c("Liver", 
                                             "Pancreas",
                                             "Muscle - Skeletal",
                                             "Heart - Atrial Appendage",
                                             "Heart - Left Ventricle",
                                             "Brain - Cerebellum",
                                             "Pituitary",
                                             "Prostate"),
                        "Other non-brain",
                        SMTSD)))

selected.table.log2 = read.delim("data/results/gtex8_processed/tfs_nontfs_selected_table_tpm_log2_final.tsv",
                                 header = T,
                                 sep = "\t",
                                 stringsAsFactors = F)

target.selected.table.log2 = selected.table.log2[, names(selected.table.log2) %in% gtex.annot.subset.filtered.target$SAMPID]

target.selected.table.log2.top = target.selected.table.log2 %>%
  rowwise() %>%
  mutate(max.expression = max(c_across(1:(ncol(target.selected.table.log2))))) %>%
  arrange(desc(max.expression)) %>%
  dplyr::select(-max.expression) %>%
  head(2000)

pca.list.tpm.log2  = prcomp(t(target.selected.table.log2.top))
pcat = as.data.frame(pca.list.tpm.log2$x[, 1:2])
pcat = pcat %>% 
  tibble::rownames_to_column(var = "SAMPID") %>%
  left_join(gtex.annot.subset.filtered.target,
            by = c("SAMPID" = "SAMPID"))

pc.annotation = explain.variation(pca.list.tpm.log2)
pc1.annotation = pc.annotation[1]
pc2.annotation = pc.annotation[2]

p = ggplot(pcat[pcat$SMTSD %in% c("Brain", "Other non-brain"), ],
           aes(x = PC1, y = PC2, color = SMTSD)) + 
  geom_point() + 
  geom_point(data = pcat[!pcat$SMTSD %in% c("Brain", "Other non-brain"), ]) +
  scale_colour_manual(values = c("pink",        # Brain
                                 "hotpink4",    # Brain - Cerebellum
                                 "darkgreen",   # Heart - Atrial Appendage
                                 "green",       # Heart - Left Ventricle
                                 "red",         # Liver
                                 "greenyellow", # Muscle - Skeletal
                                 "grey90",      # Other non-brain
                                 "blue",        # Pancreas
                                 "purple",      # Pituitary
                                 "orange")) +   # Prostate
  labs(x = pc1.annotation, y = pc2.annotation) +
  theme_classic()

ggsave("data/results/nonbrain_samples_pca2.pdf", 
       p, 
       height = 7, 
       width = 10)
```

So, skeletal muscles and heart subtissues cluster together and the pancreas and liver also cluster together, which makes sense, but these two subgroups are separated by PC2, which also makes sense, because muscle tissues are quite different physiologically from the liver and pancreas. Consequently, the separate clustering of the Non-brain 1 tissues should be caused by the switch patterns, but not expression profiles, of TFs in these tissues. The same is true about the Other tissue, as all the three of them are mostly separated by PC1. Of note, cerebellum maps to the brain group of samples, which perfectly makes sense, while the pituitary gland is mostly separated from the muscle tissues by PC2 and the prostate is mostly separated from the muscle tissues, the pituitary gland, the liver, the pancreas and the brain by PC1. This also makes sense.

## Interesting examples of switching TFs

In each cluster, try to find TFs whose switches are characterised functionally in literature and are observed in our data. This means that, in addition to some functional characterisation, such switches from literature should be:

1) Between major isoforms in different tissues;

2) Between isoforms of the same types as we observe;

3) At least partially, between the same tissues that we observe.

The presence of such TFs would support the biological relevance of major isoform switches in the clusters.

Also, in each cluster try to find TFs whose observed switches are not characterised. The presence of such TFs would motivate studying major isoform switches in the clusters.

First, define heatmap plotting functions to illustrate found switches:

```{r, include=T}
define_rank_groups = function(tf.df.heatmap, isoform.rank.heatmap) {

  isoform.rank.heatmap.vertical = as.data.frame(t(isoform.rank.heatmap)) %>%
    arrange(across(everything())) %>%
    rownames_to_column(var = "tissues")

  names(isoform.rank.heatmap.vertical) = c("tissues", rownames(isoform.rank.heatmap))

  rank.df = isoform.rank.heatmap.vertical

  for (i in 1:(nrow(rank.df))) {
    rank.df[i, names(rank.df) != "tissues"] = ifelse(rank.df[i, names(rank.df) != "tissues"] == 1,
                                                     1,
                                                     2)
  }

  unique.rank.lists = rank.df %>%
    dplyr::select(-tissues) %>%
    distinct()

  rank.groups = rep(0, ncol(isoform.rank.heatmap))
  names(rank.groups) = names(isoform.rank.heatmap)

  for (rank.group.id in 1:(nrow(unique.rank.lists))) {
    rank.list = unique.rank.lists[rank.group.id, ]
    rank.list.tissues = rank.df %>%
      filter(apply(rank.df,
                   1,
                   function(r) {
                     Reduce(function(x, y) x & y,
                            as.numeric(r[names(r) != "tissues"]) == rank.list)
                   })) %>%
      pull(tissues)

    rank.groups[names(rank.groups) %in% rank.list.tissues] = rank.group.id
  }

  return(rank.groups)
}

find_switching_tissues = function(sw.isoforms) {
  sw.tissues = c("ensembl_transcript_id")
  
  if (length(sw.isoforms) == 0) {
    return(sw.tissues)
  }
  
  sw.tissues = c(sw.tissues,
                 sort(unique(c(unlist(stringr::str_split(isoform.switch.master.table.annot.with_functions %>%
                                                           filter(ensembl_transcript_id1 %in% sw.isoforms) %>%
                                                           pull(tissue_list1),
                                                         fixed(","))),
                               unlist(stringr::str_split(isoform.switch.master.table.annot.with_functions %>%
                                                           filter(ensembl_transcript_id2 %in% sw.isoforms) %>%
                                                           pull(tissue_list2),
                                                         fixed(",")))))))
  
  return(sw.tissues)
}

generate_expression_heatmap = function(tf.df,
                                       isoform.rank.df) {

  tf.title = paste(tf.df %>% pull(humantfs_gene_name) %>% unique(),
                   paste0("(", tf.df %>% pull(ensembl_gene_id) %>% unique(), ")."),
                   tf.df %>% pull(tf_family) %>% unique(),
                   "family",
                   sep = " ")

  isoform.rank.heatmap = isoform.rank.df %>%
    dplyr::select(-ensembl_gene_id,
                  -humantfs_gene_name,
                  -tf_family,
                  -n_dbd) %>%
    filter(ensembl_transcript_id %in% switching.isoforms) %>%
    dplyr::select(., all_of(find_switching_tissues(.$ensembl_transcript_id))) %>%
    tibble::column_to_rownames(var = "ensembl_transcript_id")
  
  if (nrow(isoform.rank.heatmap) == 0) {
    stop("No switching isoforms in the TF\n")
  }

  tf.df = tf.df %>%
    mutate(dbd_present = ifelse(n_dbd > 0, T, F)) %>%
    filter(ensembl_transcript_id %in% switching.isoforms) %>%
    dplyr::select(all_of(c("ensembl_gene_id",
                           "humantfs_gene_name",
                           "ensembl_transcript_id",
                           "tf_family",
                           "n_dbd",
                           "dbd_present",
                           names(isoform.rank.heatmap))))

  tf.ha = HeatmapAnnotation(n_dbd = as.character(tf.df %>% pull(n_dbd)),
                            dbd_present = as.character(tf.df %>% pull(dbd_present)),
                            which = "row")

  tf.df.heatmap = tf.df %>%
    dplyr::select(-ensembl_gene_id,
                  -humantfs_gene_name,
                  -tf_family,
                  -n_dbd,
                  -dbd_present) %>%
    tibble::column_to_rownames(var = "ensembl_transcript_id")

  tf.df.heatmap = as.data.frame(scale(tf.df.heatmap))

  Heatmap(tf.df.heatmap,
          column_title = tf.title,
          right_annotation = tf.ha,
          column_split = define_rank_groups(tf.df.heatmap,
                                            isoform.rank.heatmap))
}
```

Cluster 1:

Cluster 2:

Uncharacterised TF:

- ZNF655:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF655"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

The full DBD+ isoform of ZNF655 binds CDK4 and may be involved in the control of the cell cycle [Houlard et al., 2004](https://www.nature.com/articles/1208043). This TF belongs to the cluster 2. It has a complicated pattern of switches, but most interestingly, it swaps one DBD- major isoform for another DBD- major isoform between different brain subtissues. Both isoforms are domainless and hence differences between them should be in their UTRs. Consequently, the two isoforms may differ in their stability or translational efficiency. In our analysis, we need to check if these domainless isoforms still have intrinsically disordered regions (IDRs) which would hint at their functional relevance. But ultimately, it would be very interesting to study them experimentally, because, although there was a publication [Houlard et al., 2004](https://www.nature.com/articles/1208043) about isoforms of this TF and their possible functions, it mentions only one DBD- isoform and does not prove its possible function.

Cluster 3:

Uncharacterised TF:

- TBX3:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "TBX3"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

As a T-box TF, TBX3 takes part in different developmental processes and is expressed in many developing tissues [Khan et al., 2020](https://www.sciencedirect.com/science/article/pii/S0378111919308820). It is also widely expressed across different adult tissues [Fan et al., 2004](https://cancerres.aacrjournals.org/content/64/15/5132.full). Two main isoforms of this TF are known. One of them, TBX3+2a, has a 20-aa insertion in its DBD (T-box), while the other, TBX3 (or TBX3-2a) lacks this insertion. To the best of our knowledge, a switch between these isoforms as major ones in different tissues is not known. However, in our data, we observe a switch between TBX3+2a in the liver and spleen and TBX3-2a in other non-brain tissues plus cortex. This switch may have some functional consequences, as [Fan et al., 2004](https://cancerres.aacrjournals.org/content/64/15/5132.full) demonstrated in the MEF cell line that while overexpression of TBX3+2a accelerated cell senescence, overexpression of TBX3-2a had the opposite effect and immortalised the cell line.

Cluster 4:

Cluster 5:

Well-studied TF with an uncharacterised DBD- isoform:

- CREM:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "CREM"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

This TF is involved in spermatogenesis and spermatid maturation. Its transcription regulation activity depends on binding of cAMP. Different DBD+ isoforms of this TF can be activators or repressors. It also plays a role in the regulation of circadian clock (GeneCards). CREM is a well-studied TF, but, to the best of our knowledge, all its functional isoforms studied so far are DBD+, while in our data CREM's major isoform in the skeletal muscles and a part of the digestive system is DBD-. So, it would be very interesting to experimentally study this major DBD- isoform.

Characterised TF:

- MEF2C:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "MEF2C"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

MEF2C activates many muscle-specific genes, takes part in myogenesis and heart development. It also plays a key role in neurodevelopment and hippocamp-dependent memory formation (GeneCards). In our data, MEF2C expresses one major DBD+ isoform in skeletal muscles and another major DBD+ isoform in all other tissues. These isoforms differ by the sequence of a protein-binding domain and hence could bind different partner proteins (or bind the same partners with different affinities). [McDermott et al., 1993](https://journals.asm.org/doi/epdf/10.1128/mcb.13.4.2564-2577.1993) also demonstrated that human MEF2C switched two DBD+ isoforms between the brain and the skeletal muscles, and the isoforms were different by the transcription activation domain sequence. Although they did not find any differences in transcription activation between the two isoforms, the authors do not exclude the possibility that the differences still exist in the native context in which the isoforms are expressed. Overall, the two switched DBD+ isoforms in our data and the two switched DBD+ studied by [McDermott et al., 1993](https://journals.asm.org/doi/epdf/10.1128/mcb.13.4.2564-2577.1993) could be the same. 

Well-studied TF with a studied switch:

- IRF3:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "IRF3"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

This interferon regulatory TF takes part in activation of interferons alpha and beta and is important for the innate immune response against viral infections (GeneCards). The DBD- isoform of IRF3 acts as a dominant negative regulator of the DBD+ isoform in a dosage-dependent manner, and the highest ratio of the DBD- isoform to the DBD+ isoform is found in the brain. The DBD+ isoform is broadly expressed across tissues and activates the interferon (IFN) beta promoter upon viral infection, while the DBD- isoform blocks this activation [Karpova et al., 2001](https://journals.asm.org/doi/10.1128/MCB.21.13.4169-4176.2001). Hence, the pathway of IFN-beta activation via IRF3 should be blocked in the adult brain.

Uncharacterised TF:

- ZNF571:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF571"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

According to GeneCards, this TF is not well studied. In our data, ZNF571 expresses a DBD+ isoform (KRAB zinc finger TF) outside of the brain and switches it for a DBD- isoform (KRAB-only) in the brain. To the best of our knowledge, this switch has not been functionally characterised. Consequently, it would be very interesting to experimentally study its functional relevance. [If ZNF571 and any other zinc finger TFs that switch in the same way are of a type that should control the transcription of transposons, then the loss of the DNA binding by these TFs in the brain is especially intriguing: Does it allow expression of some transposons in the brain, which, in general, is a known phenomenon?]

Studied TF with an unknown switch:

- TFEB:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "TFEB"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

TFEB is a "transcription factor that acts as a master regulator of lysosomal biogenesis, autophagy, lysosomal exocytosis, lipid catabolism, energy metabolism and immune response (PubMed:21617040, PubMed:22576015, PubMed:22343943, PubMed:22692423, PubMed:30120233, PubMed:31672913)." (UniProtKB summary from GeneCards). [Kuiper et al., 2004](https://academic.oup.com/nar/article/32/8/2315/2904553) described several isoforms of this TF with different 5'-UTRs. Additionally, [Pan et al., 2017](https://www.sciencedirect.com/science/article/pii/S0022282817303218) demonstrated that TFEB switches its protein isoforms in proteinopathic myocardium, although did not show functional significance of this switch. In our data, TFEB switches its major DBD+ isoforms between (1) the majority of non-brain tissues plus cerebellum and (2) the majority of the brain subtissues plus pituitary and some non-brain tissues. To the best of our knowledge, this switch has not been characterised in literature. However, it would be interesting to try to characterise its possible functional consequences experimentally, because the two major DBD+ isoforms differ in the presence / absence of a protein-binding domain.

Cluster 6:

Well-studied TF with an unknown switch:

- WT1:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "WT1"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

For a long time, this TF has been known as a regulator of sex determination and urogenital differentiation in both sexes ([Mundlos et al., 1993](https://journals.biologists.com/dev/article/119/4/1329/38141/Nuclear-localization-of-the-protein-encoded-by-the); [Shimamura et al., 1997](https://clincancerres.aacrjournals.org/content/3/12/2571); [Schedl & Hastie, 1998](https://www.sciencedirect.com/science/article/pii/S0303720798000318)) and for equally long time it has been known to have several isoforms ([Bickmore et al., 1992](https://www.science.org/doi/abs/10.1126/science.1321494); [Ullmark et al., 2017](https://haematologica.org/article/view/7973); [Schedl & Hastie, 1998](https://www.sciencedirect.com/science/article/pii/S0303720798000318)). Mostly, people studied differences in DNA binding between WT1 isoforms with and without a KTS tripeptide between two C2H2 zinc fingers, but less attention was paid to possible functional differences between isoforms with and without a 17-aa protein-binding region. It may be so, because [Natoli et al., 2002](https://journals.asm.org/doi/epub/10.1128/MCB.22.12.4433-4438.2002) demonstrated that the lack of the isoform containing this region did not preclude normal development, function and histological appearance of testes and ovaries. However, in our data WT1 swaps a major isoform containing this region in testis for a major isoform lacking this region in vagina. Consequently, it still would be interesting to double-check possible functional differences between these isoforms in human cell lines or disease association studies of mutations in corresponding splice sites, because, in principle, the role of this domain in developing and in adult tissues could be different.

Cluster 7:

TF with unknown switch:

- PRDM2:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "PRDM2"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

This is a histone methyltransferase that specifically methylates H3K9 and also can function as a TF (GeneCards). In our data, PRDM2 expresses a DBD+ (DBD-only) major isoform outside of the brain and switches it for a DBD- major isoform (catalytic domain-only) in the brain. Although several studies suggested that PRDM2 may play a role in the brain ([Geli et al., 2010](https://www.spandidos-publications.com/ijo/37/5/1323); [Zhang et al., 2015](https://bmccancer.biomedcentral.com/articles/10.1186/s12885-015-2023-1); [Gao et al., 2015](https://link.springer.com/article/10.1186/s12885-015-1267-0); [Heilig et al., 2016](https://onlinelibrary.wiley.com/doi/full/10.1111/gbb.12344); [Sorrentino et al., 2018](https://www.sciencedirect.com/science/article/pii/S1874939918300713#f0010)), this switch has not been functionally characterised, to the best of our knowledge. Hence, it would be very interesting to experimentally study its functional relevance.

## Interesting examples of switches from particular parts of the heatmap

1) The whole cluster 1 (TF gene products are not TFs in the brain and some other tissues, but turn into TFs in select tissues):

TFs from cluster 1:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 1) %>%
  pull(humantfs_gene_name) %>%
  unique()
```

Summarise their switches:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 1) %>%
  dplyr::select(ensembl_gene_id, humantfs_gene_name, tf_family, general_functional_consequence) %>%
  arrange(general_functional_consequence) %>%
  distinct()
```

Let us plot switch heatmaps for all these TF:

- FOXJ3:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "FOXJ3"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "FOXJ3")
```

So, the only switch is "TF / Domainless protein".

FOXJ3 (https://www.uniprot.org/uniprot/Q9UPW0) is a Forkhead TF which, according to UniProt, can be related by similarity to the regulation of the muscle identity and regeneration, and also to spermatogenesis. Indeed, in the muscle tissues and testis, FOXJ3 expresses a DBD+ major isoform. In contrast, in the brain and some other non-brain tissues, it expresses a DBD- major isoform and hence cannot be a TF. As this DBD- isoform is actually domainless, this switch could "turn off" the TF in the brain and some non-brain tissues where it is not necessary but "turn it on" in other non-brain tissues, including muscles and testis, where its TF function may be required.

FOXJ3 is in my compendium of TFs with studied alternative isoforms, but I did not find any experimental characterisation of this switch in literature. [Cabezas-Wallscheid et al., 2014](https://www.sciencedirect.com/science/article/pii/S1934590914003014) reported that FOXJ3 differentially uses its first exon between two undifferentiated cell types, HSCs (hematopoetic stem cells) and MPP (multipotent progenitor) cells. However, the authors did not characterise functional consequences of this switch, and the DBD is not in the first exon anyway (http://jan2020.archive.ensembl.org/Homo_sapiens/Transcript/ProteinSummary?db=core;g=ENSG00000198815;r=1:42176539-42335877;t=ENST00000372573).

- NFIB:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "NFIB"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "NFIB")
```

So the only switch is "TF / Unclear": In the brain and some non-brain tissues, the the major isoform does not have DBDs and a possible dimerization domain and retains only a domain whose function is not clear. Hence, if judged by its major isoforms, NFIB is not a TF (but a protein with an unclear function) in the brain and some non-brain tissues, while it is a TF in other non-brain tissues.

NFIB (https://www.uniprot.org/uniprot/O00712) is a "transcriptional activator of GFAP, essential for proper brain development." Hence, it could make sense to "switch off" the TF function of this protein in the adult brain.

This TF is in my compendium, although the characterised switch is different. Specifically, [Chen et al., 2014](https://www.science.org/doi/full/10.1126/science.1251033) showed that NFIB expresses a major DBD- isoform in megakaryocytes and that this isoform cannot dimerizse with NFIC. They also demonstrated that both NFIC and the DBD- isoform of NFIB play a key role in megakaryopoesis. However, in Fig. 4B, the authors also claim that the region outside of the DBD/dimerization domain is mostly a transcription regulation domain. Hence, according to our functional switch classification, the NFIB switch that we see in adult human tissues could reverse the regulatory function of this TF.

- NR2F2:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "NR2F2"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "NR2F2")
```

So the only switch is "TF / Ligand sequestering factor": While the protein is a ligand-activated TF in some non-brain tissues, it becomes a non-TF in the brain and other non-brain tissues where it probably sequesters the ligand from the chromatin.

NR2F2 (https://www.uniprot.org/uniprot/P24468) may be required "to establish ovary identity during early gonad development." 

NR2F2 is in my compendium, but I did not find any experimental characterisation of its possible switches. The only relevant article I found was [Rosa & Brivanlou, 2011](https://www.embopress.org/doi/full/10.1038/emboj.2010.319), where the authors demonstrated that NR2F2 expressed several isoforms during hESC differentiation and that the co-expression of shorter (DBD-) isoforms with the full one enhanced the activity of the latter. However, they did not investigate any switches of this TF's isoforms.

- MAZ:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "MAZ"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "MAZ")
```

Both switches are "TF / Domainless protein". Hence, in two different very small subsets of non-brain tissues it is a TF (probably, with different DNA-binding specificities, as the two DBD+ isoforms have different numbers of C2H2 zinc fingers) but turns into a domainless (hence, a non-TF) protein in the brain and some other non-brain tissues.

MAZ (https://www.uniprot.org/uniprot/P56270) is a "transcriptional regulator, potentially with dual roles in transcription initiation and termination." UniProt describes three isoforms of this TF, but all the three are DBD+. 

This TF is not in my compendium, and I did not find any experimental characterisation of its possible major isoform switches in literature.

- ZNF197:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF197"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF197")
```

So, the only switch is "Reverse regulation": In the subcutaneous adipose tissue and ovary, it expresses the full (DBD+) major isoform, and as this isoform has a KRAB domain, it probably serves as a transcription repressor; in contrast, in the brain and other non-brain tissues, this gene expresses a DBD- isoform that still contains the KRAB and the SCAN domains and hence can sequester its binding partners from the chromatin. 

ZNF197 (https://www.uniprot.org/uniprot/O14709) is not well characterised, according to UniProt.

[Li et al., 2003](https://www.embopress.org/doi/full/10.1093/emboj/cdg173) demonstrated that ZNF197 produces the two isoforms that we can see in the switch above, and that the DBD- isoform, contrary to our prediction, also can function as a transcription repressor, as it retains the KRAB domain and can be recruited by another protein complex to HIF-1A to repress it. Additionally, the authors showed that the DBD- isoforms of ZNF197 can homodimerize via the SCAN domain. However, I did not find any experimental characterisation of the switch between the two isoforms that we observe. If both the DBD+ and the DBD- isoforms are repressors, then they can have different sets of targets, as the DBD- isoform needs to be recruited to chromatin by other proteins rather than by the DNA sequence. Consequently, if judged by its major isoforms, ZNF197 may have transcription repression activities that are somehow different in (a) the brain and some non-brain tissues and (b) other non-brain tissues.

- ZNF226:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF226"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF226")
```

So, the only switch is "Reverse regulation": The DBD+ isoform expressed in some non-brain tissues and cerebellum also has a KRAB domain (and hence could bind DNA specifically and repress transcription), while the DBD- isoform is expressed in the rest of the brain and some other non-brain tissues, retains the KRAB domain and hence could sequester co-repressors from the chromatin. Consequently, we predict that in the brain and some non-brain tissues (where ZNF226 expresses a DBD- isoform), it could be an activator, while repressing transcription in other non-brain tissues and cerebellum.

ZNF226 (https://www.uniprot.org/uniprot/Q9NYT6) is not really characterised, according to UniProt.

ZNF226 is not in my compendium, and I could not find any experimental characterisation of its isoforms or their switches.

2) Cluster 7:

Select only the TFs that switch two DBD+ major isoforms between at least three brain subtissues and other tissues:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 7) %>%
  filter(((isoform_rank1 == 1) & (stringr::str_detect(tissue_list1, fixed("Brain"))) & (isoform_rank2 == 2)) | 
         ((isoform_rank2 == 1) & (stringr::str_detect(tissue_list2, fixed("Brain"))) & (isoform_rank1 == 2))) %>%
  filter((stringr::str_count(tissue_list1, fixed("Brain")) >= 3) | (stringr::str_count(tissue_list2, "Brain") >= 3)) %>%
  arrange(humantfs_gene_name)
```

Summarise this table by the functional consequence:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 7) %>%
  filter(((isoform_rank1 == 1) & (stringr::str_detect(tissue_list1, fixed("Brain"))) & (isoform_rank2 == 2)) | 
         ((isoform_rank2 == 1) & (stringr::str_detect(tissue_list2, fixed("Brain"))) & (isoform_rank1 == 2))) %>%
  filter((stringr::str_count(tissue_list1, fixed("Brain")) >= 3) | (stringr::str_count(tissue_list2, "Brain") >= 3)) %>%
  arrange(general_functional_consequence) %>%
  dplyr::select(ensembl_gene_id, humantfs_gene_name, tf_family, general_functional_consequence)
```

Mark these TFs in the switch heatmap:

```{r, include=T}
cl7.tf.ids = isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 7) %>%
  filter(((isoform_rank1 == 1) & (stringr::str_detect(tissue_list1, fixed("Brain"))) & (isoform_rank2 == 2)) | 
         ((isoform_rank2 == 1) & (stringr::str_detect(tissue_list2, fixed("Brain"))) & (isoform_rank1 == 2))) %>%
  filter((stringr::str_count(tissue_list1, fixed("Brain")) >= 3) | (stringr::str_count(tissue_list2, "Brain") >= 3)) %>%
  arrange(humantfs_gene_name) %>%
  pull(ensembl_gene_id)

ha = rowAnnotation(target.tfs = as.character(rownames(heatmap.df) %in% cl7.tf.ids),
                   col = list(target.tfs = c("TRUE" = "red", "FALSE" = "blue")))

h = Heatmap(heatmap.df, 
            name = "Isoform ranks", 
            clustering_distance_columns = function(x, y) hamming_dist(x, y),
            clustering_distance_rows = function(x, y) hamming_dist(x, y),
            col = cluster.colors,
            row_labels = rep("", nrow(heatmap.df)),
            row_split = 7, # 11
            column_split = 5) + # 4
  ha

pdf(file = "data/results/clustered_switch_heatmap_2_tissues5_cluster7_part_marked.pdf",
    width = 10,
    height = 15)
h
dev.off()
```

- ZNF274:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF274"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF274")
```

So, the only switch is a "Change in transcription repression": The two major DBD+ isoforms differ by one transcription repression domain (KRAB). Hence, in the brain and some non-brain tissues, ZNF274 may have one mode (for example, specificity or strength) of transcription repression, while in other non-brain tissues it may have another mode of transcription repression.

ZNF274 (https://www.uniprot.org/uniprot/Q96GC6) is a repressor of C2H2 ZF transcription factors. 

ZNF274 is not in my compendium, and I did not find any experimental characterisation of its possible major isoform switches.

- ZNF12:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF12"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF12")
```

So, the only switch is a "Change in transcription repression": The two major DBD+ isoforms differ by one transcription repression domain (KRAB). Hence, in the brain and some non-brain tissues, ZNF274 may have one mode (for example, specificity or strength) of transcription repression, while in other non-brain tissues it may have another mode of transcription repression.

ZNF12 (https://www.uniprot.org/uniprot/P17014) is a "transcriptional repressor which suppresses activation protein 1 (AP-1)- and serum response element (SRE)-mediated transcriptional activity."

ZNF12 is not in my compendium, and I could not find any experimental characterisation of its possible major isoform switches.

- ZNF548:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF548"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF548")
```

So, the only switch is a "Change outside of domains," and the switch happens between (a) testis and (b) brain and some non-brain tissues. Hence, the difference between the isoforms could be in some disordered regions and still be functional.

ZNF548 (https://www.uniprot.org/uniprot/Q8NEK5) is a not a well studied TF.

ZNF548 is not in my compendium, and I did not find any experimental characterisation of its possible major isoform switches.

- ZNF746:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF746"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF746")
```

So, the only switch is a "Change outside of domains," and it happens between (a) adipose tissue and (b) brain and a few non-brain tissues. Hence, the difference between the isoforms could be in some disordered regions and still be functional. 

ZNF746 (https://www.uniprot.org/uniprot/Q6NUN9) "repress transcription of PGC-1-alpha (PPARGC1A), thereby playing a role in regulation of neuron death." Hence, the switch between the brain and something else could make sense. 

ZNF746 is in my compendium, but I could not find any experimental characterisation of a switch that we observe.

- ZNF76:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF76"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF76")
```

So, the only switch is a "Change outside of domains," and it happens between (a) liver, minor salivary gland and pancreas and (b) brain and select non-brain tissues. Hence, the difference between the isoforms could be in some disordered regions and still be functional.

ZNF76 (https://www.uniprot.org/uniprot/P36508) is not a well-studied TF.

ZNF76 is in my compendium, but I could not find any experimental characterisation of a switch that we observe.

- ZNF195:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF195"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF195")
```

So, the only switch is a "Change outside of domains," which happens between (a) pituitary gland and skin and (b) brain and some other tissues. Hence, the difference between the isoforms could be in some disordered regions and still be functional.

ZNF195 (https://www.uniprot.org/uniprot/O14628) is not a well-studied TF.

ZNF195 is not in my compendium, and I could not find any experimental characterisation of its possible major isoform switches.

- ZNF721:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF721"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF721")
```

So, the only switch is a "Reverse regulation and change in DNA binding," which happens between (a) adipose tissue, tibial artery, skin and stomach and (b) brain and select non-brain tissues. Hence, in the group of tissues (a), the major isoform of ZNF721 has a C2H2 ZF and a KRAB domain and consequently could function as a transcription repressor, while in the group of tissues (b) it has only C2H2 ZFs and no KRAB domain and hence may not be able to repress transcription. Additionally, as the number of C2H2 ZFs differs greatly between the two isoforms, they probably have different DNA binding specificities (if the non-brain major isoform with just one C2H2 ZF binds DNA at all).

ZNF721 (https://www.uniprot.org/uniprot/Q8TF20) is not a well-studied TF.

ZNF721 is not in my compendium, and I could not find any experimental characterisation of its possible major isoform switches.

Additionally, let us consider TFs from this cluster that express a DBD- major isoform in the brain but switch it for a DBD+ major isoform outside of the brain:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 7) %>%
  filter(((isoform_rank1 == -2) & (stringr::str_detect(tissue_list1, fixed("Brain"))) & (isoform_rank2 == 1)) | 
         ((isoform_rank2 == -2) & (stringr::str_detect(tissue_list2, fixed("Brain"))) & (isoform_rank1 == 1))) %>%
  filter((stringr::str_count(tissue_list1, fixed("Brain")) >= 3) | (stringr::str_count(tissue_list2, "Brain") >= 3)) %>%
  arrange(humantfs_gene_name)
```

We already described the switch in PRDM2, and now let us consider the switch in SALL2, as it happens also almost exclusively between the brain and non-brain tissues.

Thus, SALL2 expresses a DBD- major isoform in some brain subtissues and also in three non-brain tissues (including the pituitary gland), while it produces a DBD+ major isoform in many non-brain tissues plus cerebellum. We predict that the functional consequence of this switch is "TF / Domainless protein," where SALL2 is a domainless protein in the brain subtissues.

SALL2 (https://www.uniprot.org/uniprot/Q9Y467) is a "probable transcription factor that plays a role in eye development before, during, and after optic fissure closure." Additionally, according to [Hermosilla et al., 2017](https://academic.oup.com/carcin/article/38/7/680/3738533), "SALL2 plays a role in neurogenesis, neuronal differentiation and eye development.

SALL2 is in my compendium, but I did not find any experimental characterisation of the switch that we observe.

3) Cluster 5:

a) DBD- in the brain vs. DBD+ outside the brain:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 5) %>%
  filter(((isoform_rank1 == -2) & (stringr::str_count(tissue_list1, fixed("Brain")) >= 5) & (isoform_rank2 == 1)) | 
         ((isoform_rank2 == -2) & (stringr::str_count(tissue_list2, fixed("Brain")) >= 5) & (isoform_rank1 == 1))) %>%
  arrange(humantfs_gene_name)
```

Summarise these switches:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 5) %>%
  filter(((isoform_rank1 == -2) & (stringr::str_count(tissue_list1, fixed("Brain")) >= 5) & (isoform_rank2 == 1)) | 
         ((isoform_rank2 == -2) & (stringr::str_count(tissue_list2, fixed("Brain")) >= 5) & (isoform_rank1 == 1))) %>%
  arrange(humantfs_gene_name) %>%
  dplyr::select(ensembl_gene_id, humantfs_gene_name, tf_family, general_functional_consequence)
```

Mark these TFs in the switch heatmap:

```{r, include=T}
cl5.mp.tf.ids = isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 5) %>%
  filter(((isoform_rank1 == -2) & (stringr::str_count(tissue_list1, fixed("Brain")) >= 5) & (isoform_rank2 == 1)) | 
         ((isoform_rank2 == -2) & (stringr::str_count(tissue_list2, fixed("Brain")) >= 5) & (isoform_rank1 == 1))) %>%
  arrange(humantfs_gene_name) %>%
  pull(ensembl_gene_id)

ha = rowAnnotation(target.tfs = as.character(rownames(heatmap.df) %in% cl5.mp.tf.ids),
                   col = list(target.tfs = c("TRUE" = "red", "FALSE" = "blue")))

h = Heatmap(heatmap.df, 
            name = "Isoform ranks", 
            clustering_distance_columns = function(x, y) hamming_dist(x, y),
            clustering_distance_rows = function(x, y) hamming_dist(x, y),
            col = cluster.colors,
            row_labels = rep("", nrow(heatmap.df)),
            row_split = 7, # 11
            column_split = 5) + # 4
  ha

pdf(file = "data/results/clustered_switch_heatmap_2_tissues5_cluster5_pm_marked.pdf",
    width = 10,
    height = 15)
h
dev.off()
```

We already described switches in IRF3 and ZNF571. Hence, let us consider the other 3 TFs:

- RARA:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "RARA"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "RARA")
```

So, the DBD+/- switch is a "TF / Ligand sequestering factor," it happens between the brain (plus skeletal muscles) and many non-brain tissues, and in the brain the major isoform loses the DBD but retains a ligand-binding domain, in this way, presumably, becoming a ligand-sequestering factor.

RARA (https://www.uniprot.org/uniprot/P10276) is a retinoic acid receptor, and such receptors usually bind as heterodimers. 

RARA is in my compendium, but I could not find any experimental characterisation of the DBD+/- switch that we observe.

- THAP4:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "THAP4"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "THAP4")
```

So, the only switch is "TF / Ligand sequestering factor," it happens between (a) the brain and select non-brain tissues and (b) other non-brain tissues plus cerebellum, and in the brain THAP4 loses the DBD while retaining the ligand-binding domain, in this way, presumably, becoming a ligand-sequestering factor.

THAP4 (https://www.uniprot.org/uniprot/Q8WY91) is a heme-binding TF. 

THAP4 is not in my compendium, and I could not find and experimental characterisation of the switch that we observe.

- ZNF470:

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF470"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF470")
```

So, the only switch is a "Reverse regulation," it happens between the brain and many non-brain tissues, and in the brain the major isoform retains only the KRAB domain, hence, possibly, sequestering co-repressors from the chromatin and in this way de-repressing ZNF470 target genes.

ZNF470 (https://www.uniprot.org/uniprot/Q6ECI4) is not a well-studied TF.

ZNF470 is not in my compendium, and I could not find any experimental characterisation of the switch that we observe.

b) DBD+/+ switches between the brain and non-bran tissues:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 5) %>%
  filter(((isoform_rank1 == 2) & (stringr::str_count(tissue_list1, fixed("Brain")) >= 6) & (isoform_rank2 == 1)) | 
         ((isoform_rank2 == 2) & (stringr::str_count(tissue_list2, fixed("Brain")) >= 6) & (isoform_rank1 == 1))) %>%
  arrange(humantfs_gene_name)
```

Summarise these switches:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 5) %>%
  filter(((isoform_rank1 == 2) & (stringr::str_count(tissue_list1, fixed("Brain")) >= 6) & (isoform_rank2 == 1)) | 
         ((isoform_rank2 == 2) & (stringr::str_count(tissue_list2, fixed("Brain")) >= 6) & (isoform_rank1 == 1))) %>%
  dplyr::select(ensembl_gene_id, humantfs_gene_name, tf_family, general_functional_consequence) %>%
    arrange(general_functional_consequence)
```

Mark these TFs in the switch heatmap:

```{r, include=T}
cl5.pp.ids = isoform.switch.master.table.annot.with_functions %>%
  filter(cluster_number == 5) %>%
  filter(((isoform_rank1 == 2) & (stringr::str_count(tissue_list1, fixed("Brain")) >= 6) & (isoform_rank2 == 1)) | 
         ((isoform_rank2 == 2) & (stringr::str_count(tissue_list2, fixed("Brain")) >= 6) & (isoform_rank1 == 1))) %>%
  arrange(humantfs_gene_name) %>%
  pull(ensembl_gene_id)

ha = rowAnnotation(target.tfs = as.character(rownames(heatmap.df) %in% cl5.pp.ids),
                   col = list(target.tfs = c("TRUE" = "red", "FALSE" = "blue")))

h = Heatmap(heatmap.df, 
            name = "Isoform ranks", 
            clustering_distance_columns = function(x, y) hamming_dist(x, y),
            clustering_distance_rows = function(x, y) hamming_dist(x, y),
            col = cluster.colors,
            row_labels = rep("", nrow(heatmap.df)),
            row_split = 7, # 11
            column_split = 5) + # 4
  ha

pdf(file = "data/results/clustered_switch_heatmap_2_tissues5_cluster5_pp_marked.pdf",
    width = 10,
    height = 15)
h
dev.off()
```

We already described the switch produced by TFEB. So, let us describe other switches:

- CAMTA2

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "CAMTA2"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "CAMTA2")
```

So, the only switch is a "Catalytic activity presence/absence," it happens between (a) the brain and the pituitary gland and (b) many non-brain tissues plus substantia nigra, and in the brain CAMTA2's major isoform loses the catalytic domain (IPR027417, "P-loop containing nucleoside triphosphate hydrolase") but gains one more protein-binding domain.

CAMTA2 (https://www.uniprot.org/uniprot/O94983) is a transcription activator.

CAMTA2 is not in my compendium, and I could not find any experimental characterisation of its major isoform switch that we observe.

- JAZF1

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "JAZF1"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "JAZF1")
```

So, the only switch is a "Change in DNA binding," it happens between (a) the brain plus liver and (b) many non-brain tissues, and the switching major isoforms differ by one C2H2 ZF domain, hence, possibly having different DNA binding specificity and consequently, different target genes.

JAZF1 (https://www.uniprot.org/uniprot/Q86VZ6) "acts as a transcriptional corepressor of orphan nuclear receptor NR2C2," among other functions.

JAZF1 is not in my compendium, and I could not find any experimental characterisation of its major isoform switches that we observe.

- MEF2A

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "MEF2A"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "MEF2A")
```

So, the only switch is a "Change in protein binding," it happens between (a) the brain and muscle tissues and (b) many non-brain tissues plus hypothalamus, spinal cord and substantia nigra. Hence, MEF2A may have different specificity or stability of binding of its partners in the most part of the brain, in comparison to other tissues.

MEF2A (https://www.uniprot.org/uniprot/Q02078) is a transcriptional activator that regulates many muscle-specific genes. Interestingly, its major isoform in muscle tissues and most part of the brain is the same.

Some MEF2 proteins are in my compendium, but I could not find any experimental characterisation of MEF2A's major isoform switch that we observe. 

- ZNF605

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF605"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF605")
```

So, the only switch is a "Change in transcription repression," it happens between (a) the brain, ovary and testis and (b) many non-brain tissues, and the difference between the switching major isoforms lies in the transcription repression (KRAB) domain. Hence, ZNF605 could have one repression mode (specificity or strength) in the brain, ovary and testis, but a different mode in other non-brain tissues.

ZNF605 (https://www.uniprot.org/uniprot/Q86T29) is not a well-studied TF.

ZNF605 is not in my compendium, and I could not find any experimental characterisation of the switch that we observe.

- CEBPG

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "CEBPG"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "CEBPG")
```

So, the only switch is a "Change in UTRs," it happens between (a) the brain, pituitary gland, heart and liver and (b) many non-brain tissues, and the switching major isoforms are different only in UTRs, which means that the mRNAs may have different localisation, stability or translation efficiency.

CEBPG (https://www.uniprot.org/uniprot/P53567) "binds to the promoter and the enhancer of the immunoglobulin heavy chain." Additionally, CEBPG "acts as a negative regulator of other C/EBP members transcriptional activities by forming heterodimers (37,38). Studies have shown that C/EBP promotes cell proliferation and inhibits senescence by interacting with C/EBP (39). C/EBP is also crucial for cellular redox homeostasis and attenuates cellular stress by forming heterodimer with ATF4, a bZIP transcription factor (40)" [Change et al., 2018](https://academic.oup.com/nar/article/46/12/5996/4992652).

CEBPG is in my compendium, and [Chang et al., 2018](https://academic.oup.com/nar/article/46/12/5996/4992652) showed that CEBPG changes the length of its 3'-UTR upon mTORC1 activation. Hence, a study of UTRs of the switching CEBPG major isoforms is necessary to determine if this switch differentially regulates CEBPG expression depending on the mTORC1 activity in the brain and other tissue.

- ZNF592

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF592"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF592")
```

So, the only switch is a "Change in UTRs," it happens between (a) a part of the brain, liver, pancreas, prostate and spleen and (b) many other non-brain tissues. Hence, the localisation, stability or translational efficiency of ZNF592 major isoform mRNAs could be different between the switching tissue sets.

ZNF592 (https://www.uniprot.org/uniprot/Q92610) is not a well-studied TF.

ZNF592 is not in my compendium, and I could not find any experimental characterisation of the major isoform switch that we observe.

- ZC3H8

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZC3H8"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZC3H8")
```

So, the only switch is a "Change in UTRs," it happens between (a) the brain and select non-brain tissues and (b) many other non-brain tissues. Hence, the localisation, stability or translational efficiency of ZC3H8 major isoform mRNAs could be different between the switching tissue sets.

ZC3H8 (https://www.uniprot.org/uniprot/Q8N5P1) is a transcriptional repressor of GATA3, a regulator of snRNA transcription and "induces thymocyte apoptosis when overexpressed."

ZC3H8 is not in my compendium, and I could not find any experimental characterisation of its major isoform switch that we observe.

- ZBTB45

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZBTB45"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZBTB45")
```

So, the only switch is a "Change in UTRs," it happens between the brain and non-brain tissues. Hence, the localisation, stability or translational efficiency of ZBTB45 major isoform mRNAs could be different between the switching tissue sets.

ZBTB45 (https://www.uniprot.org/uniprot/Q96K62) "may play a role in glial cell differentiation." 

ZBTB45 is not in my compendium, and I could not find any experimental characterisation of the switch that we observe.

- ZNF202

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF202"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF202")
```

So, the only switch is a "Change in UTRs," which happens between (a) the brain and four non-brain tissues and (b) many other non-brain tissues. Hence, the localisation, stability or translational efficiency of ZNF202 major isoform mRNAs could be different between the switching tissue sets.

ZNF202 (https://www.uniprot.org/uniprot/O95125) is a "transcriptional repressor that binds to elements found predominantly in genes that participate in lipid metabolism. Among its targets are structural components of lipoprotein particles (apolipoproteins AIV, CIII, and E), enzymes involved in lipid processing (lipoprotein lipase, lecithin cholesteryl ester transferase), transporters involved in lipid homeostasis (ABCA1, ABCG1), and several genes involved in processes related to energy metabolism and vascular disease." Hence, this switch could take part in a putative differential regulation of lipid metabolism between the brain and the majority of non-brain tissues.

ZNF202 is in my compendium, but I did not find and experimental characterisation of the switch that we observe.

- ADNP

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ADNP"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ADNP")
```

So, the only switch is a "Change in UTRs," which happens between (a) the brain, liver, pancreas and stomach and (b) many non-brain tissues. Hence, the localisation, stability or translational efficiency of ADNP major isoform mRNAs could be different between the switching tissue sets.

ADNP (https://www.uniprot.org/uniprot/Q9H2P0) "may mediate some of the neuroprotective peptide VIP-associated effects involving normal growth and cancer proliferation. When isolated from the sequence, neuroprotective peptide (NAP) provides neuroprotection against the amyloid-beta peptide." Hence, the observed switch could regulate the brain- and non-brain-specific functions of this TF.

ADNP is not in my compendium, and I could not find any experimental characterisation of the switch that we observe.

- SREBF1

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "SREBF1"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "SREBF1")
```

So, the only switch is a "Change outside of domains," which happens between (a) the brain and liver and (b) many non-brain tissues. Hence, if the change happens in an IDR or a domain not matched in our dataset, it could differentially affect the function of this TF in the brain and non-brain tissues.

The processed form of SREBF1 (https://www.uniprot.org/uniprot/P36956) is a "key transcription factor that regulates expression of genes involved in cholesterol biosynthesis and lipid homeostasis." ("Low sterol concentrations promote processing of this form, releasing the transcription factor form that translocates into the nucleus and activates transcription of genes involved in cholesterol biosynthesis and lipid homeostasis (By similarity)".) There are also two isoforms that do not require processing and hence are consitutively active transcription activators of lipogenic genes. However, SREBF1 also has a membrane attachment domain and a transcription activation domain [Hua et al., 1995](https://www.sciencedirect.com/science/article/pii/088875439580009B) which is not matched in our dataset. Consequently, the difference between the two isoforms could be in that domains. Interestingly, [Toledo et al., 2020](https://www.sciencedirect.com/science/article/pii/S2211124720305507) demonstrated that Srebf1 is required for midbrain neurogenesis.

SREBF1 is in my compendium, but I could not find any experimental charactersation of the switch that we observe.

- MEIS1

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "MEIS1"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "MEIS1")
```

So, the only switch is a "Change outside of domains," which happens between (a) the brain and heart and (b) many non-brain tissues. Hence, if the change happens in an IDR or a domain not matched in our dataset, it could differentially affect the function of this TF in the brain and non-brain tissues.

MEIS1 (https://www.uniprot.org/uniprot/O00470) "acts as a transcriptional regulator of PAX6. Acts as a transcriptional activator of PF4 in complex with PBX1 or PBX2. Required for hematopoiesis, megakaryocyte lineage development and vascular patterning. May function as a cofactor for HOXA7 and HOXA9 in the induction of myeloid leukemias." So, this TF pertains to hetamopoesis and blood vessels, and it is interesting that the switch happens not between arteries and other tissues, but between the brain plus heart and non-brain tissues, and it is actually known to play some role in the brain [Xiong et al., 2009](https://academic.oup.com/hmg/article/18/6/1065/613349).  

MEIS1 is in my compendium, but I could not find any experimental characterisation of the switch that we observe.

- ZNF384

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF384"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF384")
```

Among the five switches, let us concentrate on the DBD+/+ rank = 1 / rank = 2 switch (ENST00000319770 / ENST00000355772) as it is the most tissue-populated one. It is a "Change outside of domains," and it happens between the brain and non-brain tissues. Hence, if the change happens in an IDR or a domain not matched in our dataset, it could differentially affect the function of this TF in the brain and non-brain tissues.

ZNF384 (https://www.uniprot.org/uniprot/Q8TF68) "seems to bind and regulate the promoters of MMP1, MMP3, MMP7 and COL1A1 (By similarity)." It also recruits Ku proteins to DNA double strand breaks during the classical non-homologous end-joining [Singh et al., 2021](https://www.nature.com/articles/s41467-021-26691-0).

ZNF384 is not in my compendium, and I could not find any experimental characterisation of the switch that we observe.

- ZNF415

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF415"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF415")
```

So, the only switch is a "Reverse regulation" which happens between (a) the brain and ovary and (b) many non-brain tissues. The difference between the switching major isoforms is in the transcription repression domain (KRAB) which is present in the major isoform expressed in non-brain tissues but absent from the other major isoform expressed in the brain and ovary. Consequently, in the brain and ovary this TF may not be a transcription repressor anymore.

ZNF415 (https://www.uniprot.org/uniprot/Q09FC8) is "involved in transcriptional regulation. Transcriptional activity differed among the various isoforms. All isoforms except isoform 3 seem to suppresses the transcriptional activities of AP-1 and p53/TP53."

ZNF415 is in my compendium, but I could not find any experimental characterisation of the switch that we observe.

- ZNF529

```{r, include=T, fig.height=5, fig.width=10}
gene_name = "ZNF529"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

Switches of this TF:

```{r, include=T}
isoform.switch.master.table.annot.with_functions %>%
  filter(humantfs_gene_name == "ZNF529")
```

So, the only switch is "Reverse regulation," which happens between (a) the brain and stomach and (b) many non-brain tissues. The difference between the switching major isoforms is in the transcription repression domain (KRAB) which is present in the major isoform expressed in non-brain tissues but absent from the other major isoform expressed in the brain and stomach. Consequently, in the brain and stomach this TF may not be a transcription repressor anymore.

ZNF529 (https://www.uniprot.org/uniprot/Q6P280) does not have a good description in UniProt. It also is not described in literature, except for a thesis [Wolford, 2021](https://deepblue.lib.umich.edu/handle/2027.42/169810), where the author mentions that ZNF529 is a candidate drug target for dyslipidemia and cardiovascular diseases.

ZNF529 is not in my compendium, and I could find any experimental characterisation of the switch that we observe.

Overall, only 15 out of 34 TFs considered above in clusters 1, 5 and 7 are, to some extent, experimentally studied, and the switch of only 1 TF (IRF3) is described in the literature. Consequently, there is a vast room for exploration of these brain-non-brain switches!

Intriguingly, KRAB zinc fnger TFs ZNF226, ZNF721, ZNF571, ZNF470, ZNF415 and ZNF529 could differentially repress their targets, including transposons, between the brain and non-brain tissues. In a review [Grassi et al., 2019](https://www.sciencedirect.com/science/article/pii/S0006899318301136), the authors say that, on the one hand, KRAB zinc fingers repress transposons in the brain and their derepression may lead to brain disorders, but, on the other hand, transposons can serve as transcrption regulation elements, sources of miRNAs, lncRNAs and even proteins. Additionally, [Trizzino et al., 2018](https://link.springer.com/article/10.1186/s12864-018-4850-3) show that the brain is one of a number of tissues that are rather "permissive" for transposon transcription. Hence, the putative differential repression of transposons between the brain and non-brain tissues could still be physiological.

## Linear plots for all interesting examples

A function to make a linear plot of the expression of the top two major isoforms of a TF:

```{r, include=T}
ordered.tissue.list = tissue.names[unlist(column_order(h))]

generate_expression_linear_plot = function(tf.name) {
  top.switch = isoform.switch.master.table.annot.with_functions %>%
    filter(humantfs_gene_name == tf.name) %>%
    filter((isoform_rank1 %in% c(1, -1, 2, -2)) & 
           (isoform_rank2 %in% c(1, -1, 2, -2)))
  
  major.isoform1 = top.switch$ensembl_transcript_id1
  
  major.isoform2 = top.switch$ensembl_transcript_id2
  
  isoform1_rank = ifelse(top.switch$isoform_rank1 %in% c(-1, 1), "1", "2")
  
  isoform2_rank = ifelse(top.switch$isoform_rank2 %in% c(-1, 1), "1", "2")
  
  tf.df = isoform.expression.tf.annot %>%
    filter(ensembl_transcript_id %in% c(major.isoform1, major.isoform2)) %>%
    mutate(dbd_pos = ifelse(n_dbd > 0, "DBD+", "DBD-"))
  
  tf.df.annot = tf.df %>%
    dplyr::select(ensembl_gene_id,
                  humantfs_gene_name,
                  tf_family,
                  ensembl_transcript_id,
                  n_dbd,
                  dbd_pos)
  
  tf.df.expr = tf.df %>%
    dplyr::select(-ensembl_gene_id,
                  -humantfs_gene_name,
                  -tf_family,
                  -n_dbd,
                  -dbd_pos) %>%
    tibble::column_to_rownames("ensembl_transcript_id")
  
  tf.df.expr = tf.df.expr[, ordered.tissue.list]
  
  tf.df.final = tf.df.annot %>%
    bind_cols(tf.df.expr)
  
  p = data.frame(ensembl_transcript_id = c(rep(major.isoform1, length(ordered.tissue.list)), 
                                           rep(major.isoform2, length(ordered.tissue.list))),
                 isoform_rank = c(rep(isoform1_rank, length(ordered.tissue.list)),
                                  rep(isoform2_rank, length(ordered.tissue.list))),
                 tissue_names = rep(ordered.tissue.list, 2),
                 dbd_pos = c(rep(tf.df.annot %>% 
                                   filter(ensembl_transcript_id == major.isoform1) %>% 
                                   pull(dbd_pos), 
                                 length(ordered.tissue.list)), 
                             rep(tf.df.annot %>% 
                                   filter(ensembl_transcript_id == major.isoform2) %>% 
                                   pull(dbd_pos), 
                                 length(ordered.tissue.list))),
             bin_expr = c(as.vector(t(tf.df.expr)[, major.isoform1]),
                          as.vector(t(tf.df.expr)[, major.isoform2]))) %>%
    mutate(tissue_names = factor(tissue_names, levels = ordered.tissue.list),
           dbd_pos = factor(dbd_pos, levels = c("DBD+", "DBD-"))) %>%
    ggplot(aes(x = tissue_names,
               y = bin_expr,
               group = ensembl_transcript_id)) +
      geom_line(aes(colour = dbd_pos, linetype = isoform_rank)) +
      geom_hline(yintercept = 1, size = 0.25) +
      theme_classic() + 
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
            legend.position = "none")
  
ggsave(filename = paste0("data/results/", tf.name, "_top_switch_linear_plot.pdf"),
       plot = p,
       width = 10,
       height = 4)  
}
```

Plot the expression of the major isoforms from the top switch for several groups of interesting TFs:

```{r, include=T}
tf.names1 = c("ZNF197", "ZNF226", "FOXJ3", "MAZ", "NR2F2", "NFIB")

for (tf.name in tf.names1) {
  generate_expression_linear_plot(tf.name)
}

tf.names2 = c("ZNF274", "ZNF12", "ZNF548", "ZNF746", "ZNF76", "ZNF195", "ZNF721")

for (tf.name in tf.names2) {
  generate_expression_linear_plot(tf.name)
}

tf.names3 = c("IRF3", "RARA", "THAP4", "ZNF470", "ZNF571")

for (tf.name in tf.names3) {
  generate_expression_linear_plot(tf.name)
}

tf.names4 = c("CAMTA2", "JAZF1", "MEF2A", "ZNF605", "CEBPG", "ZNF592", "ZC3H8", "ZBTB45", "ZNF202", "ADNP", "SREBF1", "MEIS1", "ZNF384", "TFEB", "ZNF415", "ZNF529")

for (tf.name in tf.names4) {
  generate_expression_linear_plot(tf.name)
}
```

## Representation of well-studied switches in our data

Now we need to check if we expect to see switches between well-studied TF isoforms from Fig. 2 in our data as switches of major TF isoforms and if we indeed see these switches. 

Prepare functions to query TFs:

```{r, include=T}
# Upload the table of gene-transcript-protein correspondence
ensg_enst_ensp = read.delim("data/ensembl99/ensg_enst_ensp_99.tsv")

# Upload the table of all isoforms
ips_domains_ipr_ens99 = read.delim(file = "analysis/scan_interpro/output/scan_results/all_analyses_ens99/protein_fasta_tfs_99.fa.tsv", 
                                   header = F, 
                                   row.names = NULL, 
                                   col.names = c("protein_accession", "seq_md5", "seq_len", 
                                                 "db", "domain_accession", "domain_description",
                                                 "start", "stop", "evalue",
                                                 "status", "run_date", "ipr_accession",
                                                 "ipr_description", "go_terms", "pathways")) %>%
  as_tibble() %>%
  filter(!is.na(ipr_accession) & (ipr_accession != "")) %>%
  mutate(protein_accession = str_extract(protein_accession, "[A-Z0-9]+")) %>%
  left_join(ensg_enst_ensp, by = c("protein_accession" = "ensembl_peptide_id"))

# Upload the table of all isoforms with all matches, with those not included in InterPro
ips_domains_ipr_ens99_all = read.delim(file = "analysis/scan_interpro/output/scan_results/all_analyses_ens99/protein_fasta_tfs_99.fa.tsv", 
                                   header = F, 
                                   row.names = NULL, 
                                   col.names = c("protein_accession", "seq_md5", "seq_len", 
                                                 "db", "domain_accession", "domain_description",
                                                 "start", "stop", "evalue",
                                                 "status", "run_date", "ipr_accession",
                                                 "ipr_description", "go_terms", "pathways")) %>%
  as_tibble() %>%
  mutate(protein_accession = str_extract(protein_accession, "[A-Z0-9]+")) %>%
  left_join(ensg_enst_ensp, by = c("protein_accession" = "ensembl_peptide_id"))

get_id_by_name = function(tf.name) {
  return(tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
           filter(humantfs_gene_name == tf.name) %>%
           pull(ensembl_gene_id) %>%
           unique())
}

list_major_isoforms = function(ensembl.gene.id) {
  isoform.expression.tf.annot.dom.isof %>%
    filter(ensembl_gene_id == ensembl.gene.id) %>%
    left_join(tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl,
              by = c("ensembl_gene_id" = "ensembl_gene_id",
                     "humantfs_gene_name" = "humantfs_gene_name",
                     "ensembl_transcript_id" = "ensembl_transcript_id",
                     "tf_family" = "DBD")) %>%
    dplyr::select(ensembl_gene_id,
                  humantfs_gene_name,
                  tf_family,
                  ensembl_transcript_id,
                  n_dbd,
                  ipr_accession,
                  ipr_description,
                  start,
                  stop,
                  transcript_tsl,
                  transcript_tsl_code,
                  transcript_mane_select,
                  transcript_gencode_basic,
                  transcript_appris,
                  ccds) %>%
    arrange(ensembl_transcript_id,
            as.numeric(start),
            as.numeric(stop))
}

list_all_filtered_isoforms = function(ensembl.gene.id) {
  tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
    filter(ensembl_gene_id == ensembl.gene.id) %>%
    dplyr::select(ensembl_gene_id,
                  humantfs_gene_name,
                  DBD,
                  ensembl_transcript_id,
                  ipr_accession,
                  ipr_description,
                  start,
                  stop,
                  transcript_tsl,
                  transcript_tsl_code,
                  transcript_mane_select,
                  transcript_gencode_basic,
                  transcript_appris,
                  ccds) %>%
    arrange(ensembl_transcript_id, 
            as.numeric(start), 
            as.numeric(stop))
}

list_all_isoforms = function(ensembl.gene.id) {
  ips_domains_ipr_ens99 %>%
    filter(ensembl_gene_id == ensembl.gene.id) %>%
    dplyr::select(ensembl_gene_id,
                  ensembl_transcript_id,
                  ipr_accession,
                  ipr_description,
                  start,
                  stop) %>%
    arrange(ensembl_transcript_id, 
            as.numeric(start), 
            as.numeric(stop))
}

list_all_matches_in_all_isoforms = function(ensembl.gene.id) {
  ips_domains_ipr_ens99_all %>%
    filter(ensembl_gene_id == ensembl.gene.id) %>%
    arrange(ensembl_transcript_id, 
            as.numeric(start), 
            as.numeric(stop))
}

summarise_raw_matches = function(ensembl.gene.id) {
  ips_domains_ipr_ens99 %>%
    filter(ensembl_gene_id == ensembl.gene.id) %>%
    dplyr::select(ensembl_gene_id,
                  ipr_accession,
                  ipr_description) %>%
    distinct() %>%
    arrange(ipr_accession)
}

print_isoform_expression = function(ensembl.gene.id) {
  isoform.expression.tf.annot %>%
    filter(ensembl_gene_id == ensembl.gene.id)
}

print_dbd_sequences = function(ensembl.gene.id) {
  dbd_seq_table %>%
    dplyr::select(ensembl_gene_id, 
                  ensembl_transcript_id,
                  ipr_accession, 
                  ipr_description,
                  dbd_sequence) %>%
    distinct() %>%
    filter(ensembl_gene_id == ensembl.gene.id) %>%
    arrange(ensembl_transcript_id)
}

print_ipr_sequences = function(ensembl.gene.id, ipr.accession) {
  ips_domains_ipr_ens99 %>%
    dplyr::select(ensembl_gene_id, 
                  ensembl_transcript_id, 
                  protein_accession, 
                  ipr_accession, 
                  ipr_description, 
                  start, 
                  stop) %>%
    filter(ensembl_gene_id == ensembl.gene.id) %>%
    filter(ipr_accession == ipr.accession) %>%
    left_join(ens99_pep, by = c("protein_accession" = "protein_accession")) %>%
    mutate(ipr_sequence = stringr::str_sub(protein_sequence, start = start, end = stop)) %>%
    dplyr::select(ensembl_gene_id, 
                  ipr_accession,
                  ipr_description,
                  ipr_sequence) %>%
    distinct()
}
```

1) CREM (Fig. 2A, 2D):

```{r, include=T}
ensembl.gene.id = get_id_by_name("CREM")
list_major_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
print_dbd_sequences(ensembl.gene.id)
```

Fig. 2A, middle (Swap of a DBD for another DBD => Changed specificity of DNA binding):

[Laoide et al., 1993](https://www.embopress.org/doi/abs/10.1002/j.1460-2075.1993.tb05759.x) showed that in some mouse cell lines, Crem coexpresses isoforms, alpha and beta, with two different DBD sequences, which changes CREM's DNA binding specificity. This finding may be relevant to the human biology, because two long isoforms of human CREM, ENST00000354759 and ENST00000342105, have different DBDs and in this way correspond to murine Crem isoforms alpha and beta.

[Laoide et al., 1993](https://www.embopress.org/doi/abs/10.1002/j.1460-2075.1993.tb05759.x) showed that mouse Crem did not swap its isoforms across examined cell lines but rather coexpressed two or three different isoforms in some cell lines. The set of coexpressed isoforms always included a full isoform, with the P site, activation domains and a DBD. The full isoform seemed to be a major one, judging from the bands in their fig. 1B. [Stehle et al., 1993](https://www.nature.com/articles/365314a0) found that in rat, the pineal gland coexpresses ICERs (DBD+ isoforms lacking all other domains) with two different DBDs at night, and at that time it does not express any other Crem isoforms. As in the literature we did not find any evidence of switches between major DBD+ CREM isoforms with two different DBDs in adult tissues, we do not have specific expectations about these switches in our data.

InterPro only annotates a pKID and a bHLH domain in the CREM isoforms, without separate matches for activation domains that surround the pKID domain (P site). Hence, full CREM isoforms in our analysis have only the pKID and the bHLH domains. However, none of the full isoforms becomes a major one in adult human tissues. Instead, the full isoforms tend to be either not expressed or (in the case of ENST00000342105) to be second to major with a close expression value (compare a full isoform ENST00000342105 and a short, DBD-only, isoform ENST00000490511 in testis). Hence, we do not see switches of the full isoforms, but only of partial ones. However, in our data we indeed see that CREM expresses isoforms with two different DBD sequences (compare, for example, isoforms ENST00000344351 and ENST00000342105). Hence, our observations partially match expectations from [Laoide et al., 1993](https://www.embopress.org/doi/abs/10.1002/j.1460-2075.1993.tb05759.x) about expression of CREM isoforms with different DBD sequences.

Fig. 2D (Loss of a transcription activation domain in the presence of a DBD => Transcription repression):

[Molina et al., 1993](https://www.cell.com/cell/pdf/0092-8674(93)90532-U.pdf) found that mouse or rat Crem can lose a transcription activation domain in the presence of a DBD and with this switch change its regulatory function from activation (by a full isoform) to repression (by a short isoform consisting almost exclusively of a DBD and called ICER).

[Molina et al., 1993](https://www.cell.com/cell/pdf/0092-8674(93)90532-U.pdf) showed that ICERs are expressed in several different cell lines. Additionally, [Don & Stelzer, 2002](https://www.sciencedirect.com/science/article/pii/S0303720701006967) demonstrated that Sertoli cells mainly express ICERs, while germ cells mainly express activating isoforms of CREM. Additionally, [Stehle et al., 1993](https://www.nature.com/articles/365314a0) found that in rat, the pineal gland coexpresses ICERs with the two different DBDs at night, and at that time it does not express any other CREM isoforms. Also, in mouse and human, pre-meiotic germ cells express only ICERs, while post-meiotic germ cells switch to expressing only activating CREM isoforms [Peri & Serio, 2014](https://link.springer.com/article/10.1007/BF03343779). As in the literature we did not find any evidence of switches of ICERs as major isoforms in adult tissues, we do not have specific expectations about these switches in our data.

As we mentioned above, the full isoform of CREM does not become the major one in our dataset (although becomes a second to major in testis), so we do not observe a switch between a full DBD+ isoform and an ICER. However, we do have some isoforms that could be ICERs: ENST00000344351, ENST00000460270, ENST00000473940, ENST00000474931, ENST00000488328, ENST00000490511, and they become major in different tissues.

2) FOXP1 (Fig. 2A, right: Change in the DBD sequence => Changed specificity of DNA binding):

```{r, include=T}
ensembl.gene.id = get_id_by_name("FOXP1")
list_major_isoforms(ensembl.gene.id)
print_dbd_sequences(ensembl.gene.id)
print_ipr_sequences(ensembl.gene.id, "IPR001766")
```

[Gabut et al., 2011](https://www.sciencedirect.com/science/article/pii/S0092867411009494) found that human FOXP1 can change the DBD sequence and hence its DNA binding specificity.

[Gabut et al., 2011](https://www.sciencedirect.com/science/article/pii/S0092867411009494) showed that FOXP1 switches an isoform with one DBD sequence for an isoform with another DBD sequence when ESCs start to differentiate. Otherwise, we did not find evidence for FOXP1 major isoform switches across adult human tissues, and hence, we do not have any specific expectations about these switches.

In our data, FOXP1 has only one major isoform in adult human tissues and hence we do not see it switching isoforms. Additionally, all FOXP1 isoforms in our dataset have the same DBD sequence. However, if we take all FOXP1 isoforms annotated in Ensembl v99, we will see different DBD sequences. Hence, FOXP1 isoforms with the other DBD sequences did not make it through our filtering process based on quality of isoform annotation.

3) IKZF1 (Fig. 2A-B):

```{r, include=T}
ensembl.gene.id = get_id_by_name("IKZF1")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
```

Fig. 2A, left (Changing number of DBDs => Changed specificity of DNA binding):

[Hahm et al., 1994](https://journals.asm.org/doi/10.1128/mcb.14.11.7111-7123.1994) found that a difference in the number of DBDs (C2H2 ZFs) between Ikzf1 isoforms in mouse leads to a difference in their binding specificity. This result is relevant to the human biology, because [Molnar et al., 1996](https://www.jimmunol.org/content/jimmunol/156/2/585.full.pdf) found that IKZF1 isoforms are highly conserved between human and mouse.

[Hahm et al., 1994](https://journals.asm.org/doi/10.1128/mcb.14.11.7111-7123.1994) observed two IKZF1 isoforms with different numbers of C2H2 ZFs and different DNA binding specificities coexpressed in B and T lymphoid mouse cell lines. As in the literature we did not find any evidence of switches of IKZF1 major isoforms with different numbers of DBDs in adult tissues, we do not have specific expectations about these switches in our data.

According to our dataset, IKZF1 has only one major isoform in adult human tissues, and hence we do not observe switches of its major isoforms. It also has several isoforms with a smaller number of DBDs (C2H2 ZFs), although these isoforms do not become major in any tissue. In principle, this fact matches expectations from [Hahm et al., 1994](https://journals.asm.org/doi/10.1128/mcb.14.11.7111-7123.1994) about the existence of IKZF1 isoforms with different numbers of DBDs.

Fig. 2B (Loss of a DBD in the presence of a dimerisation domain => Loss of DNA binding):

[Sun et al., 1996](https://www.embopress.org/doi/epdf/10.1002/j.1460-2075.1996.tb00920.x) found that a loss of DBDs in the presence of a dimerisation domain abrogated the DNA binding of the dimer comprised of the DBD- and the DBD+ Ikzf1 isoforms in mouse. This finding is relevant to the human biology, because [Molnar et al., 1996](https://www.jimmunol.org/content/jimmunol/156/2/585.full.pdf) found that IKZF1 isoforms are highly conserved between human and mouse.

[Sun et al., 1996](https://www.embopress.org/doi/epdf/10.1002/j.1460-2075.1996.tb00920.x) found that DBD- isoforms of Ikzf1 are presumably expressed at low levels in lymphocytes and thymocytes. As in the literature we did not find any evidence of switches between major DBD+ and DBD- IKZF1 isoforms in adult tissues, we do not have specific expectations about these switches in our data.

According to our dataset, IKZF1 has only one major isoform in adult human tissues, and hence we do not observe switches of its major isoforms. Additionally, we do not separately annotate a dimerisation domain in IKZF1 isoform. According to [Sun et al., 1996](https://www.embopress.org/doi/epdf/10.1002/j.1460-2075.1996.tb00920.x), it may be the C-terminal zinc finger, but we automatically annotate all C2H2 zinc fingers as DBDs.

4) PAX6 (Fig. 2A, right: Change in the DBD sequence => Changed specificity of DNA binding):

```{r, include=T}
ensembl.gene.id = get_id_by_name("PAX6")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_dbd_sequences(ensembl.gene.id)
```

[Epstein et al., 1994](http://genesdev.cshlp.org/content/8/17/2022) showed that human PAX6 can change its DBD sequence and hence DNA binding specificity.

[Epstein et al., 1994](http://genesdev.cshlp.org/content/8/17/2022) mentioned that two PAX6 isoforms, with and without an insertion in the paired domain, are coexpressed in the developing eye, brain, spinal cord and olfactory epithelium, and that the transcript with the insertion (exon 5a) is expressed at lower levels. Later, [Azuma et al., 2005](https://academic.oup.com/hmg/article/14/6/735/2901065) showed that PAX6 swaps its major isoform during retina development. Quite expectedly, other authors ([Pinson et al., 2005](https://bmcdevbiol.biomedcentral.com/articles/10.1186/1471-213X-5-13); [Zhang et al., 2010](https://www.sciencedirect.com/science/article/pii/S1934590910001724); [Sasamoto et al., 2016](https://www.nature.com/articles/srep20807)) also studied roles of the two PAX6 isoforms in development. As in the literature we did not find any evidence of switches between major PAX6 isoforms in adult tissues, we do not have specific expectations about these switches in our data.

According to our data, PAX6 has only one major isoform in adult human tissues (a full one, with a paired and a homeobox domains), and hence we do not see a switch. It also has several other isoforms, either full or partial (with only the homeobox domain). However, most of them are not expressed in adult tissues. In accord with [Epstein et al., 1994](http://genesdev.cshlp.org/content/8/17/2022), in our data PAX6 isoforms have either one of the two different DBD sequences (see, for example, isoforms ENST00000241001 and ENST00000638914 whose paired domain sequences differ by an insertion of 14 amino acids):

```{r, include=T}
ensembl.gene.id = get_id_by_name("PAX6")

pax6.dbd.short = dbd_seq_table %>%
  dplyr::select(ensembl_gene_id, 
                ensembl_transcript_id,
                ipr_accession, 
                ipr_description,
                dbd_sequence) %>%
  distinct() %>%
  filter(ensembl_gene_id == ensembl.gene.id) %>%
  arrange(ensembl_transcript_id) %>%
  filter(ensembl_transcript_id == "ENST00000241001") %>%
  filter(ipr_accession == "IPR001523") %>%
  pull(dbd_sequence)

pax6.dbd.long = dbd_seq_table %>%
  dplyr::select(ensembl_gene_id, 
                ensembl_transcript_id,
                ipr_accession, 
                ipr_description,
                dbd_sequence) %>%
  distinct() %>%
  filter(ensembl_gene_id == ensembl.gene.id) %>%
  arrange(ensembl_transcript_id) %>%
  filter(ensembl_transcript_id == "ENST00000638914") %>%
  filter(ipr_accession == "IPR001523") %>%
  pull(dbd_sequence)

pax6.dbd.short

pax6.dbd.long
```

```{r, include=T}
cat("SHSGVNQLGGVFVNGRPLPDSTRQKIVELAHSGARPCDISRILQ--------------VSNGCVSKILGRYYETGSIRPRAIGGSKPRVATPEVVSKIAQYKRECPSIFAWEIRDRLLSEGVCTNDNIPSVSSINRVLR\n") # pax6.dbd.short
cat(pax6.dbd.long, "\n")
```

```{r, include=T}
cat("length(pax6.dbd.short) =", nchar(pax6.dbd.short), "\n")
cat("length(pax6.dbd.long)  =", nchar(pax6.dbd.long), "\n")
```

Isoforms from both groups are expressed in cerebellum:

```{r, include=T}
print_isoform_expression(ensembl.gene.id) %>%
  filter(ensembl_transcript_id %in% c("ENST00000241001",
                                      "ENST00000379109",
                                      "ENST00000419022",
                                      "ENST00000638914",
                                      "ENST00000639409"))
```

5) PAX8 (Fig. 2A, right: Change in the DBD sequence => Changed specificity of DNA binding):

```{r, include=T}
ensembl.gene.id = get_id_by_name("PAX8")
```

[Kozmik et al., 1997](http://genesdev.cshlp.org/content/8/17/2022) found that mouse Pax8 can change its DBD sequence and hence its DNA binding specificity. This finding is relevant to the human biology, because the authors of the study also found that human PAX8 had the same splicing event changing the DBD as the murine Pax8.

[Kozmik et al., 1997](http://genesdev.cshlp.org/content/8/17/2022) showed that PAX8 coexpresses two isoforms at the same ratio (5:1) throughout embryogenesis and in adult ovary and kidney. Paired domains in these isoforms differ by an insertion of one serine residue. As in the literature we did not find any evidence of PAX8 major isoform switches in adult human tissues, we do not have any specific expectations about these switches, and we also do not have any specific expectations about PAX8 expression in adult human tissues, except in ovary, where it should coexpress isoforms with both types of the paired domain.

However, according to our data, PAX8 is not expressed in adult ovary. Instead, it is expressed in testis and thyroid, and in testis is has one major isoform, while in thyroid it has another major isoform, and so we can see PAX8 switching major isoforms between these two tissues:

```{r, include=T}
print_isoform_expression(ensembl.gene.id) %>%
  filter(ensembl_transcript_id %in% c("ENST00000348715", 
                                      "ENST00000429538"))
```

But these two isoforms have the same DBD sequence:

```{r, include=T}
(print_dbd_sequences(ensembl.gene.id) %>%
  filter(ensembl_transcript_id == "ENST00000348715") %>%
  pull(dbd_sequence) %>%
  unique()) == 
(print_dbd_sequences(ensembl.gene.id) %>%
  filter(ensembl_transcript_id == "ENST00000429538") %>%
  pull(dbd_sequence) %>%
  unique())
```

Moreover, all PAX8 isoforms in our dataset, whether expressed in adult tissues or not, have the same DBD sequence:

```{r, include=T}
length(print_dbd_sequences(ensembl.gene.id) %>%
  pull(dbd_sequence) %>%
  unique()) == 1
```

This is because no PAX8 isoforms in the Ensembl v99 annotation have a paired domain with an additional serine residue between peptides ILG and RYYE. They either have a full paired domain without this serine residue (see the first two sequences in the table below) or have truncated paired domain sequences. Only isoforms with the first DBD sequence variant from the table below passed our filtering and hence were included in our dataset:

```{r include=T}
print_ipr_sequences(ensembl.gene.id, "IPR001523")
```

Hence, our results do not match expectations from [Kozmik et al., 1997](http://genesdev.cshlp.org/content/8/17/2022) due to the lack of PAX8 isoforms with a serine insertion in Ensembl v99.

6) HOXA1 (Fig. 2B: Loss of a DBD in the presence of a dimerisation domain => Loss of DNA binding):

```{r, include=T}
ensembl.gene.id = get_id_by_name("HOXA1")
list_major_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Fernandez & Gudas, 2008](https://onlinelibrary.wiley.com/doi/full/10.1002/jcb.22023) found that mouse Hoxa1 can lose a DBD in the presence of a dimerisation domain and hence abrogate the DNA binding of the dimer comprised of the DBD- and the DBD+ isoforms. This finding is relevant to the human biology, because [Hong et al., 1995](https://www.sciencedirect.com/science/article/pii/037811199592712G) found that human HOXA1 also has a long DBD+ and a short DBD- isoforms (alghouth, the authors do not mention a dimerization).

[Fernandez & Gudas, 2008](https://onlinelibrary.wiley.com/doi/full/10.1002/jcb.22023) study a DBD+ and a DBD- isoforms of Hoxa1 in F9 tetracarcinoma cells and also mention that both isoforms are detected in mouse embryos. As in the literature we did not find any evidence of HOXA1 major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data. Moreover, in general, HOX genes work in development, so we do not have any specific expectation about HOXA1 expression in adult human tissues.

In our dataset, HOXA1 is not expressed in adult human tissues, and hence we do not see it switching isoforms.

7) IKZF2 (Fig. 2B: Loss of a DBD in the presence of a dimerisation domain => Loss of DNA binding):

```{r, include=T}
ensembl.gene.id = get_id_by_name("IKZF2")
list_major_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Zhao et al., 2016](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0163328) found that human IKZF2 can lose a DBD in the presence of a dimerisation domain and hence abrogate the DNA binding of the dimer comprised of the DBD- and the DBD+ isoforms.

[Zhao et al., 2016](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0163328) showed that IKZF2 has T cell leukemia-specific DBD- isoforms. As in the literature we did not find any evidence of IKZF2 major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data.

In our dataset, IKZF2 is not expressed in adult human tissues, and hence we do not see it switching isoforms.

8) TEAD4 (Fig. 2C: Loss of a DBD in the presence of a transcription activation domain => Transcription repression):

```{r, include=T}
ensembl.gene.id = get_id_by_name("TEAD4")
list_major_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Qi et al., 2016](https://www.nature.com/articles/ncomms11840) found that human TEAD4 can lose a DBD in the presence of a transcription activation domain and in this way change its regulation mode from activation to repression.

[Qi et al., 2016](https://www.nature.com/articles/ncomms11840) showed that TEAD4 coexpresses a DBD+ and a DBD- mRNA and protein isoforms in many adult human tissues. The DBD+ mRNA isoform is always the major one, but the levels of expression of protein DBD+ and DBD- isoforms are comparable in some tissues (with the DBD+ still being major): spleen, skeletal muscle, liver. As in the literature we did not find any evidence of TEAD4 major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data.

In our dataset, we indeed have only one major isoform (ENST00000359864, DBD+) and so do not see TEAD4 switching isoforms across tissues. We also have a DBD- isoform (ENST00000397122) that is coexpressed with DBD+ isoforms in adipose visceral (omentum), esophagus muscularis, heart atrial appendage, heart left ventricle, lung, skeletal muscle, pancreas, skin and thyroid. 

9) ZGPAT (Fig. 2C: Loss of a DBD in the presence of a transcription repression domain => Transcription activation):

```{r, include=T}
ensembl.gene.id = get_id_by_name("ZGPAT")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
list_all_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Yu et al., 2010](https://www.sciencedirect.com/science/article/pii/S0021925820548771) found that human ZGPAT can lose a DBD in the presence of a transcription repression domain and in this way change its regulation mode from repression to activation.

[Yu et al., 2010](https://www.sciencedirect.com/science/article/pii/S0021925820548771) showed that the DBD+ and DBD- isoforms of ZGPAT are coexpressed in several cell lines they studied, the DBD+ isoform being the major one. They also mentioned that in their previous work [Li et al., 2009](https://www.embopress.org/doi/full/10.1038/emboj.2009.211) they found that the shorter (presumably, the DBD-) mRNA isoform of ZGPAT was coexpressed with the DBD+ mRNA isoform in liver and kidneys. As in the literature we did not find any evidence of ZGPAT major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data.

In our dataset, we have only one major ZGPAT isoform (ENST00000355969, DBD+) and so do not see this TF switching its isoforms across tissues. However, we do not have a DBD- isoform with any other domains. This is because the Ensembl v99 annotation did not have such an isoform. 

Of note, all ZGPAT isoforms in our dataset lack the Tudor domain and also the coiled-coil domain that serves as a transcription repression domain in ZGPAT according to [Li et al., 2009](https://www.embopress.org/doi/full/10.1038/emboj.2009.211). This is because neither of the two domains was included into InterPro (i. e., had an InterPro ID) for ZGPAT at the time we scanned all human protein TF isoforms (although both domains matched in ZGPAT isoforms):

```{r, include=T}
ensembl.gene.id = get_id_by_name("ZGPAT")
list_all_matches_in_all_isoforms(ensembl.gene.id)
```

10) ESR1 (Fig. 2D: Loss of a transcription activation domain in the presence of a DBD => Transcription repression):

```{r, include=T}
ensembl.gene.id = get_id_by_name("ESR1")
list_major_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Figtree et al., 2002](https://www.ahajournals.org/doi/10.1161/01.cir.0000043805.11780.f5) found that human ESR1 can lose a transcription activation domain in the presence of a DBD and consequently change its regulation mode from activation to repression.

[Figtree et al., 2002](https://www.ahajournals.org/doi/10.1161/01.cir.0000043805.11780.f5) showed that endothelial cells expressed both the full and the short isoforms of ESR1 (the latter isoform lacking the transcription activation domain). As in the literature we did not find any evidence of ESR1 major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data.

In our dataset, we can see a switch between two major ESR1 isoforms, a full one - ENST00000206249 and a short one - ENST00000427531, the latter lacking two domains. Of note, in our scanning results, we do not have a clear transcription activation domain in ESR1 isoforms.

11) PBX1 (Fig. 2E: Loss of a transcription repression domain in the presence of a DBD => Loss of the transcription repression function):

```{r, include=T}
ensembl.gene.id = get_id_by_name("PBX1")
list_major_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Asahara et al., 1999](https://journals.asm.org/doi/epub/10.1128/MCB.19.12.8219) found that human PBX1 can lose a transcription repression domain and hence lose the ability to repress transcription.

[Asahara et al., 1999](https://journals.asm.org/doi/epub/10.1128/MCB.19.12.8219) mention that the two DBD+ PBX1 isoforms, with and without a transcription repression domain, are switched between the endocrine and the exocrine compartments of the adult pancreas. Additionally, [Linares et al., 2015](https://elifesciences.org/articles/09268) showed that PBX1 switches the two isoforms when ESCs start to differentiate into NPCs. As in the literature we did not find any evidence of PBX1 major isoform switches in adult human tissues (and we do not have data on pancreas compartments), we do not have any specific expectations about these switches in our data.

In our dataset, we have only one major PBX1 isoform and so do not see switches of PBX1 major isoforms between tissues.

12) NFATC1 (Fig. 2F: Change in the sequence of a transcription activation domain in the presence of a DBD => Changed set of activated genes):

```{r, include=T}
ensembl.gene.id = get_id_by_name("NFATC1")
list_major_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Lucena et al., 2015](https://journals.asm.org/doi/10.1128/MCB.00501-15) found that mouse Nfatc1 can change the sequence of its transcription activation domain in the presence of a DBD, which consequently changes the set of genes that it activates. This result is relevant to the human biology, because the authors of the study found the same two isoforms (with two different sequences of the transcription activation domain) in human Burkitt lymphoma samples.

[Sherman et al., 1999](https://www.jimmunol.org/content/162/5/2820.long) showed that NFATC1 isoforms with different transcription activation domains are coexpressed in T cells and mast cells and that either one of these isoforms becomes major depending of the activation status of the cells. As in the literature we did not find any evidence of NFATC1 major isoform switches in adult human tissues (and we do not have data T cells and mast cells), we do not have any specific expectations about these switches in our data.

In our dataset, we do not have matches of the transcription activation domain in NFATC1 isoforms. This is due to the N-terminal and C-terminal transactivation regions of NFATC1 not matching any InterPro entries. However, they may have been partially matched by consensus disordered regions at the N and C termini of some isoforms: 

```{r, include=T}
list_all_matches_in_all_isoforms(ensembl.gene.id)
```

Hence, we do not have NFATC1 isoforms with different transactivation domains and do not see this TF switching them. Nevertheless, in our dataset, NFATC1 has two major isoforms, both with a DBD and a dimerization domain. They may or may not be different in terms of their unmatched transcription activation domains.

13) NFYA (Fig. 2F: Change in the sequence of a transcription activation domain in the presence of a DBD => Changed set of activated genes):

```{r, include=T}
ensembl.gene.id = get_id_by_name("NFYA")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Li et al., 1992](https://www.sciencedirect.com/science/article/pii/S0021925819503775) and [Dolfini et al., 2012](https://stemcellsjournals.onlinelibrary.wiley.com/doi/10.1002/stem.1232) found that mouse Nfya can change the sequence of its transcription activation domain in the presence of a DBD, which consequently changes the set of genes that it activates. This finding is relevant to the human biology, because [Li et al., 1992](https://www.sciencedirect.com/science/article/pii/S0021925819503775) found (see fig. 3 in their paper) that both human NFYA and murine Nfya genes encode for a short isoform (lacking a 28-amino acid optional exon) and a long isoform (containing the optional exon). Hence, we suppose that this is the same 28-amino acid optional peptide that [Dolfini et al., 2012](https://stemcellsjournals.onlinelibrary.wiley.com/doi/10.1002/stem.1232) found to be part of a transcription activation domain in murine Nfya.

[Li et al., 1992](https://www.sciencedirect.com/science/article/pii/S0021925819503775) showed that the long and the short isoforms of NFYA (i. e., isoforms differing by the length of the insertion in the N-terminal transcription activation domain) are switched between different malignant and normal cell types: the long isoform is expressed in fibroblasts, while the short isoform is expressed in the WEHI-22 T lymphoma and a B lymphoma. Additionally, [Dolfini et al., 2012](https://stemcellsjournals.onlinelibrary.wiley.com/doi/10.1002/stem.1232) demonstrated that the short isoform is the major one in ESCs, while the long isoform becomes the major one after ESC differentiation. Moreover, [Basile et al., 2016](https://www.sciencedirect.com/science/article/pii/S1874939916300335) found that myoblasts coexpress the short and the long isofoms, and that the long isoform is the major one. The expression level of both isoforms decreases after differentiation, and adult muscle cells do not express NFYA. As in the literature we did not find any evidence of NFYA major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data.

In our dataset, we have only one major NFYA isoform and so do not see switches of NFYA major isoforms across tissues. Of note, in skeletal muscle this isoform is expressed at roughly the same level as in other tissues (for example, liver). Also, we need to point out that our domain annotation for NFYA isoforms lacks the N-terminal transcription activation domain. This is due to this domain (a Q-rich region) not being matched by any InterPro entry:

```{r, include=T}
list_all_matches_in_all_isoforms(ensembl.gene.id)
```

14) ELF1 (Fig. 2G: Change in the sequence of a transcription repression domain in the presence of a DBD => Changed strength of transcription repression):

```{r, include=T}
ensembl.gene.id = get_id_by_name("ELF1")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Nishiyama et al., 2000](https://www.tandfonline.com/doi/pdf/10.1271/bbb.64.2601) found that rat Elf1 can change the sequence of its transcription repression domain in the presence of a DBD and consequently change the strength of transcription repression. This result is relevant to the human biology, because [Nishiyama et al., 2000](https://www.tandfonline.com/doi/pdf/10.1271/bbb.64.2601) found the same ELF1 isoforms in human.

[Nishiyama et al., 2000](https://www.tandfonline.com/doi/pdf/10.1271/bbb.64.2601) showed that at least two ELF1 isoforms with different N termini were coexpressed in various human tissues and that the samples of the same tissue could (strangely) express different major ELF1 isoforms. As in the literature we did not find any clear evidence of ELF1 major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data. 

In our dataset, we have only one major ELF1 isoform and so do not see switches of ELF1 major isoforms across tissues. 

15) HNF1A (Fig. 2G: Change in the sequence of a transcription activation domain in the presence of a DBD => Changed strength of transcription activation):

```{r, include=T}
ensembl.gene.id = get_id_by_name("HNF1A")
list_major_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Bach & Yaniv, 1993](https://www.embopress.org/doi/abs/10.1002/j.1460-2075.1993.tb06107.x) found that human HNF1A can change the sequence of its transcription activation domain and consequently change the strength of transcription activation.

[Bach & Yaniv, 1993](https://www.embopress.org/doi/abs/10.1002/j.1460-2075.1993.tb06107.x) showed that human liver coexpresses three HNF1A isoforms with different sequences of the C-terminal transcription activation domain. They also demonstrated that the longest isoform is the major one in all tissues that they examined and where the gene was expressed (namely, in liver, kidney, intestine and thymus). Additionally, [Harries et al., 2006](https://academic.oup.com/hmg/article/15/14/2216/2355944) showed that all the three HNF1A isoforms were coexpressed in liver, kidney, pancreas and isolated islets. However, HNF1A switched its major isoforms between two groups of tissues: (1) liver and kidney, where the longest isoform was the major one, and (2) adult pancreas and isolated islets where a shorter isoform became major. Interestingly, they also showed a switch of HNF1A major isoforms between the fetal pancreas (where the longest isoform is the major one) and the adult pancreas (where a shorter isoform is the major one). Hence, from these data we would expect HNF1A switch its isoforms between (1) pancreas and (2) liver, kidney, intestine and thymus.

However, in our dataset, we have only one major HNF1A isoform (ENST00000257555) and so do not see switches of HNF1A major isoforms across tissues. This could be explained either by filtering out by annotation quality of other potentially expressed HNF1A isoforms or by possible inaccuracies in the previous published or current GTEx isoform quantification data. Moreover, ENST00000257555 is the only isoform that is expressed in adult human tissues, and it is not the longest one:

```{r, include=T}
ips_domains_ipr_ens99 %>%
  filter(ensembl_gene_id == ensembl.gene.id) %>%
  filter(ensembl_transcript_id %in% (tf_coding_transcripts_final_ens99_with_fam_names_corrected.with_tsl %>%
                                       filter(ensembl_gene_id == ensembl.gene.id) %>%
                                       pull(ensembl_transcript_id) %>%
                                       unique())) %>%
  dplyr::select(ensembl_gene_id,
                ensembl_transcript_id,
                seq_len) %>%
  arrange(desc(seq_len)) %>%
  distinct()
```

16) PAX5 (Fig. 2G: Change in the sequence of a transcription activation domain in the presence of a DBD => Changed strength of transcription activation):

```{r, include=T}
ensembl.gene.id = get_id_by_name("PAX5")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Robichaud et al., 2004](https://www.sciencedirect.com/science/article/pii/S0021925820677845) found that human PAX5 can change the sequence of its transcription activation domain in the presence of a DBD and consequently change the strength of transcription activation.

[Robichaud et al., 2004](https://www.sciencedirect.com/science/article/pii/S0021925820677845) showed that PAX5 switches its major mRNA isoforms between different normal and malignant human CD19+ B cells. Additionally, [Bharti & Mishra, 2015](https://link.springer.com/article/10.1007/s11010-014-2325-7) demonstrated the presence of four PAX5 isoforms differing by their C termini and by the presence of the paired domain in a normal spleen. Two of the four isoforms differed only by their C termini. As in the literature we did not find any other evidence of PAX5 major isoform switches in adult human tissues (and we do not have data on CD19+ B cells), we do not have any specific expectations about these switches in our data. 

In our dataset, we have only one major PAX5 isoform and hence do not see switches of PAX5 major isoforms across tissues. Apart from the major isoform, we also see various other PAX5 isoforms. Of note, InterProScan did not match any transcription activation domains in our PAX5 isoforms, most probably, because InterPro does not have this domain integrated.

17) NR3C1 (Fig. S2A: Loss of a ligand-binding domain in the presence of a DBD => Switch from transcription activation to transcription repression):

```{r, include=T}
ensembl.gene.id = get_id_by_name("NR3C1")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Oakley et al., 1999](https://www.sciencedirect.com/science/article/pii/S0021925819521950) found that human NR3C1 can lose a ligand-binding domain in the presence of a DBD and consequently switch its regulation mode from activation to repression.

[Hollenberg et al., 1985](https://www.nature.com/articles/318635a0.pdf) showed that HeLa and various lymphoma cells express only the longer, LBD+, isoform. Additionally, ([Bamberger et al., 1995](https://www.jci.org/articles/view/117943/pdf); [de Castro et al., 1996](https://molmed.biomedcentral.com/articles/10.1007/BF03401643); [Oakley et al., 1997](https://academic.oup.com/endo/article/138/11/5028/2991357); [Pujols et al., 1999](https://www.atsjournals.org/doi/full/10.1165/ajrcmb.24.1.4024); [Pujols et al., 2002](https://journals.physiology.org/doi/full/10.1152/ajpcell.00363.2001)) demonstrated that multiple human tissues coexpress the LBD+ and LBD- isoforms, while ([Oakley et al., 1996](https://www.sciencedirect.com/science/article/pii/S0021925817313728); [Pujols et al., 2002](https://journals.physiology.org/doi/full/10.1152/ajpcell.00363.2001)) showed that in several human tissues the LBD+ isoform is the major one. As in the literature we did not find any evidence of NR3C1 major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data.

In our dataset, NR3C1 has two major isoforms, and hence we can see switches of NR3C1 major isoforms across tissues. Both isoforms are LBD+ and have the same domains.

18) IRF5 (Fig. S2B: Loss of a degradation-promoting domain in the presence of a DBD => Transcription activation):

```{r, include=T}
ensembl.gene.id = get_id_by_name("IRF5")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Lazzari et al., 2014](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0103609) found that human IRF5 can lose a degradation-promoting domain in the presence of a DBD and consequently become a stable transcription activator.

[Stone et al., 2013](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0054487) showed that all four IRF5 isoforms with different stability are coexpressed in monocytes both in normal physiology and systemic lupus erythematosus. As in the literature we did not find any evidence of IRF5 major isoform switches in adult human tissues (and we do not have data on immune cells), we do not have any specific expectations about these switches in our data.

In our dataset, IRF5 has just one major isoform and so we do not see switches of IRF5 major isoforms across tissues.

19) TCF3 (Fig. S2C: Loss of a DBD inhibition domain in the presence of a DBD => Gain of DNA binding):

```{r, include=T}
ensembl.gene.id = get_id_by_name("TCF3")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Sun et al., 1991](https://www.sciencedirect.com/science/article/abs/pii/009286749190653G) found that human TCF3 can lose a DBD inhibition domain in the presence of DBD and consequently gain the ability to bind DNA.

[Yamazaki et al., 2018](http://genesdev.cshlp.org/content/32/17-18/1161.full) showed that the hESC, HeLa, 293T and U87 cell lines coexpress both TCF3 isoforms, E12 and E47, that differ in their ability to bind DNA (E12 binds poorly as a homodimer, while E47 binds strongly as a homodimer or a heterodimer with E12). However, while E12 is the major isoform in hESCs, E47 becomes the major isoform in HeLa and U87 cells. Additionally, [Pfurr et al., 2017](https://journals.biologists.com/dev/article/144/21/3917/48181/The-E2A-splice-variant-E47-regulates-the) showed coexpression of E12 and E47 in embryonic neural stem cells (NSCs). As in the literature we did not find any evidence of TCF3 major isoform switches in adult human tissues, we do not have any specific expectations about these switches in our data.

In our dataset, TCF3 has 4 major isoforms, and hence we can see switches of TCF3 major isoforms across tissues. All of these isoforms have only the bHLH domain, because this is the only domain matched by InterProScan.

20) MITF (Fig. S2D: Loss of a DBD activation domain in the presence of a DBD => Unstable DNA binding):

```{r, include=T}
ensembl.gene.id = get_id_by_name("MITF")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Hemesath et al., 1994](http://genesdev.cshlp.org/content/8/22/2770.long) found that mouse Mitf can lose a DBD activation domain in the presence of a DBD, which makes its DNA binding unstable. This result is relevant for the human biology for the following reason. Originally, [Hemesath et al., 1994](http://genesdev.cshlp.org/content/8/22/2770.long) showed that the DBD activation domain is a 6-amino acid peptide located closely upstream to the bHLH and encoded by an alternative exon in murine Mitf (see their fig. 1A). Later, [Amae et al., 1998](https://www.sciencedirect.com/science/article/pii/S0006291X98988386) demonstrated that human MITF encodes for two different isoforms, MITF-A and MITF-M, that, among other differences, have an indel of 6 amino acids (ACIFPT) closely upstream to the bHLH (see their fig. 2). Finally, from [Steingrimsson et al., 1994](https://www.nature.com/articles/ng1194-256.pdf) (see their fig. 5B), one can see that the alternative murine exon encodes for exactly the same peptide (ACIFPT) using the last nucleotide of the previous exon as the start of the first triplet (for alanine). Hence, we suppose that the murine Mitf isoform with the peptide corresponds to the human isoform MITF-M with the same peptide, while the murine Mitf isoform without the peptide corresponds to the human isoform MITF-A which also does not have the peptide.

MITF has several isoforms. Isoform MITF-M has a DBD activation site, while MITF-A/H do not have it [Yasumoto et al., 1998](https://onlinelibrary.wiley.com/doi/epdf/10.1111/j.1600-0749.1998.tb00491.x). [Amae et al., 1998](https://www.sciencedirect.com/science/article/pii/S0006291X98988386) showed that MITF switches major isoforms between (1) melanoma cells where MITF-M is the major isoform and (2) HeLa and A-172 cell lines and kidney primary tissue where MITF-A/H are the major isoforms. Additionally, [Yasumoto et al., 1998](https://onlinelibrary.wiley.com/doi/epdf/10.1111/j.1600-0749.1998.tb00491.x) show that MITF-M is the major isoform in primary melanocytes. [Hershey & Fisher, 2005](https://www.sciencedirect.com/science/article/pii/S037811190400736X) demonstrated that the same switch happens also between melanoma cells and primary human osteoclasts. [Bharti et al., 2008](https://journals.biologists.com/dev/article/135/6/1169/65078/Alternative-promoter-use-in-eye-development-the) showed coexpression of different MITF isoforms in developing retinal pigment epithelial, choroid and retina. As in the literature we did not find any evidence of MITF major isoform switches in adult human tissues (and we do not have data on melanocytes, osteoclasts or eye tissues), we do not have any specific expectations about these switches in our data. 

In our datasets, we have only one major isoform, so we do not see switches of MITF major isoforms across tissues.

21) POU2F2 (Fig. S2E: Loss of a strong transcription activation domain in the presence of a weak transcription repression domain and a DBD => Switch from transcription activation to transcription repression):

```{r, include=T}
ensembl.gene.id = get_id_by_name("POU2F2")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Lillycrop et al., 1994](https://journals.asm.org/doi/epdf/10.1128/mcb.14.11.7633-7642.1994) found that human POU2F2 can lose its strong transcription activation domain in the presence of a weak transcription repression domain and a DBD and consequently switch its regulation mode from activation to repression. 

[Lillycrop & Latchman, 1992](https://www.sciencedirect.com/science/article/pii/S002192581973991X) showed that POU2F2 switched its major isoform between B cells and neuronal cells. As in the literature we did not find any evidence of POU2F2 major isoform switches in other adult human tissues (and we do not have B cells), we do not have any specific expectations about these switches in our data.

In our dataset, POU2F2 has one major isoform, and hence we do not see switches of POU2F2 major isoforms across tissues.

22) PGR (Fig. S2F: Loss of a transcription activation domain in the presence of other transcription activation domains and a DBD => Attenuated transcription activation):

```{r, include=T}
ensembl.gene.id = get_id_by_name("PGR")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Takimoto et al., 2003](https://www.sciencedirect.com/science/article/pii/S0960076003001973) found that human PGR can lose a transcription activation domain in the presence of other transcription activation domains and a DBD, which weakens its transcription activation function.

[Kaya et al., 2015](https://academic.oup.com/mend/article/29/6/882/2556297) showed that PGR switches its major isoforms between uterine glands and uterine stroma. Additionally, [Wreczycka et al., 2019](https://academic.oup.com/biolreprod/article/74/1/13/2666759) demonstrated that PGR switches its major isoforms between early and late first-trimester trophoblast cells. As in the literature we did not find any evidence of PGR major isoform switches in adult human tissues (and we do not have stratified uterine subtissues or trophoblasts), we do not have any specific expectations about these switches in our data. 

In our dataset, PGR has two major isoforms, and so we see switches of PGR major isoforms across tissues. Interestingly, one of these isoforms does not have a transcription regulation domain. Also, of note, InterPro matches only one transcription regulation domain in PGR isoforms that we study.

23) TCF4 (Fig. S2F: Loss of a transcription activation domain in the presence of other transcription activation domains and a DBD => Attenuated transcription activation):

```{r, include=T}
ensembl.gene.id = get_id_by_name("TCF4")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
list_all_matches_in_all_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Sepp et al., 2011](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0022138#pone-0022138-g003) found that human TCF4 can lose a transcription activation domain in the presence of other transcription activation domains and a DBD, which weakens its transcription activation function.

[Sepp et al., 2011](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0022138#pone-0022138-g003) showed that TCF4 coexpresses many different isoforms across many adult human tissues and that some isoforms are expressed more broadly then others, but they do not comment explicitly on major mRNA isoforms. However, they suggest the existence of switches of major protein isoforms between the following groups of tissues: (a) frontal cortex, hippocampus and cerebellum (where the major isoform should have only one transcription activation domain), (b) testis (where the major isoform should have two transcription activation domains), and (c) lung (where the major isoform again should contain only one transcription activation domain). Additionally, [Grajkowska et al., 2016](https://www.sciencedirect.com/science/article/pii/S1074761316304782) demonstrated that two isoforms of TCF4, a longer one with all transcription activation domains and a shorter one lacking one transcription activation domain, are coexpressed in interferon-producing plasmacytoid dendritic cells, but the shorter isoform becomes the major one in B cells and in two types of classical dendritic cells where the longer isoform is not expressed. Hence, in our data we would expect TCF4 to switch its major isoforms between the tissue groups (a), (b) and (c) defined above.

However, in our dataset, we have only one TCF4 major isoform (ENST00000398339), and so do not see any switches of TCF4 major isoforms across adult human tissues. This is because, according to GTEx, TCF4 expresses only one isoform in only one adult human tissue (testis). In our data, this isoform has only a bHLH, because InterPro matches only the bHLH domain but not the transcription activation domains in TCF4 isoforms that we study. Of note, isoform ENST00000398339 has several predicted disordered regions some of which could be the transcription activation domains. 

24) HNF4A (Fig. S2G: Exchange of a DBD inhibition domain for a transcription activation domain in the presence of a DBD => Transcription activation):

```{r, include=T}
ensembl.gene.id = get_id_by_name("HNF4A")
list_major_isoforms(ensembl.gene.id)
list_all_filtered_isoforms(ensembl.gene.id)
summarise_raw_matches(ensembl.gene.id)
print_isoform_expression(ensembl.gene.id)
```

[Lambert et al., 2020](https://www.mcponline.org/article/S1535-9476(20)35005-2/fulltext) found that human HNF4A can exchange a DBD inhibition domain for a transcription activation domain in the presence of a DBD and consequently become a transcription activator.

[Lambert et al., 2020](https://www.mcponline.org/article/S1535-9476(20)35005-2/fulltext) suggest that three isoforms transcribed from the promoter P1 (alpha1, alpha2 and alpha3) have a transcription activation domain, while another three isoforms transcribed from the same promoter (alpha4, alpha5 and alpha6), due to alternative splicing, swap this domain for a DBD inhibition domain. As this work is quite recent, literature still does not have data on differential expression of the first trio of isoforms versus the second one in human tissues. Previous literature either compared isoforms transcribed from the first and the second promoters ([Vuong, 2014](https://escholarship.org/uc/item/0gb1j7p3); [Chellappa et al., 2016](https://elifesciences.org/articles/10903)) or, in addition, did compare different isoforms produced from promoter P1 but the N terminus of these isoforms was shown differently ([Harries et al., 2008](https://diabetes.diabetesjournals.org/content/57/6/1745)). Hence, we do not have any specific expectations about HNF4A major isoform switches across adult human tissues.

In our dataset, HNF4A has two major isoforms, so we see switches of HNF4A major isoforms across human tissues. These isoforms have the same two domains (a DBD and an LBD). 

## Interpretation of the check results

For the majority of these TFs (22 out of 24), we do not expect to observe switches of their major isoforms between adult human tissues, because, based on literature, these TFs instead switch isoforms:

1) During cell differentiation or, more generally, in development (FOXP1, PBX1, NFYA, PAX6, HOXA1, TCF3(?), PGR); 

2) Between cancer cells and/or tissues that are not in our dataset (PBX1, NFYA, PAX5, MITF, POU2F2, CREM, PGR); 

3) Between different conditions such as before and after meiosis in the case of germ cells (CREM) or upon activation in the case of immune cells (NFATC1).

Otherwise, in literature we could not find evidence of major isoform switches for some of these TFs (IKZF1, IKZF2, ELF1, IRF5, ESR1, TEAD4, ZGPAT, PAX8, NR3C1). Finally, in the case of HNF4A, there were discrepancies in the representation of the domain structure of some isoforms in literature, and hence we did not know exactly which isoforms are expressed in which cells.

However, for 8 of these TFs (CREM, PAX8, ESR1, NFATC1, NR3C1, TCF3, PGR, HNF4A) we observed switches of their major isoforms between adult human tissues, although, these isoforms were different from the ones described in examples from literature.

For the other two TFs, HNF1A and TCF4, we expected to observe switches of their major isoforms between some adult human tissues. However, we did not observe any switches, because, according to the isoform quantification from GTEx and our isoform filtering by the quality of their annotation in Ensembl v99, these TFs have only one isoform expressed in adult human tissues.

Importantly, our analysis is limited by the fact that some types of domains are rarely described as such in InterPro or are rarely included in this database. For example, the coiled-coil transcription repression domain in ZGPAT, although matched as a coiled-coil region, does not have an InterPro ID and is not described as a transcription repression domain. Additionally, isoforms of NFYA, PAX5, ESR1, NFATC1 and TCF4 do not have matches of transcription activation domains, although these TFs are known to have these domains. Finally, TCF3 isoforms do not have matches of a DBD inhibition domain known to be present in this TF. 

Overall, in our data we do not observe switches of the isoforms described in the well-studied examples, because we consider only switches of major isoforms between adult human tissues, while the described isoforms may not become major or may switch in different contexts. Additionally, some known domains are not matched by InterPro, which precludes us from observing possible switches in other cases. However, instead, in some cases we observe switches of other isoforms. Hence, in general, if the well-studied alternative isoforms either do not take part in switches or do not switch as major ones between adult tissues, then our analysis may predict a considerable number of novel switches.

## DBD sequence changes in switches

Find TFs that change DBD sequences in the switches of their one-DBD major isoforms:

```{r, include=T}
isoform.switch.master.table.annot.with_functions.dbd_seq = isoform.switch.master.table.annot.with_functions %>% 
  filter((n_dbd1 == 1) & (n_dbd2 == 1)) %>%
  filter(dbd_seq_list1 != dbd_seq_list2) %>%
  arrange(tf_family)

isoform.switch.master.table.annot.with_functions.dbd_seq %>%
  dplyr::select(ensembl_gene_id, humantfs_gene_name, tf_family, cluster_number, ensembl_transcript_id1, ensembl_transcript_id2)
```

Hence, the following number of TFs change their DBD sequences in the switches of major one-DBD isoforms:

```{r, include=T}
isoform.switch.master.table.annot.with_functions.dbd_seq %>%
  pull(ensembl_gene_id) %>%
  unique() %>%
  length()
```

From these, select TFs whose structures with DNA are included in NPIDB (version from 27/12/2021):

1) ERF: 7jsa (reduced form - as a dimer, Cys72 is in a reduced state)
2) ERF: 7jsl (oxidized form - as a tetramer via Cys72-Cys72)
3) TCF3: 2ypb (E47 isoform)
4) MBD2: 7mwm (methylated DNA)
5) MBD2: 6c1u (methylated DNA)
6) MBD2: 6c1a (methylated DNA)
7) MBD2: 6cnp (methylated CpG island)
8) PGR: 2c7a (as a dimer with a zinc molecule)
9) PITX2: 2lkx
10) RARB: 5uan (as a heterodimer with RXR-alpha)
11) NFATC1: 1a66

We put the lists of protein-DNA contacts into `data/results/npidb_tf_models_2021_12_27`.

Hence, we have protein-DNA structures for 7 TFs for further analysis.

We downloaded FASTA files, corresponding to the structures, from PDB into `data/results/tf_dbd_sequences/chain_sequences` and manually curated them to shorten headers and select only protein sequences of our 7 TFs of interest. We put the curated FASTA files into `data/results/tf_dbd_sequences/chain_sequences_curated` directory.

Next, we added DBD and full isoform sequences to the FASTA files with the corresponding TF names:

```{r, include=T}
library(seqinr)

path.to.fasta = "data/results/tf_dbd_sequences"

fasta.file.list = list.files(file.path(path.to.fasta, "chain_sequences_curated"))

tf.ids = c("ENSG00000105722", # ERF
           "ENSG00000071564", # TCF3
           "ENSG00000134046", # MBD2
           "ENSG00000082175", # PGR
           "ENSG00000077092", # RARB
           "ENSG00000131196", # NFATC1
           "ENSG00000164093") # PITX2
  
for (tf.id in tf.ids) {
  tf.name = isoform.switch.master.table.annot.with_functions.dbd_seq %>%
    filter(ensembl_gene_id == tf.id) %>%
    pull(humantfs_gene_name) %>%
    unique()
  
  tf.isoform.ids = c(isoform.switch.master.table.annot.with_functions.dbd_seq %>%
                       filter(ensembl_gene_id == tf.id) %>%
                       pull(ensembl_transcript_id1),
                     isoform.switch.master.table.annot.with_functions.dbd_seq %>%
                       filter(ensembl_gene_id == tf.id) %>%
                       pull(ensembl_transcript_id2))
  
  tf.dbd.sequences = c(isoform.switch.master.table.annot.with_functions.dbd_seq %>%
                         filter(ensembl_gene_id == tf.id) %>%
                         pull(dbd_seq_list1),
                       isoform.switch.master.table.annot.with_functions.dbd_seq %>%
                         filter(ensembl_gene_id == tf.id) %>%
                         pull(dbd_seq_list2))
  
  tf.isoform.sequences = c(isoform.switch.master.table.annot.with_functions.dbd_seq %>%
                             filter(ensembl_gene_id == tf.id) %>%
                             pull(protein_sequence1),
                           isoform.switch.master.table.annot.with_functions.dbd_seq %>%
                             filter(ensembl_gene_id == tf.id) %>%
                             pull(protein_sequence2))
  
  tf_df = data.frame(ensembl_transcript_id = tf.isoform.ids,
                     dbd_sequences = tf.dbd.sequences,
                     isoform_sequences = tf.isoform.sequences) %>%
    distinct()
  
  isoform_headers = c(stringr::str_c(tf.name, "_", tf_df$ensembl_transcript_id),
                      stringr::str_c(tf.name, "_", tf_df$ensembl_transcript_id, "_dbd"))
  
  isoform_sequences_whole = tf_df$isoform_sequences
  
  isoform_sequences_dbds = tf_df$dbd_sequences
  
  all_sequences = c(isoform_sequences_whole,
                    isoform_sequences_dbds)
  
  dir.create(file.path(path.to.fasta, "all_sequences"))
  
  for (fasta.file in fasta.file.list[stringr::str_detect(fasta.file.list, paste0(tf.name, "_"))]) {
    combined.fasta.name = paste0(unlist(stringr::str_split(fasta.file, fixed(".")))[1], "_with_isoforms.fa")
    file.copy(file.path(path.to.fasta, "chain_sequences_curated", fasta.file), 
              file.path(path.to.fasta, "all_sequences", combined.fasta.name))
    write.fasta(as.list(all_sequences),
                as.list(isoform_headers),
                file.path(path.to.fasta, "all_sequences", combined.fasta.name),
                open = "a")
  }
}

detach("package:seqinr", unload = TRUE)
```

Do multiple sequence alignments of DBD sequences of TF isoforms from our data and the corresponding protein chains from the structures:

```{r, include=T}
library(msa)

path_to_fasta = "data/results/tf_dbd_sequences/all_sequences"

path_to_msa = "data/results/msa_results"

dir.create(file.path(path_to_msa))

for (fasta_file in list.files(path_to_fasta)) {
  m = msa(inputSeqs = file.path(path_to_fasta, fasta_file),
          type = "protein",
          method = "ClustalOmega",
          order = "input")
  
  m_name = unlist(stringr::str_split(fasta_file, pattern = fixed(".")))[1]
  
  m_df = as.data.frame(m@unmasked)
  
  write.table(m_df, 
              file.path(path_to_msa, 
                        paste0(m_name, "_msa.txt")),
              row.names = F,
              col.names = F,
              quote = F)
}
detach("package:msa", unload = TRUE)
```

We marked specifically and unspecifically contacting amino acids according to the highlighting by NPIDB on each model page. Specific contacts are those made to nucleotide bases, while unspecific ones are made to the sugar-phosphate backbone). We used only H-bond lists.

The alignments show that switching isoforms of MBD2 and TCF3 actually have the same DBD sequence, but for some reason the boundaries of the DBD match differ by one amino acid. In addition, In PITX2 isoforms, DBD matches are very "sensitive" to the changes in the putative flanks of the DBD, so we exclude it either. Hence, we  proceed with the four remaining TFs (ERF, NFATC1, RARB, PGR):

1) ERF:

a) Dimer: Isoform ERF-202 loses a big part of its DBD (along with the whole N-terminal part of the protein) but still retains amino acids that facilitate DNA binding specificity. Hence, if ERF-202 still folds so that the retained amino acids can bind DNA in the same way they did in the longer isoform, then ERF-202 retains the same DNA binding specificity as ERF-201 which has the full DBD (an Ets domain, a winged-helix type of a DBD). 

```{r, include=T}
gene_name = "ERF"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

This switch happens between two different parts of the brain (caudate and putamen) and many other non-brain tissues plus cerebellum. The switch belongs to the cluster 5. 

b) Tetramer: In the tetrametic configuration, an ERF molecule has three additional amino acids that bind to DNA. Two of them, a specific and a non-specific, are retained in both isoforms, while the third one is also non-specific and is lost in ERF-202. Hence, tetrameric DNA binding of ERF also retains specificity in this switch.

2) RARB:

Isoform RARB-204 loses all amino acids that confer DNA binding specificity of its DBD (a nuclear receptor-type zinc finger) and hence, if the short retained part of the DBD still binds DNA, it should bind unspecifically or have a different specificity. 

```{r, include=T}
gene_name = "RARB"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

The switch between RARB-201 and RARB-204 occurs between (a) putamen, caudate, nucleus accumbens and spleen, and (b) many non-brain tissues. Hence, the shorter DBD presumably facilitates different DNA binding properties of RARB in the basal ganglia and spleen. Additionally, RARB usually binds DNA as a heterodimer, and hence its changed DNA binding specificity could affect the DNA binding of its partner TFs. This switch belongs to cluster 3.

To the best of our knowledge, these changes in the DBDs of ERF and RARB and their possible functional consequences have not been described in literature.

3) PGR:

Isoform PGR-207 loses almost a half of the full DBD sequence, along with the whole N-terminal part of the protein. With this, it loses two out of the three amino acids that confer DNA binding specificity but retains the majority of its "non-specific" residues. Hence, if the truncated DBD still binds DNA, it probably binds it unspecifically or with a different specificity.

```{r, include=T}
gene_name = "PGR"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

The switch occurs between testis and many other non-brain tissues, including ovary, vagina, uterus and prostate. This switch belongs to cluster 3.

4) NFATC1:

Isoform NFATC-207 loses a long N-terminal part of the full DBD sequence, along with the whole N-terminal part of the protein. With this, it loses three out of the four amino acids that confer DNA binding specificity but retains the majority of its "non-specific" residues. Hence, if the truncated DBD still binds DNA, it probably binds it unspecifically or with a different specificity.

```{r, include=T}
gene_name = "NFATC1"
pdf(file = paste0("data/results/", gene_name, "_switch_heatmap.pdf"), 
    height = 5, width = 10)
generate_expression_heatmap(isoform.expression.tf.annot.bin %>%
                              filter(humantfs_gene_name == gene_name),
                            isoform.expression.tf.annot.ranks %>%
                              filter(humantfs_gene_name == gene_name))
dev.off()
```

The switch occurs between two groups of non-brain tissues and belongs to cluster 3.
